// @statoscope/stats-validator-reporter-stats-report/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const open_1 = __importDefault(require("open"));
const json_ext_1 = require("@discoveryjs/json-ext");
const utils_1 = require("@statoscope/report-writer/dist/utils");
const package_json_1 = __importDefault(require("@statoscope/stats/package.json"));
const generator_1 = __importDefault(require("@statoscope/stats-extension-stats-validation-result/dist/generator"));
const normalizeCompilation_1 = __importDefault(require("@statoscope/webpack-model/dist/normalizeCompilation"));
// eslint-disable-next-line @typescript-eslint/no-var-requires
const version = require('../package.json');
class ConsoleReporter {
    constructor(options) {
        this.options = options;
    }
    async run(result) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        var _k;
        console.log(`Preparing data for Statoscope report...`);
        const generator = new generator_1.default({
            adapter: version.adapter,
            author: version.author,
            description: version.description,
            homepage: version.homepage,
            name: version.name,
            version: version.version,
        });
        for (const rule of result.rules) {
            const storage = rule.api.getStorage();
            const ruleDescriptor = rule.api.getRuleDescriptor();
            if (ruleDescriptor) {
                generator.handleRule(rule.name, ruleDescriptor);
            }
            for (const entry of storage) {
                generator.handleEntry(rule.name, entry);
            }
        }
        const parsedInput = (await (0, json_ext_1.parseChunked)(fs_1.default.createReadStream(result.files.input)));
        const meta = {
            descriptor: { name: package_json_1.default.name, version: package_json_1.default.version },
            extensions: [],
        };
        (_a = parsedInput.__statoscope) !== null && _a !== void 0 ? _a : (parsedInput.__statoscope = meta);
        (_b = (_k = parsedInput.__statoscope).extensions) !== null && _b !== void 0 ? _b : (_k.extensions = []);
        parsedInput.__statoscope.extensions.push(generator.get());
        let parsedReference;
        if (result.files.reference) {
            parsedReference = await (0, json_ext_1.parseChunked)(fs_1.default.createReadStream(result.files.reference));
        }
        const id = path_1.default.basename(result.files.input, '.json');
        const reportPath = ((_c = this.options) === null || _c === void 0 ? void 0 : _c.saveReportTo) ||
            path_1.default.join(os_1.default.tmpdir(), `statoscope-report-${id}-${Date.now()}.html`);
        const statsPath = (_d = this.options) === null || _d === void 0 ? void 0 : _d.saveStatsTo;
        if (statsPath) {
            console.log(`Generating stats...`);
            const toDir = path_1.default.dirname(statsPath);
            if (!fs_1.default.existsSync(toDir)) {
                fs_1.default.mkdirSync(toDir, { recursive: true });
            }
            const statsFileStream = fs_1.default.createWriteStream(statsPath);
            const statStream = (0, json_ext_1.stringifyStream)(parsedInput, (0, utils_1.makeReplacer)(parsedInput.__statoscope.context, '.', ['context', 'source']));
            statStream.pipe(statsFileStream);
            await (0, utils_1.waitFinished)(statsFileStream);
            console.log(`Stats saved into ${statsPath}`);
        }
        if (((_e = this.options) === null || _e === void 0 ? void 0 : _e.saveReportTo) || ((_f = this.options) === null || _f === void 0 ? void 0 : _f.open)) {
            console.log(`Generating Statoscope report...`);
            (0, normalizeCompilation_1.default)(parsedInput);
            if (parsedReference) {
                (0, normalizeCompilation_1.default)(parsedReference);
            }
            const reportFilename = await (0, utils_1.transform)({
                writer: {
                    scripts: [{ type: 'path', path: require.resolve('@statoscope/webpack-ui') }],
                    init: `function (data) {
            Statoscope.default(data.map((item) => ({ name: item.id, data: item.data })));
          }`,
                    dataCompression: !this.options.disableReportCompression,
                },
            }, [
                {
                    type: 'data',
                    filename: 'input.json',
                    data: parsedInput,
                    replacer: (0, utils_1.makeReplacer)((_g = parsedInput.__statoscope) === null || _g === void 0 ? void 0 : _g.context, '.', [
                        'context',
                        'source',
                    ]),
                },
                parsedReference
                    ? {
                        type: 'data',
                        filename: 'reference.json',
                        data: parsedReference,
                        replacer: (0, utils_1.makeReplacer)((_h = parsedReference.__statoscope) === null || _h === void 0 ? void 0 : _h.context, '.', [
                            'context',
                            'source',
                        ]),
                    }
                    : null,
            ].filter(Boolean), reportPath);
            console.log(`Statoscope report saved into ${reportFilename}`);
            if ((_j = this.options) === null || _j === void 0 ? void 0 : _j.open) {
                (0, open_1.default)(reportFilename);
            }
        }
    }
}
exports.default = ConsoleReporter;
