// @gravity-ui/table/build/esm/components/TableSettingsColumn/TableSettingsColumn.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { SortableContext, useSortable, verticalListSortingStrategy } from '@dnd-kit/sortable';
import { Grip } from '@gravity-ui/icons';
import { Checkbox, Divider, Icon, Text } from '@gravity-ui/uikit';
import { b } from "./TableSettingsColumn.classname.js";
import { getColumnTitle, getIsVisible, isEnabledHiding } from "./TableSettingsColumn.utils.js";
import "./TableSettingsColumn.css";
export const TableSettingsColumn = ({ column, header, children, visibilityState, filterable, sortable, disabled = false, activeDepth, withOffset = false, onVisibilityToggle, }) => {
    const innerColumns = column.getLeafColumns();
    const isVisible = React.useMemo(() => innerColumns.some((innerColumn) => getIsVisible(innerColumn, visibilityState)), [innerColumns, visibilityState]);
    const isIndeterminate = React.useMemo(() => isVisible &&
        innerColumns.some((innerColumn) => !getIsVisible(innerColumn, visibilityState)), [isVisible, innerColumns, visibilityState]);
    const isParent = innerColumns.length > 1;
    const toggle = () => {
        if (isParent) {
            const newState = innerColumns.reduce((acc, innerColumn) => {
                const result = { ...acc };
                if (isEnabledHiding(innerColumn))
                    result[innerColumn.id] = !(isVisible && !isIndeterminate);
                return result;
            }, {});
            onVisibilityToggle((prevState) => ({ ...prevState, ...newState }));
        }
        else {
            onVisibilityToggle((prevState) => ({ ...prevState, [column.id]: !isVisible }));
        }
    };
    const renderSpacers = () => {
        const spacers = [];
        for (let index = 0; index < column.depth; index++) {
            spacers.push(_jsx("div", { className: b('vertical-spacer') }, index));
        }
        return spacers;
    };
    const isDisabledContext = typeof activeDepth === 'number' ? activeDepth !== column.depth : false;
    const isRoot = column.depth === 0;
    const { attributes, listeners, setNodeRef, transform, transition, isDragging } = useSortable({
        id: column.id,
        disabled: {
            draggable: isDisabledContext,
            droppable: isDisabledContext,
        },
    });
    const style = {
        transform: `translate3d(0px, ${transform?.y ?? 0}px, 0)`,
        transition,
    };
    return (_jsxs("div", { style: style, ref: setNodeRef, className: b({ dragging: isDragging, root: isRoot, sortable }), children: [_jsx("div", { className: b('background'), children: _jsxs("div", { ...attributes, ...(sortable ? { ...listeners } : {}), className: b('layout', { dragging: isDragging }), children: [_jsxs("div", { className: b('content', { disabled }), children: [filterable && (
                                // eslint-disable-next-line jsx-a11y/no-static-element-interactions
                                _jsx("div", { onMouseDown: (e) => {
                                        e.stopPropagation();
                                    }, onTouchStart: (e) => {
                                        e.stopPropagation();
                                    }, children: _jsx(Checkbox, { checked: isVisible, disabled: !isEnabledHiding(column), onChange: toggle, indeterminate: isIndeterminate, className: b('checkbox', { 'unset-coursor': isDragging }) }) })), !filterable && withOffset && _jsx("div", { className: b('checkbox-spacer') }), renderSpacers(), _jsx(Text, { variant: "body-1", className: b('name', { parent: isParent }), children: getColumnTitle(column, header) }), sortable ? (_jsx("span", { className: b('drag-handle', { 'unset-coursor': isDragging }), ...listeners, children: _jsx(Icon, { data: Grip, size: 16 }) })) : null] }), _jsx(SortableContext, { id: column.id, items: column.columns?.map(({ id }) => id), strategy: verticalListSortingStrategy, children: children })] }) }), isRoot && !isDragging ? (_jsx("div", { className: b('divider'), children: _jsx(Divider, {}) })) : null] }));
};
