// @gravity-ui/navigation/build/cjs/index.js
'use strict';

var index = require('./index-CLbIRuop.js');
require('react');
require('@bem-react/classname');
require('@gravity-ui/uikit');
require('@gravity-ui/icons');
require('@gravity-ui/uikit/i18n');
require('react/jsx-runtime');
require('react-dom');
require('@gravity-ui/uikit/legacy');



exports.ActionBar = index.PublicActionBar;
exports.AsideFallback = index.AsideFallback;
exports.AsideHeader = index.AsideHeader;
exports.AsideHeaderContextProvider = index.AsideHeaderContextProvider;
exports.Drawer = index.Drawer;
exports.DrawerItem = index.DrawerItem;
exports.Footer = index.Footer;
exports.FooterItem = index.FooterItem;
exports.HotkeysPanel = index.HotkeysPanel;
exports.Logo = index.Logo;
exports.MOBILE_HEADER_EVENT_NAMES = index.MOBILE_HEADER_EVENT_NAMES;
exports.MobileFooter = index.MobileFooter;
exports.MobileHeader = index.MobileHeader;
exports.MobileHeaderFooterItem = index.FooterItem$1;
exports.MobileLogo = index.MobileLogo;
exports.PageLayout = index.PageLayout;
exports.PageLayoutAside = index.PageLayoutAside;
exports.Settings = index.Settings;
exports.Title = index.Title;
exports.getMobileHeaderCustomEvent = index.getMobileHeaderCustomEvent;
exports.useAsideHeaderContext = index.useAsideHeaderContext;
exports.useSettingsContext = index.useSettingsContext;
exports.useSettingsSelectionContext = index.useSettingsSelectionContext;
