// @gravity-ui/table/build/esm/hooks/useSortableList.js
import * as React from 'react';
import { useDndMonitor } from '@dnd-kit/core';
export const useSortableList = ({ items, onDragStart, onDragEnd, enableNesting, childModeOffset = 20, nextChildModeOffset = 10, }) => {
    const [activeItemKey, setActiveItemKey] = React.useState(null);
    const [targetItemIndex, setTargetItemIndex] = React.useState(-1);
    const [isParentMode, setIsParentMode] = React.useState(false);
    const [isChildMode, setIsChildMode] = React.useState(false);
    const [isNextChildMode, setIsNextChildMode] = React.useState(false);
    const itemIndexMap = React.useMemo(() => {
        const map = new Map();
        items.forEach((item, index) => {
            map.set(item, index);
        });
        return map;
    }, [items]);
    const getItemIndex = (item) => {
        if (typeof item !== 'undefined' && item !== null) {
            return itemIndexMap.get(item) ?? -1;
        }
        return -1;
    };
    const resetState = () => {
        setActiveItemKey(null);
        setTargetItemIndex(-1);
        setIsParentMode(false);
        setIsChildMode(false);
        setIsNextChildMode(false);
    };
    const getTargetItemKey = (activeId, overId) => {
        if (overId) {
            const overItemIndex = getItemIndex(overId);
            const activeItemIndex = getItemIndex(activeId);
            if (overItemIndex <= activeItemIndex) {
                return items[overItemIndex - 1];
            }
        }
        return overId;
    };
    useDndMonitor({
        onDragStart: (event) => {
            setActiveItemKey(event.active.id);
            onDragStart?.(event.active.id);
            document.body.style.setProperty('cursor', 'grabbing');
        },
        onDragMove: (event) => {
            if (enableNesting) {
                setIsParentMode(event.delta.x < -childModeOffset);
                setIsChildMode(event.delta.x > childModeOffset);
                setIsNextChildMode(event.delta.x > nextChildModeOffset && event.delta.x <= childModeOffset);
            }
        },
        onDragOver: (event) => {
            const targetItemKey = getTargetItemKey(event.active.id, event.over?.id);
            setTargetItemIndex(targetItemKey && targetItemKey !== event.active.id
                ? getItemIndex(targetItemKey)
                : -1);
        },
        onDragEnd: (event) => {
            document.body.style.setProperty('cursor', 'default');
            if (!event.over) {
                resetState();
                return;
            }
            const draggedItemKey = event.active.id;
            const targetItemKey = getTargetItemKey(draggedItemKey, event.over.id);
            if (targetItemKey === draggedItemKey) {
                resetState();
                return;
            }
            if (isChildMode) {
                onDragEnd?.({
                    draggedItemKey,
                    targetItemKey,
                    enableNesting,
                });
            }
            else {
                onDragEnd?.({
                    draggedItemKey,
                    baseItemKey: targetItemKey,
                    baseNextItemKey: items[targetItemKey ? getItemIndex(targetItemKey) + 1 : 0],
                    enableNesting,
                    nextChild: isNextChildMode,
                    pullFromParent: isParentMode,
                });
            }
            resetState();
        },
        onDragCancel: resetState,
    });
    return {
        activeItemKey,
        activeItemIndex: getItemIndex(activeItemKey),
        targetItemIndex,
        isParentMode,
        isChildMode,
        isNextChildMode,
    };
};
