// @gravity-ui/table/build/esm/hooks/useTable.js
import { getCoreRowModel, getExpandedRowModel, getGroupedRowModel, getSortedRowModel, useReactTable, } from '@tanstack/react-table';
export const useTable = (options) => {
    const tableOptions = {
        ...options,
        enableColumnPinning: options.enableColumnPinning ?? false,
        enableColumnResizing: options.enableColumnResizing ?? false,
        enableExpanding: options.enableExpanding ?? false,
        enableGrouping: options.enableGrouping ?? false,
        enableMultiRowSelection: options.enableMultiRowSelection ?? false,
        enableRowSelection: options.enableRowSelection ?? false,
        enableSorting: options.enableSorting ?? false,
        getCoreRowModel: options.getCoreRowModel ?? getCoreRowModel(),
        getExpandedRowModel: options.enableExpanding
            ? (options.getExpandedRowModel ?? getExpandedRowModel())
            : undefined,
        getGroupedRowModel: options.enableGrouping
            ? (options.getGroupedRowModel ?? getGroupedRowModel())
            : undefined,
        getSortedRowModel: options.enableSorting
            ? (options.getSortedRowModel ?? getSortedRowModel())
            : undefined,
        manualGrouping: options.manualGrouping ?? false,
        manualSorting: options.manualSorting ?? false,
    };
    return useReactTable(tableOptions);
};
