// @gravity-ui/table/build/esm/hooks/useTableSettings.js
import * as React from 'react';
export const useTableSettings = ({ initialOrdering = [], initialVisibility = {}, onVisibilityChange, onOrderingChange } = {
    initialOrdering: [],
    initialVisibility: {},
}) => {
    const [visibilityState, setVisibilityState] = React.useState(initialVisibility);
    const [orderingState, setOrderState] = React.useState(initialOrdering);
    const onColumnVisibilityChange = React.useCallback((value) => {
        setVisibilityState(value);
        if (onVisibilityChange) {
            const newValue = typeof value === 'function' ? value(visibilityState) : value;
            onVisibilityChange(newValue);
        }
    }, [onVisibilityChange, visibilityState]);
    const onColumnOrderChange = React.useCallback((value) => {
        setOrderState(value);
        if (onOrderingChange) {
            const newValue = typeof value === 'function' ? value(orderingState) : value;
            onOrderingChange(newValue);
        }
    }, [onOrderingChange, orderingState]);
    return {
        state: {
            columnVisibility: visibilityState,
            columnOrder: orderingState,
        },
        callbacks: {
            onColumnVisibilityChange,
            onColumnOrderChange,
        },
    };
};
