// @gravity-ui/date-utils/build/dateTime/parse.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTimestampFromArray = getTimestampFromArray;
exports.getTimestampFromObject = getTimestampFromObject;
var timeZone_1 = require("../timeZone");
var utils_1 = require("../utils");
var locale_1 = require("../utils/locale");
function getTimestampFromArray(input, timezone, offset) {
    if (input.length === 0) {
        return getTimestampFromObject({}, timezone, offset);
    }
    var dateParts = input.map(Number);
    var year = dateParts[0], _a = dateParts[1], month = _a === void 0 ? 0 : _a, _b = dateParts[2], date = _b === void 0 ? 1 : _b, _c = dateParts[3], hour = _c === void 0 ? 0 : _c, _d = dateParts[4], minute = _d === void 0 ? 0 : _d, _e = dateParts[5], second = _e === void 0 ? 0 : _e, _f = dateParts[6], millisecond = _f === void 0 ? 0 : _f;
    return getTimestampFromObject({ year: year, month: month, date: date, hour: hour, minute: minute, second: second, millisecond: millisecond }, timezone, offset);
}
var defaultUnitValues = {
    year: 1,
    month: 0,
    date: 1,
    hour: 0,
    minute: 0,
    second: 0,
    millisecond: 0,
};
var defaultWeekUnitValues = {
    weekYear: 1,
    weekNumber: 1,
    weekday: 1,
    hour: 0,
    minute: 0,
    second: 0,
    millisecond: 0,
};
var defaultOrdinalUnitValues = {
    year: 1,
    dayOfYear: 1,
    hour: 0,
    minute: 0,
    second: 0,
    millisecond: 0,
};
var orderedUnits = ['year', 'month', 'date', 'hour', 'minute', 'second', 'millisecond'];
var orderedWeekUnits = [
    'weekYear',
    'weekNumber',
    'weekday',
    'hour',
    'minute',
    'second',
    'millisecond',
];
var orderedOrdinalUnits = [
    'year',
    'dayOfYear',
    'hour',
    'minute',
    'second',
    'millisecond',
];
function getTimestampFromObject(input, timezone, offset) {
    var _a;
    var normalized = (0, utils_1.normalizeDateComponents)(input, utils_1.normalizeComponent);
    normalized.date = (_a = normalized.day) !== null && _a !== void 0 ? _a : normalized.date;
    var definiteWeekStuff = normalized.weekNumber !== undefined || normalized.weekYear !== undefined;
    var containsDayOfYear = normalized.dayOfYear !== undefined;
    var containsYear = normalized.year !== undefined;
    var containsMonthOrDate = normalized.month !== undefined || normalized.date !== undefined;
    var containsYearOrMonthDay = containsYear || containsMonthOrDate;
    if ((containsYearOrMonthDay || containsDayOfYear) && definiteWeekStuff) {
        throw new Error("Can't mix weekYear/weekNumber units with year/month/day or ordinals");
    }
    if (containsMonthOrDate && containsDayOfYear) {
        throw new Error("Can't mix ordinal dates with month/day");
    }
    var useWeekData = definiteWeekStuff || (normalized.weekday && !containsYearOrMonthDay);
    var isFixedOffset = offset !== undefined;
    var likelyOffset = isFixedOffset ? offset : (0, timeZone_1.timeZoneOffset)(timezone, Date.now());
    var objNow = (0, utils_1.tsToObject)(Date.now(), likelyOffset);
    if (useWeekData) {
        var localeData = (0, locale_1.getLocaleData)('en'); // TODO: locale
        var _b = (0, locale_1.getLocaleWeekValues)(localeData), minDaysInFirstWeek = _b.minDaysInFirstWeek, startOfWeek = _b.startOfWeek;
        objNow = __assign(__assign({}, objNow), (0, utils_1.gregorianToWeek)(objNow, minDaysInFirstWeek, startOfWeek));
        setDefaultValues(normalized, objNow, orderedWeekUnits, defaultWeekUnitValues);
        if (!isValidWeekData(normalized, minDaysInFirstWeek, startOfWeek)) {
            return [NaN, NaN];
        }
        normalized = __assign(__assign({}, normalized), (0, utils_1.weekToGregorian)(normalized, minDaysInFirstWeek, startOfWeek));
    }
    else if (containsDayOfYear) {
        objNow = __assign(__assign({}, objNow), (0, utils_1.gregorianToOrdinal)(objNow));
        setDefaultValues(normalized, objNow, orderedOrdinalUnits, defaultOrdinalUnitValues);
        if (!isValidOrdinalData(normalized)) {
            return [NaN, NaN];
        }
        normalized = __assign(__assign({}, normalized), (0, utils_1.uncomputeOrdinal)(__assign(__assign({}, normalized), { ordinal: normalized.dayOfYear })));
    }
    else {
        setDefaultValues(normalized, objNow, orderedUnits, defaultUnitValues);
    }
    if (!isValidDateData(normalized) || !isValidTimeData(normalized)) {
        return [NaN, NaN];
    }
    var ts = (0, utils_1.objToTS)(normalized);
    if (isFixedOffset) {
        return [ts - offset * 60 * 1000, offset];
    }
    return (0, timeZone_1.fixOffset)(ts, likelyOffset, timezone);
}
function setDefaultValues(value, now, units, defaultValues) {
    var foundFirst = false;
    for (var _i = 0, units_1 = units; _i < units_1.length; _i++) {
        var unit = units_1[_i];
        if (value[unit] !== undefined) {
            foundFirst = true;
        }
        else if (foundFirst) {
            value[unit] = defaultValues[unit];
        }
        else {
            value[unit] = now[unit];
        }
    }
}
function isValidWeekData(weekData, minDaysInFirstWeek, startOfWeek) {
    if (minDaysInFirstWeek === void 0) { minDaysInFirstWeek = 4; }
    if (startOfWeek === void 0) { startOfWeek = 1; }
    return (weekData.weekYear !== undefined &&
        weekData.weekNumber !== undefined &&
        weekData.weekday !== undefined &&
        Number.isInteger(weekData.weekYear) &&
        Number.isInteger(weekData.weekNumber) &&
        weekData.weekNumber >= 1 &&
        weekData.weekNumber <=
            (0, utils_1.weeksInWeekYear)(weekData.weekYear, minDaysInFirstWeek, startOfWeek) &&
        Number.isInteger(weekData.weekday) &&
        weekData.weekday >= 1 &&
        weekData.weekday <= 7);
}
function isValidOrdinalData(ordinalData) {
    return (ordinalData.year !== undefined &&
        ordinalData.dayOfYear !== undefined &&
        Number.isInteger(ordinalData.year) &&
        Number.isInteger(ordinalData.dayOfYear) &&
        ordinalData.dayOfYear >= 1 &&
        ordinalData.dayOfYear <= (0, utils_1.daysInYear)(ordinalData.year));
}
function isValidDateData(dateData) {
    return (dateData.year !== undefined &&
        dateData.month !== undefined &&
        dateData.date !== undefined &&
        Number.isInteger(dateData.year) &&
        Number.isInteger(dateData.month) &&
        dateData.month >= 0 &&
        dateData.month <= 11 &&
        Number.isInteger(dateData.date) &&
        dateData.date >= 1 &&
        dateData.date <= (0, utils_1.daysInMonth)(dateData.year, dateData.month));
}
function isValidTimeData(timeData) {
    return (timeData.hour !== undefined &&
        timeData.minute !== undefined &&
        timeData.second !== undefined &&
        timeData.millisecond !== undefined &&
        Number.isInteger(timeData.hour) &&
        Number.isInteger(timeData.minute) &&
        Number.isInteger(timeData.second) &&
        Number.isInteger(timeData.millisecond) &&
        ((timeData.hour >= 0 && timeData.hour <= 23) ||
            (timeData.hour === 24 &&
                timeData.minute === 0 &&
                timeData.second === 0 &&
                timeData.millisecond === 0)) &&
        timeData.minute >= 0 &&
        timeData.minute <= 59 &&
        timeData.second >= 0 &&
        timeData.second <= 59 &&
        timeData.millisecond >= 0 &&
        timeData.millisecond <= 999);
}
