// @gravity-ui/date-utils/build/dateTime/regexParse.js
"use strict";
// Copyright 2019 JS Foundation and other contributors
// Copyright 2024 YANDEX LLC
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseISODate = parseISODate;
exports.parseRFC2822Date = parseRFC2822Date;
exports.parseHTTPDate = parseHTTPDate;
exports.parseISOTimeOnly = parseISOTimeOnly;
exports.parseDateString = parseDateString;
var constants_1 = require("../constants");
var English = __importStar(require("../locale/english"));
function combineExtractors() {
    var extractors = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        extractors[_i] = arguments[_i];
    }
    return function (m) {
        var res = extractors.reduce(function (_a, extractor) {
            var dateObj = _a[0], timezoneOffset = _a[1], cursor = _a[2];
            var _b = extractor(m, cursor), nextDateObj = _b[0], nextTimezoneOffset = _b[1], nextCursor = _b[2];
            return [
                __assign(__assign({}, dateObj), nextDateObj),
                nextTimezoneOffset !== null && nextTimezoneOffset !== void 0 ? nextTimezoneOffset : timezoneOffset,
                nextCursor,
            ];
        }, [{}, null, 1]);
        return [res[0], res[1]];
    };
}
function parse(input) {
    var patterns = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        patterns[_i - 1] = arguments[_i];
    }
    if (!input) {
        return [null, null];
    }
    for (var _a = 0, patterns_1 = patterns; _a < patterns_1.length; _a++) {
        var _b = patterns_1[_a], regex = _b[0], extractor = _b[1];
        var match = regex.exec(input);
        if (match) {
            return extractor(match);
        }
    }
    return [null, null];
}
// IANA time zone format: https://en.wikipedia.org/wiki/List_of_IANA_time_zones
var ianaRegex = /[A-Za-z_+-]{1,256}(?::?\/[A-Za-z0-9_+-]{1,256}(?:\/[A-Za-z0-9_+-]{1,256})?)?/;
// Z or ±00:00
var offsetRegex = /(?:(Z)|([+-]\d\d)(?::?(\d\d))?)/;
var isoExtendedZone = "(?:".concat(offsetRegex.source, "?(?:\\[(").concat(ianaRegex.source, ")\\])?)?");
// hh:mm:ss.sss
var isoTimeBaseRegex = /(\d\d)(?::?(\d\d)(?::?(\d\d)(?:[.,](\d{1,30}))?)?)?/;
var isoTimeRegex = RegExp("".concat(isoTimeBaseRegex.source).concat(isoExtendedZone));
var isoTimeOnly = RegExp("^T?".concat(isoTimeBaseRegex.source, "$"));
// Thh:mm:ss.sss±00:00
var isoTimeExtensionRegex = RegExp("(?:T".concat(isoTimeRegex.source, ")?"));
// YYYY-MM-DD
var isoYmdRegex = /([+-]\d{6}|\d{4})(?:-?(\d\d)(?:-?(\d\d))?)?/;
// YYYY-Www-D
var isoWeekRegex = /(\d{4})-?W(\d\d)(?:-?(\d))?/;
// YYYY-DDD
var isoOrdinalRegex = /(\d{4})-?(\d{3})/;
var isoYmdWithTimeExtensionRegex = new RegExp("^".concat(isoYmdRegex.source).concat(isoTimeExtensionRegex.source, "$"));
var isoWeekWithTimeExtensionRegex = new RegExp("^".concat(isoWeekRegex.source).concat(isoTimeExtensionRegex.source, "$"));
var isoOrdinalWithTimeExtensionRegex = new RegExp("^".concat(isoOrdinalRegex.source).concat(isoTimeExtensionRegex.source, "$"));
var isoTimeFullRegex = new RegExp("^".concat(isoTimeRegex.source, "$"));
// https://datatracker.ietf.org/doc/html/rfc2822#section-4.3
var obsOffsets = {
    GMT: 0,
    UT: 0,
    EDT: -4 * 60,
    EST: -5 * 60,
    CDT: -5 * 60,
    CST: -6 * 60,
    MDT: -6 * 60,
    MST: -7 * 60,
    PDT: -7 * 60,
    PST: -8 * 60,
};
// RFC 2822/5322 https://datatracker.ietf.org/doc/html/rfc2822
// Fri, 19 Nov 82 16:14:55 GMT
var rfc2822 = /^(?:(Mon|Tue|Wed|Thu|Fri|Sat|Sun),\s)?(\d{1,2})\s(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s(\d{2,4})\s(\d\d):(\d\d)(?::(\d\d))?\s(?:(UT|GMT|[ECMP][SD]T)|([Zz])|(?:([+-]\d\d)(\d\d)))$/;
function preprocessRFC2822(s) {
    // Remove comments and folding whitespace and replace multiple-spaces with a single space
    return s
        .replace(/\([^()]*\)|[\n\t]/g, ' ')
        .replace(/(\s\s+)/g, ' ')
        .trim();
}
function extractRfc2822(match) {
    var weekday = match[1], day = match[2], month = match[3], year = match[4], hour = match[5], minute = match[6], second = match[7], obsOffset = match[8], zeroOffset = match[9], offsetHours = match[10], offsetMinutes = match[11];
    var result = stringsToDateObject(weekday, year, month, day, hour, minute, second);
    var offset;
    if (obsOffset) {
        offset = obsOffsets[obsOffset];
    }
    else if (zeroOffset) {
        offset = 0;
    }
    else {
        var hours = parseInt(offsetHours, 10);
        var sign = hours < 0 || Object.is(hours, -0) ? -1 : 1;
        var minutes = parseInt(offsetMinutes, 10) || 0;
        offset = (hours || 0) * 60 + sign * minutes;
    }
    return [result, offset];
}
function extractISOYmd(match, cursor) {
    var item = {
        year: parseInteger(match[cursor]),
        month: parseInteger(match[cursor + 1], 1) - 1,
        day: parseInteger(match[cursor + 2], 1),
    };
    return [item, null, cursor + 3];
}
function extractISOTime(match, cursor) {
    var item = {
        hour: parseInteger(match[cursor], 0),
        minute: parseInteger(match[cursor + 1], 0),
        second: parseInteger(match[cursor + 2], 0),
        millisecond: parseMilliseconds(match[cursor + 3]),
    };
    return [item, null, cursor + 4];
}
function extractISOOffset(match, cursor) {
    var local = !match[cursor] && !match[cursor + 1];
    if (local) {
        return [{}, null, cursor + 3];
    }
    if (match[cursor]) {
        return [{}, 0, cursor + 3];
    }
    var fullOffset = signedOffset(match[cursor + 1], match[cursor + 2]);
    return [{}, fullOffset, cursor + 3];
}
function extractIANAZone(match, cursor) {
    var zone = match[cursor] || null;
    return [{}, zone, cursor + 1];
}
function extractISOWeekData(match, cursor) {
    var item = {
        weekYear: parseInteger(match[cursor]),
        weekNumber: parseInteger(match[cursor + 1], 1),
        weekday: parseInteger(match[cursor + 2], 1),
    };
    return [item, null, cursor + 3];
}
function extractISOOrdinalData(match, cursor) {
    var item = {
        year: parseInteger(match[cursor]),
        dayOfYear: parseInteger(match[cursor + 1], 1),
    };
    return [item, null, cursor + 2];
}
var extractISOYmdTimeAndOffset = combineExtractors(extractISOYmd, extractISOTime, extractISOOffset, extractIANAZone);
var extractISOWeekTimeAndOffset = combineExtractors(extractISOWeekData, extractISOTime, extractISOOffset, extractIANAZone);
var extractISOOrdinalDateAndTime = combineExtractors(extractISOOrdinalData, extractISOTime, extractISOOffset, extractIANAZone);
var extractISOTimeAndOffset = combineExtractors(extractISOTime, extractISOOffset, extractIANAZone);
// https://datatracker.ietf.org/doc/html/rfc1123#page-55
var rfc1123 = /^(Mon|Tue|Wed|Thu|Fri|Sat|Sun), (\d\d) (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) (\d{4}) (\d\d):(\d\d):(\d\d) GMT$/;
// https://datatracker.ietf.org/doc/html/rfc850#section-2.1.4
var rfc850 = /^(Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday), (\d\d)-(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)-(\d\d) (\d\d):(\d\d):(\d\d) GMT$/;
// Fri Nov 19 16:59:30 1982
var ascii = /^(Mon|Tue|Wed|Thu|Fri|Sat|Sun) (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) ( \d|\d\d) (\d\d):(\d\d):(\d\d) (\d{4})$/;
function extractRFC1123Or850(match) {
    var weekdayStr = match[1], dayStr = match[2], monthStr = match[3], yearStr = match[4], hourStr = match[5], minuteStr = match[6], secondStr = match[7], result = stringsToDateObject(weekdayStr, yearStr, monthStr, dayStr, hourStr, minuteStr, secondStr);
    return [result, constants_1.UtcTimeZone];
}
function extractASCII(match) {
    var weekdayStr = match[1], monthStr = match[2], dayStr = match[3], hourStr = match[4], minuteStr = match[5], secondStr = match[6], yearStr = match[7], result = stringsToDateObject(weekdayStr, yearStr, monthStr, dayStr, hourStr, minuteStr, secondStr);
    return [result, constants_1.UtcTimeZone];
}
function parseInteger(str, defaultValue) {
    return str ? parseInt(str, 10) : defaultValue;
}
function parseMilliseconds(str) {
    return str ? Math.floor(parseFloat("0.".concat(str)) * 1000) : undefined;
}
function signedOffset(offsetHours, offsetMinutes) {
    var hours = parseInt(offsetHours, 10);
    var sign = hours < 0 || Object.is(hours, -0) ? -1 : 1;
    var minutes = parseInt(offsetMinutes, 10) || 0;
    var offset = (hours || 0) * 60 + sign * minutes;
    return offset;
}
function stringsToDateObject(weekdayStr, yearStr, monthStr, dayStr, hourStr, minuteStr, secondStr) {
    var res = {
        year: yearStr.length === 2
            ? fullYearFromTwoDigitYear(parseInteger(yearStr))
            : parseInteger(yearStr),
        month: monthStr.length > 3
            ? English.monthsLong.indexOf(monthStr)
            : English.monthsShort.indexOf(monthStr),
        date: parseInteger(dayStr),
        hour: parseInteger(hourStr),
        minute: parseInteger(minuteStr),
        second: parseInteger(secondStr),
    };
    if (weekdayStr) {
        res.weekday =
            (weekdayStr.length > 3
                ? English.weekdaysLong.indexOf(weekdayStr)
                : English.weekdaysShort.indexOf(weekdayStr)) + 1;
    }
    return res;
}
function fullYearFromTwoDigitYear(year) {
    if (!year || year > 99) {
        return year;
    }
    // TODO: add two digit cutoff year to settings
    return year > 49 ? 1900 + year : 2000 + year;
}
function parseISODate(s) {
    return parse(s, [isoYmdWithTimeExtensionRegex, extractISOYmdTimeAndOffset], [isoWeekWithTimeExtensionRegex, extractISOWeekTimeAndOffset], [isoOrdinalWithTimeExtensionRegex, extractISOOrdinalDateAndTime], [isoTimeFullRegex, extractISOTimeAndOffset]);
}
function parseRFC2822Date(s) {
    return parse(preprocessRFC2822(s), [rfc2822, extractRfc2822]);
}
function parseHTTPDate(s) {
    return parse(s, [rfc1123, extractRFC1123Or850], [rfc850, extractRFC1123Or850], [ascii, extractASCII]);
}
function parseISOTimeOnly(s) {
    return parse(s, [isoTimeOnly, combineExtractors(extractISOTime)]);
}
function parseDateString(input) {
    var _a, _b, _c;
    var _d = parseISODate(input), obj = _d[0], offset = _d[1];
    if (obj !== null) {
        return [obj, offset];
    }
    _a = parseRFC2822Date(input), obj = _a[0], offset = _a[1];
    if (obj !== null) {
        return [obj, offset];
    }
    _b = parseHTTPDate(input), obj = _b[0], offset = _b[1];
    if (obj !== null) {
        return [obj, offset];
    }
    _c = parseISOTimeOnly(input), obj = _c[0], offset = _c[1];
    if (obj !== null) {
        return [obj, offset];
    }
    return [{}, null];
}
