// @gravity-ui/date-components/dist/cjs/components/RangeDateSelection/components/SelectionControl/SelectionControl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectionControl = SelectionControl;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = tslib_1.__importDefault(require("react"));
const useMove_1 = require("../../../../hooks/useMove.js");
const cn_1 = require("../../../../utils/cn.js");
const i18n_1 = require("../../i18n/index.js");
const span_1 = require("../../utils/span.js");
const Ruler_1 = require("../Ruler/Ruler.js");
require("./SelectionControl.css");
const b = (0, cn_1.block)('range-date-selection-control');
const MIN_SELECTION_WIDTH = 12; // px;
const MIN_DRAG_SCALE = 0.01;
const MAX_DRAG_SCALE = 100;
const DRAG_SCALE_STRENGTH = 20;
function SelectionControl(_a) {
    var _b;
    var { state, style, className } = _a, props = tslib_1.__rest(_a, ["state", "style", "className"]);
    const { start: startDate, end: endDate } = (0, Ruler_1.useViewportInterval)();
    const viewport = (0, Ruler_1.useViewportDimensions)();
    const [isDragging, setIsDragging] = react_1.default.useState(null);
    const moveDataRef = react_1.default.useRef({
        isZooming: false,
    });
    const visualMinDuration = (MIN_SELECTION_WIDTH * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
    const handleMoveEnd = () => {
        state.endDragging();
        setIsDragging(null);
    };
    const { moveProps } = (0, useMove_1.useMove)({
        onMoveStart: (e) => {
            setIsDragging('dragger');
            state.startDragging();
            moveDataRef.current = {
                isZooming: state.canResize && (e.ctrlKey || e.metaKey),
            };
        },
        onMoveEnd: handleMoveEnd,
        onMove: (e) => {
            if (e.deltaX === 0) {
                return;
            }
            if (moveDataRef.current.isZooming) {
                let scale = Math.exp((e.deltaX * DRAG_SCALE_STRENGTH) / viewport.width);
                scale = Math.min(Math.max(scale, MIN_DRAG_SCALE), MAX_DRAG_SCALE);
                state.scale(scale, { visualMinDuration });
            }
            else {
                let delta = (e.deltaX * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
                if (e.pointerType === 'keyboard') {
                    delta =
                        delta > 0
                            ? Math.max(delta, state.align)
                            : Math.min(delta, -1 * state.align);
                }
                state.move(delta, { visualMinDuration });
            }
        },
    });
    const { moveProps: moveStartProps } = (0, useMove_1.useMove)({
        onMoveStart: () => {
            setIsDragging('start');
            state.startDragging();
        },
        onMoveEnd: handleMoveEnd,
        onMove: (e) => {
            if (e.deltaX === 0) {
                return;
            }
            let delta = (e.deltaX * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
            if (e.pointerType === 'keyboard') {
                delta =
                    delta > 0 ? Math.max(delta, state.align) : Math.min(delta, -1 * state.align);
            }
            state.moveStart(delta, { visualMinDuration });
        },
    });
    const { moveProps: moveEndProps } = (0, useMove_1.useMove)({
        onMoveStart: () => {
            setIsDragging('end');
            state.startDragging();
        },
        onMoveEnd: handleMoveEnd,
        onMove: (e) => {
            if (e.deltaX === 0) {
                return;
            }
            let delta = (e.deltaX * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
            if (e.pointerType === 'keyboard') {
                delta =
                    delta > 0 ? Math.max(delta, state.align) : Math.min(delta, -1 * state.align);
            }
            state.moveEnd(delta, { visualMinDuration });
        },
    });
    const { start: startValue, end: endValue } = state.value;
    const startX = (0, span_1.lerp)({ start: startDate.valueOf(), end: endDate.valueOf() }, { start: 0, end: viewport.width }, startValue.valueOf());
    const endX = (0, span_1.lerp)({ start: startDate.valueOf(), end: endDate.valueOf() }, { start: 0, end: viewport.width }, endValue.valueOf());
    const draggerInputRef = react_1.default.useRef(null);
    const startInputRef = react_1.default.useRef(null);
    const endInputRef = react_1.default.useRef(null);
    let id = react_1.default.useId();
    id = (_b = props.id) !== null && _b !== void 0 ? _b : id;
    const { t } = i18n_1.i18n.useTranslation();
    return ((0, jsx_runtime_1.jsxs)("div", { className: b(null, className), style: Object.assign(Object.assign({}, style), { left: startX, width: endX - startX }), children: [(0, jsx_runtime_1.jsx)("div", Object.assign({ className: b('dragger', {
                    dragging: isDragging === 'dragger',
                    disabled: Boolean(isDragging && isDragging !== 'dragger'),
                }) }, moveProps, { onClick: () => {
                    var _a;
                    (_a = draggerInputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
                }, children: (0, jsx_runtime_1.jsx)("div", { className: b('slider-input-wrapper'), children: (0, jsx_runtime_1.jsx)("input", { id: `${id}-range`, ref: draggerInputRef, className: b('slider-input'), type: "range", step: 1, "aria-label": t('Range'), "aria-labelledby": [`${id}-range`, props['aria-labelledby']]
                            .filter(Boolean)
                            .join(' '), "aria-orientation": "horizontal", "aria-valuetext": `${startValue.format()} - ${endValue.format()}`, value: 0, onChange: () => { } }) }) })), state.canResize ? (
            /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
            (0, jsx_runtime_1.jsx)("div", Object.assign({ className: b('start', {
                    dragging: isDragging === 'start',
                    disabled: Boolean(isDragging && isDragging !== 'start'),
                }) }, moveStartProps, { onClick: () => {
                    var _a;
                    (_a = startInputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
                }, children: (0, jsx_runtime_1.jsx)("div", { className: b('slider-input-wrapper'), children: (0, jsx_runtime_1.jsx)("input", { id: `${id}-start`, ref: startInputRef, className: b('slider-input'), type: "range", step: 1, "aria-label": t('Start of range'), "aria-labelledby": [`${id}-start`, props['aria-labelledby']]
                            .filter(Boolean)
                            .join(' '), "aria-orientation": "horizontal", "aria-valuetext": `${startValue.format()}`, value: 0, onChange: () => { } }) }) }))) : null, state.canResize ? (
            /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
            (0, jsx_runtime_1.jsx)("div", Object.assign({ className: b('end', {
                    dragging: isDragging === 'end',
                    disabled: Boolean(isDragging && isDragging !== 'end'),
                }) }, moveEndProps, { onClick: () => {
                    var _a;
                    (_a = endInputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
                }, children: (0, jsx_runtime_1.jsx)("div", { className: b('slider-input-wrapper'), children: (0, jsx_runtime_1.jsx)("input", { id: `${id}-end`, ref: endInputRef, className: b('slider-input'), type: "range", step: 1, "aria-label": t('End of range'), "aria-labelledby": [`${id}-end`, props['aria-labelledby']]
                            .filter(Boolean)
                            .join(' '), "aria-orientation": "horizontal", "aria-valuetext": `${endValue.format()}`, value: 0, onChange: () => { } }) }) }))) : null] }));
}
