// @gravity-ui/date-components/dist/esm/components/RelativeDatePicker/RelativeDatePicker.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { Calendar as CalendarIcon, Clock as ClockIcon, Function as FunctionIcon, } from '@gravity-ui/icons';
import { Button, Icon, Popup, TextInput, useForkRef, useMobile } from '@gravity-ui/uikit';
import { block } from "../../utils/cn.js";
import { Calendar } from "../Calendar/index.js";
import { DateField } from "../DateField/index.js";
import { MobileCalendar } from "../DatePicker/MobileCalendar.js";
import { StubButton } from "../DatePicker/StubButton.js";
import { HiddenInput } from "../HiddenInput/HiddenInput.js";
import { useRelativeDatePickerProps } from "./hooks/useRelativeDatePickerProps.js";
import { useRelativeDatePickerState } from "./hooks/useRelativeDatePickerState.js";
import "./RelativeDatePicker.css";
const b = block('relative-date-picker');
export function RelativeDatePicker(props) {
    var _a;
    const state = useRelativeDatePickerState(props);
    const { groupProps, fieldProps, modeSwitcherProps, calendarButtonProps, popupProps, calendarProps, timeInputProps, } = useRelativeDatePickerProps(state, props);
    const [anchor, setAnchor] = React.useState(null);
    const handleRef = useForkRef(groupProps.ref, setAnchor);
    const isMobile = useMobile();
    const isOnlyTime = state.datePickerState.formatInfo.hasTime && !state.datePickerState.formatInfo.hasDate;
    return (_jsxs("div", Object.assign({}, groupProps, { ref: handleRef, className: b(null, props.className), children: [isMobile && state.mode === 'absolute' && (_jsx(MobileCalendar, { state: state.datePickerState, props: {
                    id: props.id,
                    disabled: props.disabled,
                    readOnly: props.readOnly,
                    placeholderValue: props.placeholderValue,
                    timeZone: props.timeZone,
                } })), _jsx(TextInput, Object.assign({}, fieldProps, { controlProps: Object.assign(Object.assign({}, fieldProps.controlProps), { disabled: isMobile && state.mode === 'absolute', className: b('input', { mobile: isMobile && state.mode === 'absolute' }) }), hasClear: props.hasClear && !(isMobile && state.mode === 'absolute'), startContent: _jsx(Button, Object.assign({}, modeSwitcherProps, { children: _jsx(Icon, { data: FunctionIcon }) })), endContent: _jsxs(React.Fragment, { children: [!isMobile && !isOnlyTime && (_jsx(Button, Object.assign({}, calendarButtonProps, { children: _jsx(Icon, { data: CalendarIcon }) }))), !isMobile && isOnlyTime && (_jsx(StubButton, { size: calendarButtonProps.size, icon: ClockIcon })), isMobile && state.mode === 'absolute' && (_jsx(StubButton, { size: calendarButtonProps.size, icon: isOnlyTime ? ClockIcon : CalendarIcon }))] }) })), _jsx(HiddenInput, { name: props.name, value: (_a = state.value) === null || _a === void 0 ? void 0 : _a.type, disabled: state.disabled, form: props.form }), _jsx(HiddenInput, { name: props.name, value: state.value, toStringValue: (value) => getNativeValue(value), onReset: (value) => {
                    state.setValue(value);
                }, disabled: state.disabled, form: props.form }), !isMobile && !isOnlyTime && (_jsx(Popup, Object.assign({}, popupProps, { anchorElement: anchor, children: _jsxs("div", { className: b('popup-content'), children: [typeof props.children === 'function' ? (props.children(calendarProps)) : (_jsx(Calendar, Object.assign({}, calendarProps))), state.datePickerState.formatInfo.hasTime && (_jsx("div", { className: b('time-field-wrapper'), children: _jsx(DateField, Object.assign({}, timeInputProps)) }))] }) })))] })));
}
function getNativeValue(value) {
    if (!value) {
        return '';
    }
    if (value.type === 'relative') {
        return value.value;
    }
    return value.value.toISOString();
}
