// @gravity-ui/date-components/dist/esm/components/RelativeRangeDatePicker/components/PickerDialog/PickerDoc.js
'use client';
import { __rest } from "tslib";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { CircleQuestion } from '@gravity-ui/icons';
import { Button, Icon, Popover, Sheet, Table, useMobile } from '@gravity-ui/uikit';
import { block } from "../../../../utils/cn.js";
import { getButtonSizeForInput } from "../../../utils/getButtonSizeForInput.js";
import { PresetTitle } from "../Presets/defaultPresets.js";
import { i18n } from "../Presets/i18n/index.js";
import "./PickerDoc.css";
const b = block('relative-range-date-picker-doc');
const data = [
    {
        title: _jsx(PresetTitle, { title: "Last 5 minutes" }),
        from: 'now - 5m',
        to: 'now',
    },
    {
        title: _jsx(PresetTitle, { title: "From start of day" }),
        from: 'now/d',
        to: 'now',
    },
    {
        title: _jsx(PresetTitle, { title: "This week" }),
        from: 'now/w',
        to: 'now/w',
    },
    {
        title: _jsx(PresetTitle, { title: "From start of week" }),
        from: 'now/w',
        to: 'now',
    },
    {
        title: _jsx(PresetTitle, { title: "Previous month" }),
        from: 'now - 1M/M',
        to: 'now - 1M/M',
    },
];
function DocContent({ size, docs, onStartUpdate, onEndUpdate }) {
    const isMobile = useMobile();
    const { t } = i18n.useTranslation();
    const columns = React.useMemo(() => [
        {
            id: 'title',
            name: () => {
                return t('Range');
            },
        },
        {
            id: 'from',
            name: () => {
                return t('From');
            },
            template: (item) => (_jsx(Button, { size: isMobile ? 'l' : getButtonSizeForInput(size), onClick: () => onStartUpdate(item.from), children: item.from })),
        },
        {
            id: 'to',
            name: () => {
                return t('To');
            },
            template: (item) => (_jsx(Button, { size: isMobile ? 'l' : getButtonSizeForInput(size), onClick: () => onEndUpdate(item.to), children: item.to })),
        },
    ], [isMobile, onEndUpdate, onStartUpdate, size, t]);
    return _jsx(Table, { columns: columns, data: docs, className: b('table', { size }), wordWrap: true });
}
function DesktopDoc(_a) {
    var { className, size } = _a, props = __rest(_a, ["className", "size"]);
    return (_jsx(Popover, { className: b(null), hasArrow: false, placement: ['right-start', 'left-start'], content: _jsx(DocContent, Object.assign({ size: size }, props)), children: _jsx(Button, { className: b('button', className), view: "flat-secondary", size: getButtonSizeForInput(size), children: _jsx(Icon, { data: CircleQuestion }) }) }));
}
function MobileDoc(_a) {
    var { className } = _a, props = __rest(_a, ["className"]);
    const [open, setOpen] = React.useState(false);
    return (_jsxs("div", { className: b(null, className), children: [_jsx(Button, { className: b('button'), view: "flat-secondary", size: "l", onClick: () => {
                    setOpen(true);
                }, children: _jsx(Icon, { data: CircleQuestion }) }), _jsx(Sheet, { visible: open, onClose: () => setOpen(false), children: _jsx(DocContent, Object.assign({}, props)) })] }));
}
export function PickerDoc(_a) {
    var { docs = data } = _a, props = __rest(_a, ["docs"]);
    const isMobile = useMobile();
    if (!Array.isArray(docs) || docs.length === 0) {
        return null;
    }
    if (isMobile) {
        return _jsx(MobileDoc, Object.assign({}, props, { docs: docs }));
    }
    return _jsx(DesktopDoc, Object.assign({}, props, { docs: docs }));
}
