// @gravity-ui/date-components/dist/esm/components/RelativeRangeDatePicker/components/PickerDialog/PickerForm.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Button, Text } from '@gravity-ui/uikit';
import { block } from "../../../../utils/cn.js";
import { RelativeDatePicker } from "../../../RelativeDatePicker/index.js";
import { Presets } from "../Presets/Presets.js";
import { Zones } from "../Zones/Zones.js";
import { PickerDoc } from "./PickerDoc.js";
import { i18n } from "./i18n/index.js";
import { useRelativeRangeDatePickerDialogState } from "./useRelativeRangeDatePickerDialogState.js";
import "./PickerForm.css";
const b = block('relative-range-date-picker-form');
export function PickerForm(props) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const state = useRelativeRangeDatePickerDialogState(props);
    const fieldProps = {
        timeZone: state.timeZone,
        format: props.format,
        minValue: props.minValue,
        maxValue: props.maxValue,
        hasClear: props.allowNullableValues,
        readOnly: props.readOnly,
        size: props.size,
        errorPlacement: 'inside',
    };
    const { isDateUnavailable, withHeader = false } = props;
    const { t } = i18n.useTranslation();
    return (_jsxs("div", { className: b({ size: props.size }, props.className), style: props.style, children: [withHeader && (_jsxs("div", { className: b('header'), children: [_jsx(Text, { variant: props.size === 'xl' ? 'subheader-3' : 'subheader-2', children: t('Select the interval') }), _jsx(PickerDoc, { size: props.size, docs: props.docs, onStartUpdate: (start) => {
                            state.setStart({ type: 'relative', value: start });
                        }, onEndUpdate: (end) => {
                            state.setEnd({ type: 'relative', value: end });
                        } })] })), _jsxs("div", { className: b('pickers'), children: [_jsx(RelativeDatePicker, Object.assign({}, fieldProps, { isDateUnavailable: isDateUnavailable ? (date) => isDateUnavailable(date, 'start') : undefined, validationState: ((_a = state.startValidation) === null || _a === void 0 ? void 0 : _a.isInvalid) ? 'invalid' : undefined, errorMessage: ((_c = (_b = state.startValidation) === null || _b === void 0 ? void 0 : _b.errors) === null || _c === void 0 ? void 0 : _c.join('\n')) || t('Value is incorrect.'), placeholderValue: (_d = props.placeholderValue) === null || _d === void 0 ? void 0 : _d.startOf('day'), label: t('From'), value: state.start, onUpdate: state.setStart })), _jsx(RelativeDatePicker, Object.assign({}, fieldProps, { isDateUnavailable: isDateUnavailable ? (date) => isDateUnavailable(date, 'end') : undefined, validationState: ((_e = state.endValidation) === null || _e === void 0 ? void 0 : _e.isInvalid) ? 'invalid' : undefined, errorMessage: ((_g = (_f = state.endValidation) === null || _f === void 0 ? void 0 : _f.errors) === null || _g === void 0 ? void 0 : _g.join('\n')) || t('Value is incorrect.'), placeholderValue: (_h = props.placeholderValue) === null || _h === void 0 ? void 0 : _h.endOf('day'), label: t('To'), value: state.end, onUpdate: state.setEnd, roundUp: true }))] }), props.withApplyButton && !props.readOnly ? (_jsx(Button, { disabled: state.isInvalid, size: props.size, onClick: () => {
                    state.applyValue();
                    props.onApply();
                }, className: b('apply'), width: "max", children: t('Apply') })) : null, props.withPresets && !props.readOnly ? (_jsx(Presets, { size: props.size, presetTabs: props.presetTabs, onChoosePreset: (start, end) => {
                    state.setRange({ type: 'relative', value: start }, { type: 'relative', value: end }, props.applyPresetsImmediately);
                    if (!props.withApplyButton || props.applyPresetsImmediately) {
                        props.onApply();
                    }
                }, minValue: props.minValue, className: b('presets') })) : null, props.withZonesList ? (_jsx("div", { className: b('zone'), children: _jsx(Zones, { value: state.timeZone, onUpdate: state.setTimeZone, disabled: props.readOnly, size: props.size }) })) : null] }));
}
