// @gravity-ui/date-components/dist/cjs/components/DatePicker/MobileCalendar.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MobileCalendar = MobileCalendar;
const jsx_runtime_1 = require("react/jsx-runtime");
const date_utils_1 = require("@gravity-ui/date-utils");
const cn_1 = require("../../utils/cn.js");
const dates_1 = require("../utils/dates.js");
require("./MobileCalendar.css");
const b = (0, cn_1.block)('mobile-calendar');
function MobileCalendar({ props, state }) {
    var _a, _b;
    let type = 'date';
    if (state.formatInfo.hasTime && state.formatInfo.hasDate) {
        type = 'datetime-local';
    }
    else if (state.formatInfo.hasTime) {
        type = 'time';
    }
    return ((0, jsx_runtime_1.jsx)("input", { className: b(), disabled: props.disabled, type: type, value: formatNative(state.dateFieldState.value, type), id: props.id, min: formatNative((_a = props.minValue) === null || _a === void 0 ? void 0 : _a.timeZone(state.timeZone), type), max: formatNative((_b = props.maxValue) === null || _b === void 0 ? void 0 : _b.timeZone(state.timeZone), type), tabIndex: -1, onChange: (e) => {
            var _a, _b;
            if (props.readOnly) {
                return;
            }
            const newValue = e.target.value;
            if (newValue) {
                const localDate = (0, date_utils_1.dateTime)({
                    input: newValue,
                    format: getDateFormat(type),
                    timeZone: 'system',
                }).timeZone(state.timeZone, true);
                let newDate = state.formatInfo.hasDate
                    ? localDate
                    : (0, dates_1.createPlaceholderValue)({
                        placeholderValue: (_a = props.placeholderValue) === null || _a === void 0 ? void 0 : _a.timeZone(state.timeZone),
                        timeZone: state.timeZone,
                    });
                if (state.formatInfo.hasTime) {
                    newDate = (0, dates_1.mergeDateTime)(newDate, localDate);
                }
                else if (state.value) {
                    newDate = (0, dates_1.mergeDateTime)(newDate, state.value.timeZone(state.timeZone));
                }
                else {
                    newDate = (0, dates_1.mergeDateTime)(newDate, (0, dates_1.createPlaceholderValue)({
                        placeholderValue: (_b = props.placeholderValue) === null || _b === void 0 ? void 0 : _b.timeZone(state.timeZone),
                        timeZone: state.timeZone,
                    }));
                }
                state.setValue(newDate);
            }
            else {
                state.setValue(null);
            }
        } }));
}
function getDateFormat(type) {
    switch (type) {
        case 'time': {
            return 'HH:mm';
        }
        case 'datetime-local': {
            return 'YYYY-MM-DDTHH:mm';
        }
        default: {
            return 'YYYY-MM-DD';
        }
    }
}
function formatNative(date, type) {
    if (!date) {
        return '';
    }
    const format = getDateFormat(type);
    return date.format(format);
}
