// @gravity-ui/date-components/dist/cjs/components/RangeDatePicker/hooks/useRangeDatePickerState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRangeDatePickerState = void 0;
const utils_1 = require("../../DateField/utils.js");
const datePickerStateFactory_1 = require("../../DatePicker/hooks/datePickerStateFactory.js");
const utils_2 = require("../../DatePicker/utils/index.js");
const RangeDateField_1 = require("../../RangeDateField/index.js");
const utils_3 = require("../../RangeDateField/utils/index.js");
const dates_1 = require("../../utils/dates.js");
exports.useRangeDatePickerState = (0, datePickerStateFactory_1.datePickerStateFactory)({
    getPlaceholderTime,
    mergeDateTime: mergeRangeDateTime,
    setTimezone,
    getDateTime: utils_2.getDateTimeValue,
    useDateFieldState: RangeDateField_1.useRangeDateFieldState,
    adjustDateToFormat: adjustRangeToFormat,
});
function getPlaceholderTime(placeholderValue, timeZone) {
    return (0, utils_3.createPlaceholderRangeValue)({ placeholderValue, timeZone });
}
function mergeRangeDateTime(date, time) {
    const start = (0, dates_1.mergeDateTime)(date.start, time.start);
    const end = (0, dates_1.mergeDateTime)(date.end, time.end);
    return { start, end };
}
function setTimezone(date, timeZone) {
    const start = date.start.timeZone(timeZone);
    const end = date.end.timeZone(timeZone);
    return { start, end };
}
function adjustRangeToFormat(date, sectionsInfo) {
    const start = (0, utils_1.adjustDateToFormat)(date.start, sectionsInfo, 'startOf');
    const end = (0, utils_1.adjustDateToFormat)(date.end, sectionsInfo, 'endOf');
    return { start, end };
}
