// @gravity-ui/date-components/dist/esm/components/RelativeRangeDatePicker/components/Control/Control.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { Calendar as CalendarIcon } from '@gravity-ui/icons';
import { Button, Icon, TextInput, useLang } from '@gravity-ui/uikit';
import { block } from "../../../../utils/cn.js";
import { getButtonSizeForInput } from "../../../utils/getButtonSizeForInput.js";
import { getDefaultTitle } from "../../utils.js";
import { i18n as i18nPresets } from "../Presets/i18n/index.js";
import { i18n } from "./i18n/index.js";
import "./Control.css";
const b = block('relative-range-date-picker-control');
export const Control = React.forwardRef(({ props, state, open, setOpen, isMobile, onClick, onKeyDown, onFocus, onClickCalendar, onUpdate, }, ref) => {
    var _a;
    const { alwaysShowAsAbsolute, presetTabs, getRangeTitle } = props;
    const format = props.format || 'L';
    const { t } = i18n.useTranslation();
    const { t: presetsTranslations } = i18nPresets.useTranslation();
    const { lang } = useLang();
    const text = React.useMemo(() => typeof getRangeTitle === 'function'
        ? getRangeTitle(state.value, state.timeZone)
        : getDefaultTitle({
            value: state.value,
            timeZone: state.timeZone,
            alwaysShowAsAbsolute,
            format,
            presets: presetTabs === null || presetTabs === void 0 ? void 0 : presetTabs.flatMap(({ presets }) => presets),
            presetsTranslations,
            lang,
        }), [
        alwaysShowAsAbsolute,
        format,
        getRangeTitle,
        lang,
        presetTabs,
        presetsTranslations,
        state.timeZone,
        state.value,
    ]);
    const validationState = props.validationState || (state.isInvalid ? 'invalid' : undefined);
    const errorMessage = (_a = props.errorMessage) !== null && _a !== void 0 ? _a : state.errors.join('\n');
    const triggerProps = {
        id: props.id,
        role: 'combobox',
        'aria-haspopup': 'dialog',
        'aria-expanded': open,
        'aria-label': props['aria-label'],
        'aria-labelledby': props['aria-labelledby'],
        'aria-describedby': props['aria-describedby'],
        'aria-details': props['aria-details'],
        disabled: props.disabled,
        readOnly: props.readOnly,
        onClick: onClickCalendar,
        onKeyDown,
    };
    return props.renderControl ? (props.renderControl({
        ref,
        value: state.value,
        title: text,
        validationState,
        errorMessage,
        open,
        setOpen,
        triggerProps,
    })) : (_jsxs(React.Fragment, { children: [_jsx(TextInput, { id: props.id, autoFocus: props.autoFocus, controlRef: ref, value: text, placeholder: props.placeholder, onUpdate: onUpdate, controlProps: Object.assign(Object.assign({ className: b('input', { mobile: isMobile }) }, triggerProps), { disabled: isMobile, onClick }), onKeyDown: onKeyDown, onFocus: onFocus, validationState: validationState, errorMessage: errorMessage, errorPlacement: props.errorPlacement, pin: props.pin, size: props.size, label: props.label, hasClear: props.hasClear, disabled: props.disabled, endContent: _jsx(Button, { view: "flat-secondary", size: getButtonSizeForInput(props.size), disabled: props.disabled, "aria-haspopup": "dialog", "aria-expanded": open, "aria-label": t('Range date picker'), onClick: onClickCalendar, tabIndex: -1, children: _jsx(Icon, { data: CalendarIcon }) }) }), isMobile ? (_jsx("button", { className: b('mobile-trigger', {
                    'has-clear': Boolean(props.hasClear && state.value),
                    'has-errors': state.isInvalid && props.errorPlacement === 'inside',
                    size: props.size,
                }), onClick: onClick })) : null] }));
});
Control.displayName = 'Control';
