// @gravity-ui/date-components/dist/cjs/components/RangeDateSelection/components/Ticks/Ticks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeSlitTicksGeometry = exports.makeMiddleTicksGeometry = void 0;
exports.Ticks = Ticks;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = tslib_1.__importDefault(require("react"));
const date_utils_1 = require("@gravity-ui/date-utils");
const cn_1 = require("../../../../utils/cn.js");
const span_1 = require("../../utils/span.js");
const Ruler_1 = require("../Ruler/Ruler.js");
const utils_1 = require("./utils.js");
Object.defineProperty(exports, "makeMiddleTicksGeometry", { enumerable: true, get: function () { return utils_1.makeMiddleTicksGeometry; } });
Object.defineProperty(exports, "makeSlitTicksGeometry", { enumerable: true, get: function () { return utils_1.makeSlitTicksGeometry; } });
require("./Ticks.css");
const b = (0, cn_1.block)('timeline-ticks');
function Ticks({ theme, minTickWidth, maxTickWidth, hasLabels, geometry, className, labelsClassName, formatTime, timeZone, }) {
    const viewport = (0, Ruler_1.useViewportDimensions)();
    const interval = (0, Ruler_1.useViewportInterval)();
    const tickTimes = react_1.default.useMemo(() => (0, utils_1.calculateTickTimes)({
        minTickWidth,
        maxTickWidth,
        viewportStart: interval.start,
        viewportEnd: interval.end,
        viewportWidth: viewport.width,
        timeZone,
    }), [maxTickWidth, minTickWidth, interval, viewport.width, timeZone]);
    const tickCoords = react_1.default.useMemo(() => {
        const timeToXCoeff = (0, span_1.getLerpCoeff)({ start: interval.start.valueOf(), end: interval.end.valueOf() }, { start: 0, end: viewport.width });
        return tickTimes.map((t) => 0 + Math.round((t.valueOf() - interval.start.valueOf()) * timeToXCoeff));
    }, [tickTimes, interval, viewport.width]);
    const ticksGeometry = react_1.default.useMemo(() => {
        return tickCoords.map(geometry).join('');
    }, [geometry, tickCoords]);
    const ticksPath = (0, jsx_runtime_1.jsx)("path", { className: b({ theme }, className), d: ticksGeometry });
    const tickLabels = react_1.default.useMemo(() => {
        if (!hasLabels || tickCoords.length < 2) {
            return null;
        }
        const result = [];
        for (let i = 0; i < tickCoords.length; i++) {
            const x = tickCoords[i];
            const t = tickTimes[i];
            const dt = (0, date_utils_1.dateTime)({ input: t, timeZone });
            result.push((0, jsx_runtime_1.jsx)("text", { className: b('label', labelsClassName), dominantBaseline: "middle", x: x, y: "50%", filter: "url(#g-date-tt-ticklabelbg)", children: formatTime ? formatTime(dt) : (0, utils_1.formatTick)(dt) }, i));
        }
        return result;
    }, [hasLabels, tickCoords, tickTimes, timeZone, labelsClassName, formatTime]);
    return ((0, jsx_runtime_1.jsxs)(react_1.default.Fragment, { children: [ticksPath, tickLabels] }));
}
