// @gravity-ui/date-components/dist/esm/components/DatePicker/hooks/datePickerStateFactory.js
import React from 'react';
import { useControlledState } from '@gravity-ui/uikit';
import { useDefaultTimeZone } from "../../utils/useDefaultTimeZone.js";
export function datePickerStateFactory({ getPlaceholderTime, mergeDateTime, setTimezone, getDateTime, useDateFieldState, adjustDateToFormat, }) {
    return function useDatePickerState(props) {
        var _a, _b;
        const { disabled, readOnly } = props;
        const [isOpen, _setOpen] = useControlledState(props.open, (_a = props.defaultOpen) !== null && _a !== void 0 ? _a : false, props.onOpenChange);
        const setOpen = _setOpen;
        const [value, setValue] = useControlledState(props.value, (_b = props.defaultValue) !== null && _b !== void 0 ? _b : null, props.onUpdate);
        const [selectedDateInner, setSelectedDate] = React.useState(null);
        const [selectedTimeInner, setSelectedTime] = React.useState(null);
        const inputTimeZone = useDefaultTimeZone(getDateTime(props.value) || getDateTime(props.defaultValue) || props.placeholderValue);
        const timeZone = props.timeZone || inputTimeZone;
        let selectedDate = selectedDateInner;
        let selectedTime = selectedTimeInner;
        const format = props.format || 'L';
        const commitValue = (date, time) => {
            if (disabled || readOnly) {
                return;
            }
            setValue(setTimezone(mergeDateTime(date, time), inputTimeZone));
            setSelectedDate(null);
            setSelectedTime(null);
        };
        const dateFieldState = useDateFieldState(Object.assign(Object.assign({}, props), { value,
            onUpdate(date) {
                if (date) {
                    commitValue(date, date);
                }
                else {
                    setValue(null);
                }
            },
            disabled,
            readOnly, validationState: props.validationState, minValue: props.minValue, maxValue: props.maxValue, isDateUnavailable: props.isDateUnavailable, format, placeholderValue: props.placeholderValue, timeZone }));
        const timeFormat = React.useMemo(() => {
            if (!dateFieldState.formatInfo.hasTime) {
                return undefined;
            }
            const timeFormatParts = [];
            const hours = dateFieldState.sections.find((s) => s.type === 'hour');
            if (hours) {
                timeFormatParts.push(hours.format);
            }
            const minutes = dateFieldState.sections.find((s) => s.type === 'minute');
            if (minutes) {
                timeFormatParts.push(minutes.format);
            }
            const seconds = dateFieldState.sections.find((s) => s.type === 'second');
            if (seconds) {
                timeFormatParts.push(seconds.format);
            }
            const dayPeriod = dateFieldState.sections.find((s) => s.type === 'dayPeriod');
            return timeFormatParts.join(':') + (dayPeriod ? ` ${dayPeriod.format}` : '');
        }, [dateFieldState.formatInfo.hasTime, dateFieldState.sections]);
        if (value) {
            selectedDate = setTimezone(value, timeZone);
            if (dateFieldState.formatInfo.hasTime) {
                selectedTime = selectedDate;
            }
        }
        // Intercept setValue to make sure the Time section is not changed by date selection in Calendar
        const selectDate = (newValue) => {
            if (disabled || readOnly) {
                return;
            }
            const newDate = adjustDateToFormat(newValue, dateFieldState.formatInfo);
            const shouldClose = !dateFieldState.formatInfo.hasTime;
            if (dateFieldState.formatInfo.hasTime) {
                if (selectedTime || shouldClose) {
                    commitValue(newDate, selectedTime || newDate);
                }
                else {
                    setSelectedDate(newDate);
                }
            }
            else {
                commitValue(newDate, newDate);
            }
            if (shouldClose) {
                setOpen(false, 'ValueSelected');
            }
        };
        const selectTime = (newValue) => {
            if (disabled || readOnly) {
                return;
            }
            const newTime = adjustDateToFormat(newValue !== null && newValue !== void 0 ? newValue : getPlaceholderTime(props.placeholderValue, timeZone), dateFieldState.formatInfo);
            if (selectedDate) {
                commitValue(selectedDate, newTime);
            }
            else {
                setSelectedTime(newTime);
            }
        };
        if (dateFieldState.formatInfo.hasTime && !selectedTime) {
            selectedTime = dateFieldState.displayValue;
        }
        return {
            value,
            setValue(newDate) {
                if (props.readOnly || props.disabled) {
                    return;
                }
                if (newDate) {
                    setValue(setTimezone(newDate, inputTimeZone));
                }
                else {
                    setValue(null);
                }
            },
            dateValue: selectedDate,
            timeValue: selectedTime,
            setDateValue: selectDate,
            setTimeValue: selectTime,
            disabled,
            readOnly,
            format,
            formatInfo: dateFieldState.formatInfo,
            hasDate: dateFieldState.formatInfo.hasDate,
            hasTime: dateFieldState.formatInfo.hasTime,
            timeFormat,
            timeZone,
            isOpen,
            setOpen(newIsOpen, reason) {
                if (!newIsOpen && !value && selectedDate && dateFieldState.formatInfo.hasTime) {
                    commitValue(selectedDate, selectedTime || getPlaceholderTime(props.placeholderValue, props.timeZone));
                }
                setOpen(newIsOpen, reason);
            },
            dateFieldState,
        };
    };
}
