// @gravity-ui/date-components/dist/esm/components/DatePicker/hooks/useDatePickerProps.js
import { __rest } from "tslib";
import React from 'react';
import { isDateTime } from '@gravity-ui/date-utils';
import { useFocusWithin, useForkRef } from '@gravity-ui/uikit';
import { useDateFieldProps } from "../../DateField/index.js";
import { filterDOMProps } from "../../utils/filterDOMProps.js";
import { getButtonSizeForInput } from "../../utils/getButtonSizeForInput.js";
import { mergeProps } from "../../utils/mergeProps.js";
import { i18n } from "../i18n/index.js";
import { getCalendarModes, getDateTimeValue } from "../utils/index.js";
export function useDatePickerProps(state, _a) {
    var { onFocus, onBlur } = _a, props = __rest(_a, ["onFocus", "onBlur"]);
    const [isActive, setActive] = React.useState(false);
    const [focusedDate, setFocusedDate] = React.useState(isDateTime(state.dateFieldState.displayValue)
        ? state.dateFieldState.displayValue
        : state.dateFieldState.displayValue.start);
    const [prevDateValue, setPrevDateValue] = React.useState(state.dateFieldState.displayValue);
    if (isDateTime(prevDateValue)) {
        if (isDateTime(state.dateFieldState.displayValue) &&
            !state.dateFieldState.displayValue.isSame(prevDateValue, 'day')) {
            setPrevDateValue(state.dateFieldState.displayValue);
            setFocusedDate(state.dateFieldState.displayValue);
        }
    }
    else if (!isDateTime(state.dateFieldState.displayValue)) {
        if (!state.dateFieldState.displayValue.start.isSame(prevDateValue.start, 'day')) {
            setPrevDateValue(state.dateFieldState.displayValue);
            setFocusedDate(state.dateFieldState.displayValue.start);
        }
        else if (!state.dateFieldState.displayValue.end.isSame(prevDateValue.end, 'day')) {
            setPrevDateValue(state.dateFieldState.displayValue);
            setFocusedDate(state.dateFieldState.displayValue.end);
        }
    }
    const { focusWithinProps } = useFocusWithin({
        onFocusWithin: onFocus,
        onBlurWithin: onBlur,
        onFocusWithinChange(isFocusWithin) {
            setActive(isFocusWithin);
            if (!isFocusWithin) {
                state.setOpen(false, 'FocusOut');
            }
        },
    });
    const { inputProps } = useDateFieldProps(state.dateFieldState, props);
    const inputRef = React.useRef(null);
    const handleRef = useForkRef(inputRef, inputProps.controlRef);
    const calendarRef = React.useRef(null);
    const calendarButtonRef = React.useRef(null);
    const groupRef = React.useRef(null);
    function focusInput() {
        setTimeout(() => {
            var _a;
            (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus({ preventScroll: true });
        });
    }
    const onlyTime = state.formatInfo.hasTime && !state.formatInfo.hasDate;
    const calendarModes = getCalendarModes(state.formatInfo);
    const { t } = i18n.useTranslation();
    const DOMProps = filterDOMProps(props);
    delete DOMProps.id;
    return {
        groupProps: Object.assign(Object.assign(Object.assign(Object.assign({}, DOMProps), { ref: groupRef, tabIndex: -1, role: 'group' }), focusWithinProps), { style: props.style, 'aria-disabled': state.disabled || undefined, onKeyDown: (e) => {
                if (!onlyTime && e.altKey && (e.key === 'ArrowDown' || e.key === 'ArrowUp')) {
                    e.preventDefault();
                    e.stopPropagation();
                    state.setOpen(true, 'ShortcutKeyDown');
                }
            } }),
        fieldProps: mergeProps(inputProps, state.dateFieldState.isEmpty && !isActive && props.placeholder
            ? { value: '' }
            : undefined, {
            controlRef: handleRef,
            controlProps: { role: 'combobox', 'aria-expanded': state.isOpen },
        }),
        calendarButtonProps: {
            ref: calendarButtonRef,
            size: getButtonSizeForInput(props.size),
            disabled: state.disabled,
            'aria-label': t('Calendar'),
            'aria-haspopup': 'dialog',
            'aria-expanded': state.isOpen,
            view: 'flat-secondary',
            onClick: () => {
                setActive(true);
                state.setOpen(!state.isOpen, 'TriggerButtonClick');
            },
        },
        popupProps: {
            open: state.isOpen,
            onOpenChange: (open, e, reason) => {
                var _a, _b;
                if (!open) {
                    if (reason === 'escape-key') {
                        state.setOpen(false, 'EscapeKeyDown');
                        focusInput();
                    }
                    else if (reason === 'outside-press') {
                        if (!(e === null || e === void 0 ? void 0 : e.target) || !((_a = calendarButtonRef.current) === null || _a === void 0 ? void 0 : _a.contains(e.target))) {
                            state.setOpen(false, 'ClickOutside');
                        }
                        if ((e === null || e === void 0 ? void 0 : e.target) && ((_b = groupRef.current) === null || _b === void 0 ? void 0 : _b.contains(e.target))) {
                            focusInput();
                        }
                    }
                }
            },
            onTransitionOutComplete: () => {
                setFocusedDate(isDateTime(state.dateFieldState.displayValue)
                    ? state.dateFieldState.displayValue
                    : state.dateFieldState.displayValue.start);
            },
            modal: !props.disableFocusTrap,
            returnFocus: !props.disableFocusTrap,
            disablePortal: props.disablePortal,
            placement: props.popupPlacement,
            offset: props.popupOffset,
            className: props.popupClassName,
            style: props.popupStyle,
        },
        calendarProps: {
            ref: calendarRef,
            autoFocus: true,
            size: props.size === 's' ? 'm' : props.size,
            disabled: props.disabled,
            readOnly: props.readOnly,
            onUpdate: (d) => {
                state.setDateValue(d);
                if (!state.formatInfo.hasTime) {
                    focusInput();
                }
            },
            value: state.dateFieldState.displayValue,
            minValue: props.minValue,
            maxValue: props.maxValue,
            isDateUnavailable: props.isDateUnavailable,
            timeZone: state.timeZone,
            focusedValue: focusedDate,
            onFocusUpdate: setFocusedDate,
            modes: calendarModes,
        },
        timeInputProps: {
            value: state.timeValue,
            placeholderValue: getDateTimeValue(state.dateFieldState.displayValue),
            onUpdate: state.setTimeValue,
            format: state.timeFormat,
            readOnly: state.readOnly,
            disabled: state.disabled,
            timeZone: state.timeZone,
            hasClear: props.hasClear,
            size: props.size,
        },
    };
}
