// @statoscope/cli/dist/commands/templates/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginTemplate = void 0;
const templateParts_1 = require("./templateParts");
const pluginTemplate = (templateOptions) => {
    const plugin = (0, templateParts_1.templatePartsByOptions)(templateOptions, {
        js: jsPlugin,
        ts: tsPlugin,
    });
    return (0, templateParts_1.concatTemplateParts)(plugin);
};
exports.pluginTemplate = pluginTemplate;
const jsPlugin = {
    imports: {
        commonjs: `const asyncChunkRule = require('./rule.statoscope.js')`,
        esm: `import asyncChunkRule from './rule.statoscope.js';`,
    },
    body: `/**
  * @typedef {import('@statoscope/webpack-model').Prepared} Prepared
  * @returns {import('@statoscope/stats-validator/dist/plugin').PluginDescription<Prepared>} PluginDescription
 */
const newStatoscopePlugin = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
    exports: {
        commonjs: 'module.exports = newStatoscopePlugin',
        esm: 'export default newStatoscopePlugin',
    },
};
const tsPlugin = {
    imports: `import type { PluginFn } from '@statoscope/stats-validator/dist/plugin';
import type { Prepared } from '@statoscope/webpack-model';
import asyncChunkRule from './rule.statoscope.ts';`,
    body: `
const newStatoscopePlugin: PluginFn<Prepared> = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
    exports: 'export default newStatoscopePlugin',
};
