// @statoscope/cli/dist/commands/templates/reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reporterTemplate = void 0;
const templateParts_1 = require("./templateParts");
const reporterTemplate = (templateOptions) => {
    const reporter = (0, templateParts_1.templatePartsByOptions)(templateOptions, {
        js: jsReporter,
        ts: tsReporter,
    });
    return (0, templateParts_1.concatTemplateParts)(reporter);
};
exports.reporterTemplate = reporterTemplate;
const tsReporter = {
    imports: `import type { Reporter } from "@statoscope/types/types/validation/reporter";
import type { Result } from "@statoscope/types/types/validation/result";`,
    body: `class NewReporter implements Reporter {
  constructor() {}
  async run(result: Result): Promise<void> {
    console.log("Hello from new custom reporter!");
    console.log({ result });
  }
}`,
    exports: `export default NewReporter;`,
};
const jsReporter = {
    imports: {
        commonjs: '',
        esm: '',
    },
    body: `class NewReporter {
  constructor() {}
  /**
   * @typedef {import('@statoscope/types/types/validation/result').Result} Result
   * @param {Result} result
   * @returns {Promise<void>}
   */
  async run(result) {
    console.log("Hello from new custom reporter!");
    console.log({ result });
  }
}`,
    exports: {
        commonjs: `module.exports = NewReporter;`,
        esm: `export default NewReporter;`,
    },
};
