// @statoscope/cli/dist/commands/templates/rule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ruleTemplate = void 0;
const templateParts_1 = require("./templateParts");
const ruleTemplate = (templateOptions) => {
    const rule = (0, templateParts_1.templatePartsByOptions)(templateOptions, {
        js: jsRule,
        ts: tsRule,
    });
    return (0, templateParts_1.concatTemplateParts)(rule);
};
exports.ruleTemplate = ruleTemplate;
const jsRuleBody = `/**
 * @typedef {import('@statoscope/types/types/validation/api').API} API
 * @typedef {import('@statoscope/webpack-model').Prepared} Prepared
 * @param  {string[]} ruleParams
 * @param  {Prepared} data
 * @param  {API} api
 * @returns void
 */
const asyncChunkRule = (ruleParams, data, api) => {
  const normalizedParams = ruleParams ? ruleParams : [];
  // we can use jora-queries in rules
  // const chunks = data.query('compilations.chunks');
  const chunks = [...new Set(data.compilations.map((c) => c.chunks).flat())];

  chunks
    .filter((c) => c.names.some((n) => normalizedParams.includes(n) && c.initial))
    .forEach((f) => {
      api.message('Chunk ' + f.names[0] + ' should be async.',{
         compilation:data.compilations[0].hash,
         filename:"All build files"
      });
    });
};`;
const jsRule = {
    imports: {
        commonjs: '',
        esm: '',
    },
    body: jsRuleBody,
    exports: {
        commonjs: 'module.exports = asyncChunkRule;',
        esm: 'export default asyncChunkRule',
    },
};
const tsRuleBody = `type WebpackRule<TParams> = Rule<TParams, Prepared>;
type Params = string[];

const asyncChunkRule: WebpackRule<Params> = (ruleParams, data, api): void => {
  const normalizedParams: string[] = ruleParams ? ruleParams : [];
  // we can use jora-queries in rules
  // const chunks = data.query('compilations.chunks');
  const chunks = [...new Set(data.compilations.map((c) => c.chunks).flat())];

  chunks
    .filter((c) =>
      c.names.some((n) => normalizedParams.includes(n) && c.initial)
    )
    .forEach((f) => {
      api.message('Chunk ' + f.names[0] + ' should be async.',{
         compilation:data.compilations[0].hash,
         filename:"All build files"
      });
    });
};`;
const tsRule = {
    imports: `import type { Rule } from "@statoscope/stats-validator/dist/rule";
import type { Prepared } from "@statoscope/webpack-model";`,
    body: tsRuleBody,
    exports: 'export default asyncChunkRule;',
};
