// @gravity-ui/table/build/cjs/components/TableSettings/TableSettings.utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNestedColumnsCount = exports.isFilteredColumn = exports.isDisplayedColumn = exports.getInitialOrderItems = exports.orderStateToColumnOrder = exports.useOrderedItems = exports.findContainer = void 0;
const React = __importStar(require("react"));
const sortable_1 = require("@dnd-kit/sortable");
const filterColumns = (column) => Boolean(column);
const orderItem = (item, state) => {
    const childrenOrder = state[item.id] ?? [];
    const orderedChildren = childrenOrder
        .map((orderId) => item.columns?.find(({ id }) => id === orderId))
        .filter(filterColumns);
    return {
        ...item,
        columns: orderedChildren?.map((child) => orderItem(child, state)) ?? [],
    };
};
const findContainer = (itemsMap, id) => {
    if (!id)
        return undefined;
    return Object.keys(itemsMap).find((key) => itemsMap[key].includes(id.toString()));
};
exports.findContainer = findContainer;
const useOrderedItems = (items, orderState, setOrderState) => {
    const [activeDepth, setActiveDepth] = React.useState();
    const depthMap = React.useMemo(() => {
        const stack = [...items];
        const result = {};
        while (stack.length) {
            const item = stack.pop();
            if (!item)
                continue;
            result[item.id] = item.depth;
            if (item.columns)
                stack.push(...item.columns);
        }
        return result;
    }, [items]);
    const orderedItems = React.useMemo(() => {
        return (orderState['root']
            .map((orderId) => items.find(({ id }) => id === orderId))
            .filter(filterColumns)
            .map((item) => orderItem(item, orderState)) ?? []);
    }, [orderState, items]);
    const handleDragEnd = ({ active, over }) => {
        setActiveDepth(undefined);
        const activeContainer = (0, exports.findContainer)(orderState, active.id);
        const overContainer = (0, exports.findContainer)(orderState, over?.id);
        if (!activeContainer || !overContainer || activeContainer !== overContainer)
            return;
        const activeIndex = orderState[activeContainer].indexOf(active.id.toString());
        const overIndex = over ? orderState[overContainer].indexOf(over?.id.toString()) : -1;
        if (activeIndex !== overIndex) {
            setOrderState((prevState) => ({
                ...prevState,
                [overContainer]: (0, sortable_1.arrayMove)(orderState[overContainer], activeIndex, overIndex),
            }));
        }
    };
    const handleDragStart = ({ active }) => {
        setActiveDepth(depthMap[active.id]);
    };
    const handleDragCancel = () => {
        setActiveDepth(undefined);
    };
    return { orderedItems, activeDepth, handleDragEnd, handleDragStart, handleDragCancel };
};
exports.useOrderedItems = useOrderedItems;
const orderStateToColumnOrder = (state) => {
    const result = [];
    const stack = [...state['root']];
    while (stack.length) {
        const item = stack.shift();
        if (!item)
            continue;
        const children = state[item];
        if (children.length)
            stack.unshift(...children);
        else
            result.push(item);
    }
    return result;
};
exports.orderStateToColumnOrder = orderStateToColumnOrder;
const getInitialOrderItems = (treeItems, initialOrder) => {
    const orderMap = Object.fromEntries(initialOrder.map((id, index) => [id, index]));
    const stack = [...treeItems];
    const result = { root: treeItems.map(({ id }) => id) };
    const parentNodes = [];
    while (stack.length) {
        const item = stack.shift();
        if (!item)
            continue;
        result[item.id] = item?.columns?.map(({ id }) => id) ?? [];
        if (item.columns.length) {
            stack.push(...item.columns);
            parentNodes.push(item.id);
        }
    }
    parentNodes.reverse().forEach((parent) => {
        const childs = result[parent];
        const childsOrders = childs.map((id) => orderMap[id] ?? -1);
        const minIndex = Math.min(...childsOrders);
        orderMap[parent] = minIndex;
    });
    Object.keys(result).forEach((key) => {
        result[key] = result[key].sort((a, b) => orderMap[a] - orderMap[b]);
    });
    return result;
};
exports.getInitialOrderItems = getInitialOrderItems;
const isDisplayedColumn = (column) => !column.columnDef.meta?.hideInSettings;
exports.isDisplayedColumn = isDisplayedColumn;
const isFilteredColumn = (title, search) => (typeof title === 'string' ? title.toLowerCase().includes(search.toLowerCase()) : false);
exports.isFilteredColumn = isFilteredColumn;
const getNestedColumnsCount = (column) => {
    return column.columns
        ? column.columns.reduce((count, current) => {
            count++;
            if (current.columns.length > 0) {
                return count + (0, exports.getNestedColumnsCount)(current);
            }
            return count;
        }, 0)
        : 0;
};
exports.getNestedColumnsCount = getNestedColumnsCount;
