// @gravity-ui/table/build/cjs/components/TableSettings/TableSettings.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableSettings = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const React = __importStar(require("react"));
const core_1 = require("@dnd-kit/core");
const sortable_1 = require("@dnd-kit/sortable");
const icons_1 = require("@gravity-ui/icons");
const uikit_1 = require("@gravity-ui/uikit");
const react_table_1 = require("@tanstack/react-table");
const debounce_js_1 = __importDefault(require("lodash/debounce.js"));
const TableSettingsColumn_1 = require("../TableSettingsColumn/TableSettingsColumn.js");
const TableSettings_classname_1 = require("./TableSettings.classname.js");
const TableSettings_utils_1 = require("./TableSettings.utils.js");
const hooks_1 = require("./hooks/index.js");
const i18n_1 = __importDefault(require("./i18n/index.js"));
require("./TableSettings.css");
const POPUP_PLACEMENT = ['bottom-end', 'bottom', 'top-end', 'top', 'left', 'right'];
const NESTED_COLUMNS_PREFIX = 'nested_columns_count_';
const TableSettings = ({ table, sortable = true, filterable = true, enableSearch = false, searchPlaceholder = '', onSettingsApply, }) => {
    const anchorRef = React.useRef(null);
    const [open, setOpen] = React.useState(false);
    const [search, setSearch] = React.useState('');
    const [expandNestedColumns, setExpandNestedColumns] = React.useState(false);
    const allColumns = React.useMemo(table.getAllColumns, [table]);
    const filteredColumns = React.useMemo(() => allColumns.filter(TableSettings_utils_1.isDisplayedColumn), [allColumns]);
    const headers = table.getFlatHeaders();
    const headersById = React.useMemo(() => {
        return headers.reduce((acc, header) => {
            const result = { ...acc };
            result[header.column.id] = header;
            return result;
        }, {});
    }, [headers]);
    const { hiddenNodes, partiallyHiddenNodes } = (0, hooks_1.useFilterTableSettings)({
        filteredColumns,
        search,
        expandNestedColumns,
        headersById,
    });
    const [visibilityState, setVisibilityState] = React.useState(() => table.getState().columnVisibility);
    const [orderState, setOrderState] = React.useState(() => (0, TableSettings_utils_1.getInitialOrderItems)(filteredColumns, table.getState().columnOrder));
    const sensors = (0, core_1.useSensors)((0, core_1.useSensor)(core_1.MouseSensor), (0, core_1.useSensor)(core_1.TouchSensor));
    const { orderedItems, activeDepth, handleDragEnd, handleDragStart, handleDragCancel } = (0, TableSettings_utils_1.useOrderedItems)(filteredColumns, orderState, setOrderState);
    const orderedItemIds = React.useMemo(() => orderedItems.map(({ id }) => id), [orderedItems]);
    const renderColumns = (renderedColumns, innerColumn) => {
        const columns = [...renderedColumns];
        const isSortableColumn = sortable && !search.length;
        const level = innerColumn?.depth ?? 0;
        const displayedColumns = columns.filter((column) => !hiddenNodes.has(column.id));
        if (innerColumn && !expandNestedColumns && partiallyHiddenNodes.has(innerColumn.id)) {
            let count = (0, TableSettings_utils_1.getNestedColumnsCount)(innerColumn);
            if (displayedColumns.length) {
                count -= displayedColumns.length;
            }
            displayedColumns.push((0, react_table_1.createColumn)(table, {
                id: `${NESTED_COLUMNS_PREFIX}${innerColumn.id}`,
                header: () => ((0, jsx_runtime_1.jsx)(uikit_1.Text, { variant: "body-1", color: "secondary", children: (0, i18n_1.default)('nested_columns_count', {
                        count,
                    }) })),
                columns: [],
                enableHiding: false,
            }, level + 1, innerColumn));
        }
        return displayedColumns.map((innerColumn) => {
            const children = innerColumn.columns.length
                ? renderColumns(innerColumn.columns.filter(TableSettings_utils_1.isDisplayedColumn), innerColumn)
                : null;
            const header = headersById[innerColumn.id] ?? innerColumn.columnDef.header;
            const isFilterableColumn = innerColumn.id.startsWith(NESTED_COLUMNS_PREFIX)
                ? false
                : filterable;
            return ((0, jsx_runtime_1.jsx)(TableSettingsColumn_1.TableSettingsColumn, { column: innerColumn, header: header, visibilityState: visibilityState, sortable: isSortableColumn, filterable: isFilterableColumn, activeDepth: activeDepth, withOffset: true, onVisibilityToggle: setVisibilityState, children: children }, innerColumn.id));
        });
    };
    const resetSettings = () => {
        setVisibilityState(table.getState().columnVisibility);
        setOrderState((0, TableSettings_utils_1.getInitialOrderItems)(filteredColumns, table.getState().columnOrder));
        setSearch('');
    };
    const cancelEditing = () => {
        setOpen(false);
        resetSettings();
    };
    const togglePopup = () => {
        if (open)
            cancelEditing();
        else
            setOpen(true);
    };
    const applyNewSettings = () => {
        const columnOrder = (0, TableSettings_utils_1.orderStateToColumnOrder)(orderState);
        if (onSettingsApply)
            onSettingsApply({ visibilityState, columnOrder });
        if (filterable)
            table.setColumnVisibility(visibilityState);
        if (sortable)
            table.setColumnOrder(columnOrder);
        setOpen(false);
        setSearch('');
    };
    const updateSearch = (0, debounce_js_1.default)((value) => {
        setSearch(value);
    }, 200);
    const toggleNestedColumns = () => setExpandNestedColumns(!expandNestedColumns);
    const emptyResult = search.length > 0 && !orderedItems.some((item) => !hiddenNodes.has(item.id));
    return ((0, jsx_runtime_1.jsxs)(React.Fragment, { children: [(0, jsx_runtime_1.jsxs)(uikit_1.Popup, { open: open, onClose: cancelEditing, anchorRef: anchorRef, placement: POPUP_PLACEMENT, className: (0, TableSettings_classname_1.b)(), 
                // @ts-ignore
                contentClassName: (0, TableSettings_classname_1.b)(), children: [(0, jsx_runtime_1.jsxs)("div", { className: (0, TableSettings_classname_1.b)('popover-content'), children: [enableSearch && ((0, jsx_runtime_1.jsx)(uikit_1.TextInput, { placeholder: searchPlaceholder, className: (0, TableSettings_classname_1.b)('search-input'), endContent: partiallyHiddenNodes.size > 0 && ((0, jsx_runtime_1.jsx)(uikit_1.Button, { view: "flat-secondary", onClick: toggleNestedColumns, children: expandNestedColumns ? ((0, jsx_runtime_1.jsxs)(uikit_1.Flex, { alignItems: "center", children: [(0, jsx_runtime_1.jsx)(uikit_1.Icon, { data: icons_1.ChevronsDown }), (0, i18n_1.default)('collapse_button')] })) : ((0, jsx_runtime_1.jsxs)(uikit_1.Flex, { alignItems: "center", children: [(0, jsx_runtime_1.jsx)(uikit_1.Icon, { data: icons_1.ChevronsRight }), (0, i18n_1.default)('expand_button')] })) })), onUpdate: updateSearch })), emptyResult ? ((0, jsx_runtime_1.jsx)(uikit_1.Text, { variant: "body-1", color: "secondary", className: (0, TableSettings_classname_1.b)('empty-message'), children: (0, i18n_1.default)('not_found') })) : ((0, jsx_runtime_1.jsx)(core_1.DndContext, { onDragEnd: handleDragEnd, onDragStart: handleDragStart, onDragCancel: handleDragCancel, sensors: sensors, children: (0, jsx_runtime_1.jsx)(sortable_1.SortableContext, { items: orderedItemIds, strategy: sortable_1.verticalListSortingStrategy, children: renderColumns(orderedItems) }) }))] }), (0, jsx_runtime_1.jsx)(uikit_1.Divider, {}), !emptyResult && ((0, jsx_runtime_1.jsx)("div", { className: (0, TableSettings_classname_1.b)('popover-actions'), children: (0, jsx_runtime_1.jsx)(uikit_1.Button, { view: "action", size: "m", onClick: applyNewSettings, width: "max", children: (0, i18n_1.default)('button_apply') }) }))] }), (0, jsx_runtime_1.jsx)(uikit_1.Button, { view: "flat-secondary", size: "m", ref: anchorRef, onClick: togglePopup, children: (0, jsx_runtime_1.jsx)(uikit_1.Icon, { data: icons_1.Gear }) })] }));
};
exports.TableSettings = TableSettings;
