// @statoscope/stats-validator/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const json_ext_1 = require("@discoveryjs/json-ext");
const path_1 = require("@statoscope/config/dist/path");
const api_1 = require("./api");
class Validator {
    constructor(config, rootDir = process.cwd()) {
        this.plugins = {};
        this.applyConfig(config, rootDir);
    }
    applyConfig(config, rootDir) {
        var _a;
        this.rootDir = rootDir;
        this.config = config;
        this.require = (0, path_1.makeRequireFromPath)(rootDir);
        if (this.config.plugins) {
            for (const pluginDefinition of this.config.plugins) {
                const pluginAlias = Array.isArray(pluginDefinition)
                    ? pluginDefinition[0]
                    : pluginDefinition;
                const normalizedPluginPath = (0, path_1.resolveAliasPackage)(path_1.PackageAliasType.PLUGIN, pluginAlias, rootDir);
                const localPluginAlias = Array.isArray(pluginDefinition)
                    ? pluginDefinition[1]
                    : pluginDefinition;
                const resolvedPluginPath = this.require.resolve(normalizedPluginPath);
                const pluginNS = this.require(resolvedPluginPath);
                const plugin = ((_a = pluginNS.default) !== null && _a !== void 0 ? _a : pluginNS);
                const pluginDescription = plugin();
                this.plugins[resolvedPluginPath] = {
                    rules: {},
                    aliases: [localPluginAlias],
                    prepare: pluginDescription.prepare,
                };
                if (pluginDescription.rules) {
                    for (const [name, rule] of Object.entries(pluginDescription.rules)) {
                        this.plugins[resolvedPluginPath].rules[name] = rule;
                    }
                }
            }
        }
    }
    resolveRule(name) {
        const ruleRx = /^(?:(@.+?[/\\][^/\\\s]+|[^/\\\s]+)[/\\])?(.+)/;
        const [, pluginAlias, ruleName] = name.match(ruleRx) || [];
        for (const [, pluginDesc] of Object.entries(this.plugins)) {
            if (pluginDesc.aliases.includes(pluginAlias)) {
                if (Object.prototype.hasOwnProperty.call(pluginDesc.rules, ruleName)) {
                    return pluginDesc.rules[ruleName];
                }
            }
        }
        return null;
    }
    async validate(input, reference) {
        var _a, _b;
        const parsedInput = [];
        let preparedInput;
        parsedInput.push({
            name: 'input.json',
            data: await (0, json_ext_1.parseChunked)(fs_1.default.createReadStream(input)),
        });
        if (reference) {
            parsedInput.push({
                name: 'reference.json',
                data: await (0, json_ext_1.parseChunked)(fs_1.default.createReadStream(reference)),
            });
        }
        const result = {
            rules: [],
            files: {
                input: input,
                reference: reference,
            },
        };
        for (const [, plugin] of Object.entries(this.plugins)) {
            if (typeof plugin.prepare === 'function') {
                preparedInput = await plugin.prepare(parsedInput);
            }
        }
        preparedInput !== null && preparedInput !== void 0 ? preparedInput : (preparedInput = parsedInput);
        const rules = Object.entries((_b = (_a = this.config) === null || _a === void 0 ? void 0 : _a.rules) !== null && _b !== void 0 ? _b : {});
        if (!rules.length) {
            console.log('[WARN] No rules has specified in statoscope config');
        }
        for (const [ruleName, ruleDesc] of rules) {
            let ruleParams;
            let execParams;
            if (Array.isArray(ruleDesc)) {
                [execParams, ruleParams] = ruleDesc;
            }
            else {
                execParams = ruleDesc;
            }
            const normalizedExecParams = this.normalizeExecParams(execParams);
            if (normalizedExecParams.mode === 'off') {
                continue;
            }
            const resolvedRule = this.resolveRule(ruleName);
            if (!resolvedRule) {
                throw new Error(`Can't resolve rule ${ruleName}`);
            }
            const api = await this.execRule(resolvedRule, ruleParams, preparedInput);
            result.rules.push({ name: ruleName, api, execParams: normalizedExecParams });
        }
        return result;
    }
    async execRule(rule, params, data) {
        const api = (0, api_1.makeAPI)();
        await rule(params, data, api);
        return api;
    }
    normalizeExecParams(execParams) {
        let mode = typeof execParams === 'string' ? execParams : 'error';
        if (mode === 'warn' && this.config.warnAsError) {
            mode = 'error';
        }
        return {
            mode,
        };
    }
}
exports.default = Validator;
