// @gravity-ui/date-components/dist/cjs/components/RelativeRangeDatePicker/components/Control/Control.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Control = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = tslib_1.__importDefault(require("react"));
const icons_1 = require("@gravity-ui/icons");
const uikit_1 = require("@gravity-ui/uikit");
const cn_1 = require("../../../../utils/cn.js");
const getButtonSizeForInput_1 = require("../../../utils/getButtonSizeForInput.js");
const utils_1 = require("../../utils.js");
const i18n_1 = require("../Presets/i18n/index.js");
const i18n_2 = require("./i18n/index.js");
require("./Control.css");
const b = (0, cn_1.block)('relative-range-date-picker-control');
exports.Control = react_1.default.forwardRef(({ props, state, open, setOpen, isMobile, onClick, onKeyDown, onFocus, onClickCalendar, onUpdate, }, ref) => {
    var _a;
    const { alwaysShowAsAbsolute, presetTabs, getRangeTitle } = props;
    const format = props.format || 'L';
    const { t } = i18n_2.i18n.useTranslation();
    const { t: presetsTranslations } = i18n_1.i18n.useTranslation();
    const { lang } = (0, uikit_1.useLang)();
    const text = react_1.default.useMemo(() => typeof getRangeTitle === 'function'
        ? getRangeTitle(state.value, state.timeZone)
        : (0, utils_1.getDefaultTitle)({
            value: state.value,
            timeZone: state.timeZone,
            alwaysShowAsAbsolute,
            format,
            presets: presetTabs === null || presetTabs === void 0 ? void 0 : presetTabs.flatMap(({ presets }) => presets),
            presetsTranslations,
            lang,
        }), [
        alwaysShowAsAbsolute,
        format,
        getRangeTitle,
        lang,
        presetTabs,
        presetsTranslations,
        state.timeZone,
        state.value,
    ]);
    const validationState = props.validationState || (state.isInvalid ? 'invalid' : undefined);
    const errorMessage = (_a = props.errorMessage) !== null && _a !== void 0 ? _a : state.errors.join('\n');
    const triggerProps = {
        id: props.id,
        role: 'combobox',
        'aria-haspopup': 'dialog',
        'aria-expanded': open,
        'aria-label': props['aria-label'],
        'aria-labelledby': props['aria-labelledby'],
        'aria-describedby': props['aria-describedby'],
        'aria-details': props['aria-details'],
        disabled: props.disabled,
        readOnly: props.readOnly,
        onClick: onClickCalendar,
        onKeyDown,
    };
    return props.renderControl ? (props.renderControl({
        ref,
        value: state.value,
        title: text,
        validationState,
        errorMessage,
        open,
        setOpen,
        triggerProps,
    })) : ((0, jsx_runtime_1.jsxs)(react_1.default.Fragment, { children: [(0, jsx_runtime_1.jsx)(uikit_1.TextInput, { id: props.id, autoFocus: props.autoFocus, controlRef: ref, value: text, placeholder: props.placeholder, onUpdate: onUpdate, controlProps: Object.assign(Object.assign({ className: b('input', { mobile: isMobile }) }, triggerProps), { disabled: isMobile, onClick }), onKeyDown: onKeyDown, onFocus: onFocus, validationState: validationState, errorMessage: errorMessage, errorPlacement: props.errorPlacement, pin: props.pin, size: props.size, label: props.label, hasClear: props.hasClear, disabled: props.disabled, endContent: (0, jsx_runtime_1.jsx)(uikit_1.Button, { view: "flat-secondary", size: (0, getButtonSizeForInput_1.getButtonSizeForInput)(props.size), disabled: props.disabled, "aria-haspopup": "dialog", "aria-expanded": open, "aria-label": t('Range date picker'), onClick: onClickCalendar, tabIndex: -1, children: (0, jsx_runtime_1.jsx)(uikit_1.Icon, { data: icons_1.Calendar }) }) }), isMobile ? ((0, jsx_runtime_1.jsx)("button", { className: b('mobile-trigger', {
                    'has-clear': Boolean(props.hasClear && state.value),
                    'has-errors': state.isInvalid && props.errorPlacement === 'inside',
                    size: props.size,
                }), onClick: onClick })) : null] }));
});
exports.Control.displayName = 'Control';
