// @gravity-ui/date-components/dist/esm/components/RangeDateSelection/components/Ruler/Ruler.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import ReactDOM from 'react-dom';
import { useResizeObserver } from "../../../../hooks/useResizeObserver.js";
import { block } from "../../../../utils/cn.js";
import { NowLine } from "../NowLine/NowLine.js";
import { RulerViewport } from "../RulerViewport/RulerViewport.js";
import { MiddleTicks } from "../Ticks/MiddleTicks.js";
import { SlitTicks } from "../Ticks/SlitTicks.js";
import { UnavailableTicks } from "../Ticks/UnavialableTicks.js";
import { makeUnavailableTicksGeometry } from "../Ticks/utils.js";
const b = block('ruler');
const viewportDimensionsContext = React.createContext(null);
const viewportIntervalContext = React.createContext(null);
export function DateTimeRuler({ className, children, start, end, minValue, maxValue, onMove, onMoveStart, onMoveEnd, displayNow, formatTime, timeZone, dragDisabled, renderAdditionalRulerContent, }) {
    const [container, setContainer] = React.useState(null);
    const viewportInterval = React.useMemo(() => ({ start, end }), [start, end]);
    const [viewportDimensions, setViewportDimensions] = React.useState({
        width: 0,
        height: 0,
    });
    const onResize = React.useCallback(() => {
        if (!container) {
            return;
        }
        const { width, height } = container.getBoundingClientRect();
        setViewportDimensions({ width, height });
    }, [container]);
    const containerRef = React.useMemo(() => ({ current: container }), [container]);
    useResizeObserver({ ref: containerRef, onResize });
    const hasConstrains = Boolean(minValue || maxValue);
    return (_jsx("div", { className: b(null, className), children: _jsx(viewportIntervalContext.Provider, { value: viewportInterval, children: _jsxs(viewportDimensionsContext.Provider, { value: viewportDimensions, children: [_jsxs(RulerViewport, { ref: setContainer, onMove: onMove, onMoveStart: onMoveStart, onMoveEnd: onMoveEnd, dragDisabled: dragDisabled, children: [_jsx(MiddleTicks, { timeZone: timeZone }), _jsx(SlitTicks, { formatTime: formatTime, timeZone: timeZone }), hasConstrains ? (_jsx(UnavailableTicks, { theme: "dim", minValue: minValue, maxValue: maxValue, geometry: makeUnavailableTicksGeometry({
                                    tickHeight: 10,
                                    viewportHeight: viewportDimensions.height,
                                }), tickWidth: viewportDimensions.height + 10 })) : null, displayNow ? _jsx(NowLine, {}) : null, renderAdditionalRulerContent === null || renderAdditionalRulerContent === void 0 ? void 0 : renderAdditionalRulerContent({
                                interval: viewportInterval,
                                dimensions: viewportDimensions,
                            })] }), container ? ReactDOM.createPortal(children, container) : null] }) }) }));
}
export function useViewportDimensions() {
    const viewport = React.useContext(viewportDimensionsContext);
    if (!viewport) {
        throw new Error('useViewportDimensions must be used within a RulerViewport');
    }
    return viewport;
}
export function useViewportInterval() {
    const viewport = React.useContext(viewportIntervalContext);
    if (!viewport) {
        throw new Error('useViewportInterval must be used within a RulerViewport');
    }
    return viewport;
}
