// @statoscope/cli/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.combineCustomReports = exports.isCustomReport = exports.createDestStatReportPath = exports.transform = exports.mergeCustomExtensionsIntoCompilation = exports.mergeCustomReportsIntoCompilation = void 0;
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const fs_1 = __importDefault(require("fs"));
const utils_1 = require("@statoscope/report-writer/dist/utils");
const json_ext_1 = require("@discoveryjs/json-ext");
const normalizeCompilation_1 = __importDefault(require("@statoscope/webpack-model/dist/normalizeCompilation"));
const generator_1 = __importDefault(require("@statoscope/stats-extension-custom-reports/dist/generator"));
function mergeCustomReportsIntoCompilation(parsed, reports) {
    var _a, _b;
    var _c;
    (_a = parsed.__statoscope) !== null && _a !== void 0 ? _a : (parsed.__statoscope = {});
    (_b = (_c = parsed.__statoscope).extensions) !== null && _b !== void 0 ? _b : (_c.extensions = []);
    const customReportsExtensionIx = parsed.__statoscope.extensions.findIndex((ext) => ext.descriptor.name === '@statoscope/stats-extension-custom-reports');
    const customReportsExtension = customReportsExtensionIx > -1
        ? parsed.__statoscope.extensions[customReportsExtensionIx]
        : null;
    const customReportGenerator = new generator_1.default();
    if (customReportsExtension === null || customReportsExtension === void 0 ? void 0 : customReportsExtension.payload) {
        const payload = customReportsExtension.payload;
        for (const compilationItem of payload.compilations) {
            for (const report of compilationItem.reports) {
                customReportGenerator.handleReport(report);
            }
        }
    }
    for (const report of reports) {
        if (isCustomReport(report)) {
            customReportGenerator.handleReport(report);
        }
        else {
            throw new Error(`Can't add a report. A valid report should contain id and view fields`);
        }
    }
    if (customReportsExtension) {
        parsed.__statoscope.extensions.splice(customReportsExtensionIx, 1);
    }
    parsed.__statoscope.extensions.push(customReportGenerator.get());
    return parsed;
}
exports.mergeCustomReportsIntoCompilation = mergeCustomReportsIntoCompilation;
function mergeCustomExtensionsIntoCompilation(parsed, extensions) {
    var _a, _b;
    var _c;
    (_a = parsed.__statoscope) !== null && _a !== void 0 ? _a : (parsed.__statoscope = {});
    (_b = (_c = parsed.__statoscope).extensions) !== null && _b !== void 0 ? _b : (_c.extensions = []);
    for (const ext of extensions) {
        if (parsed.__statoscope.extensions.some((e) => e.descriptor.name === ext.descriptor.name)) {
            console.log(`Extension ${ext.descriptor.name} already exists. Ignoring`);
            continue;
        }
        parsed.__statoscope.extensions.push(ext);
    }
    return parsed;
}
exports.mergeCustomExtensionsIntoCompilation = mergeCustomExtensionsIntoCompilation;
async function transform(from, to, customReports = [], compression) {
    var _a;
    const normalizedFrom = [];
    for (const item of from) {
        const filename = typeof item === 'string' ? item : item.name;
        const as = typeof item === 'string' ? item : item.as;
        let parsed = await (0, json_ext_1.parseChunked)(fs_1.default.createReadStream(filename));
        if (customReports.length) {
            parsed = mergeCustomReportsIntoCompilation(parsed, customReports);
        }
        (0, normalizeCompilation_1.default)(parsed);
        normalizedFrom.push({
            type: 'data',
            filename: as,
            data: parsed,
            replacer: (0, utils_1.makeReplacer)((_a = parsed.__statoscope) === null || _a === void 0 ? void 0 : _a.context, '.', ['context', 'source']),
        });
    }
    return (0, utils_1.transform)({
        writer: {
            scripts: [{ type: 'path', path: require.resolve('@statoscope/webpack-ui') }],
            init: `function (data) {
          Statoscope.default(data.map((item) => ({ name: item.id, data: item.data })));
        }`,
            dataCompression: compression,
        },
    }, normalizedFrom, to);
}
exports.transform = transform;
const createDestStatReportPath = (from, to) => {
    const id = path_1.default.basename(from[0], '.json');
    return to || path_1.default.join(os_1.default.tmpdir(), `statoscope-report-${id}-${Date.now()}.html`);
};
exports.createDestStatReportPath = createDestStatReportPath;
/**
 * Return true for objects with properties expected from custom report.
 * Type guard.
 */
function isCustomReport(report) {
    return !!(report &&
        typeof report === 'object' &&
        report.constructor === Object &&
        'id' in report &&
        'view' in report);
}
exports.isCustomReport = isCustomReport;
/**
 * Combine custom user reports from statoscope config and command arguments
 * (report json file paths).
 */
function combineCustomReports(config, reportArg = []) {
    var _a;
    const reports = [];
    if (reportArg) {
        let malformedReport = false;
        const reportsFromFiles = reportArg.flatMap((filepath) => require(path_1.default.resolve(filepath)));
        for (const report of reportsFromFiles) {
            if (isCustomReport(report)) {
                reports.push(report);
            }
            else {
                malformedReport = true;
            }
        }
        if (malformedReport) {
            console.warn('One or more custom reports from files passed are malformed');
        }
    }
    if ((_a = config.generate) === null || _a === void 0 ? void 0 : _a.reports) {
        reports.push(...config.generate.reports);
    }
    return reports;
}
exports.combineCustomReports = combineCustomReports;
