// @gravity-ui/date-components/dist/esm/components/RangeDateSelection/components/Ticks/Ticks.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { dateTime } from '@gravity-ui/date-utils';
import { block } from "../../../../utils/cn.js";
import { getLerpCoeff } from "../../utils/span.js";
import { useViewportDimensions, useViewportInterval } from "../Ruler/Ruler.js";
import { calculateTickTimes, formatTick as defaultFormatTime, makeMiddleTicksGeometry, makeSlitTicksGeometry, } from "./utils.js";
import "./Ticks.css";
const b = block('timeline-ticks');
export { makeMiddleTicksGeometry, makeSlitTicksGeometry };
export function Ticks({ theme, minTickWidth, maxTickWidth, hasLabels, geometry, className, labelsClassName, formatTime, timeZone, }) {
    const viewport = useViewportDimensions();
    const interval = useViewportInterval();
    const tickTimes = React.useMemo(() => calculateTickTimes({
        minTickWidth,
        maxTickWidth,
        viewportStart: interval.start,
        viewportEnd: interval.end,
        viewportWidth: viewport.width,
        timeZone,
    }), [maxTickWidth, minTickWidth, interval, viewport.width, timeZone]);
    const tickCoords = React.useMemo(() => {
        const timeToXCoeff = getLerpCoeff({ start: interval.start.valueOf(), end: interval.end.valueOf() }, { start: 0, end: viewport.width });
        return tickTimes.map((t) => 0 + Math.round((t.valueOf() - interval.start.valueOf()) * timeToXCoeff));
    }, [tickTimes, interval, viewport.width]);
    const ticksGeometry = React.useMemo(() => {
        return tickCoords.map(geometry).join('');
    }, [geometry, tickCoords]);
    const ticksPath = _jsx("path", { className: b({ theme }, className), d: ticksGeometry });
    const tickLabels = React.useMemo(() => {
        if (!hasLabels || tickCoords.length < 2) {
            return null;
        }
        const result = [];
        for (let i = 0; i < tickCoords.length; i++) {
            const x = tickCoords[i];
            const t = tickTimes[i];
            const dt = dateTime({ input: t, timeZone });
            result.push(_jsx("text", { className: b('label', labelsClassName), dominantBaseline: "middle", x: x, y: "50%", filter: "url(#g-date-tt-ticklabelbg)", children: formatTime ? formatTime(dt) : defaultFormatTime(dt) }, i));
        }
        return result;
    }, [hasLabels, tickCoords, tickTimes, timeZone, labelsClassName, formatTime]);
    return (_jsxs(React.Fragment, { children: [ticksPath, tickLabels] }));
}
