// @gravity-ui/table/build/cjs/components/BaseRow/BaseRow.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const React = __importStar(require("react"));
const uikit_1 = require("@gravity-ui/uikit");
const BaseCell_1 = require("../BaseCell/index.js");
const BaseGroupHeader_1 = require("../BaseGroupHeader/index.js");
const BaseTable_classname_1 = require("../BaseTable/BaseTable.classname.js");
exports.BaseRow = React.forwardRef(({ cellClassName, className: classNameProp, getGroupTitle, getIsCustomRow, getIsGroupHeaderRow, groupHeaderClassName, onClick, renderCustomRowContent, renderGroupHeader, renderGroupHeaderRowContent, row, rowVirtualizer, style, virtualItem, attributes: attributesProp, cellAttributes, table: _, ...restProps }, ref) => {
    const rowRef = (0, uikit_1.useForkRef)(rowVirtualizer?.measureElement, ref);
    const attributes = typeof attributesProp === 'function' ? attributesProp(row) : attributesProp;
    const className = typeof classNameProp === 'function' ? classNameProp(row) : classNameProp;
    const handleClick = React.useCallback((event) => {
        onClick?.(row, event);
    }, [onClick, row]);
    const renderRowContent = () => {
        if (getIsGroupHeaderRow?.(row)) {
            return renderGroupHeaderRowContent ? (renderGroupHeaderRowContent({
                row,
                Cell: BaseCell_1.BaseCell,
                cellClassName,
                getGroupTitle,
            })) : ((0, jsx_runtime_1.jsx)(BaseCell_1.BaseCell, { className: cellClassName, colSpan: row.getVisibleCells().length, attributes: cellAttributes, "aria-colindex": 1, children: renderGroupHeader ? (renderGroupHeader({
                    row,
                    className: (0, BaseTable_classname_1.b)('group-header', groupHeaderClassName),
                    getGroupTitle,
                })) : ((0, jsx_runtime_1.jsx)(BaseGroupHeader_1.BaseGroupHeader, { row: row, className: (0, BaseTable_classname_1.b)('group-header', groupHeaderClassName), getGroupTitle: getGroupTitle })) }));
        }
        if (getIsCustomRow?.(row) && renderCustomRowContent) {
            return renderCustomRowContent({ row, Cell: BaseCell_1.BaseCell, cellClassName });
        }
        return row
            .getVisibleCells()
            .map((cell) => ((0, jsx_runtime_1.jsx)(BaseCell_1.BaseCell, { cell: cell, className: cellClassName, attributes: cellAttributes, "aria-colindex": cell.column.getIndex() + 1 }, cell.id)));
    };
    return ((0, jsx_runtime_1.jsx)("tr", { ref: rowRef, className: (0, BaseTable_classname_1.b)('row', {
            selected: row.getIsSelected(),
            interactive: Boolean(onClick),
        }, className), onClick: handleClick, "data-index": virtualItem?.index, ...restProps, ...attributes, style: {
            top: rowVirtualizer && virtualItem
                ? virtualItem.start - rowVirtualizer.options.scrollMargin
                : undefined,
            ...style,
            ...attributes?.style,
        }, children: renderRowContent() }));
});
exports.BaseRow.displayName = 'BaseRow';
