// monaco-yql-languages/build/_.contribution.js
import { getCompletionItemsProvider } from './completion';
import { Emitter, languages } from './fillers/monaco-editor-core';
const languageDefinitions = {};
const lazyLanguageLoaders = {};
class LazyLanguageLoader {
    static getOrCreate(languageId) {
        if (!lazyLanguageLoaders[languageId]) {
            lazyLanguageLoaders[languageId] = new LazyLanguageLoader(languageId);
        }
        return lazyLanguageLoaders[languageId];
    }
    constructor(languageId) {
        this._languageId = languageId;
        this._loadingTriggered = false;
        this._lazyLoadPromise = new Promise((resolve, reject) => {
            this._lazyLoadPromiseResolve = resolve;
            this._lazyLoadPromiseReject = reject;
        });
    }
    whenLoaded() {
        return this._lazyLoadPromise;
    }
    load() {
        if (!this._loadingTriggered) {
            this._loadingTriggered = true;
            languageDefinitions[this._languageId].loader().then((mod) => this._lazyLoadPromiseResolve(mod), (err) => this._lazyLoadPromiseReject(err));
        }
        return this._lazyLoadPromise;
    }
}
export function loadLanguage(languageId) {
    return LazyLanguageLoader.getOrCreate(languageId).load();
}
export function registerLanguage(def) {
    const languageId = def.id;
    languageDefinitions[languageId] = def;
    languages.register(def);
    const lazyLanguageLoader = LazyLanguageLoader.getOrCreate(languageId);
    languages.setMonarchTokensProvider(languageId, lazyLanguageLoader.whenLoaded().then((mod) => mod.language));
    languages.onLanguage(languageId, () => {
        lazyLanguageLoader.load().then((mod) => {
            languages.setLanguageConfiguration(languageId, mod.conf);
        });
    });
    lazyLanguageLoader.whenLoaded().then((mod) => {
        if (mod.completions) {
            registerCompletionItemProvider(languageId, mod.completions);
        }
    });
}
function registerCompletionItemProvider(languageId, completions) {
    const disposables = [];
    function unregister() {
        var _a;
        while (disposables.length > 0) {
            (_a = disposables.pop()) === null || _a === void 0 ? void 0 : _a.dispose();
        }
    }
    const defaults = languages[languageId];
    if (!defaults || defaults.modeConfiguration.completionItems) {
        disposables.push(languages.registerCompletionItemProvider(languageId, getCompletionItemsProvider(completions, languages.CompletionItemKind)));
    }
    if (defaults) {
        disposables.push(defaults.onDidChange(() => {
            unregister();
            registerCompletionItemProvider(languageId, completions);
        }));
    }
}
export class LanguageServiceDefaultsImpl {
    constructor(languageId, diagnosticsOptions, modeConfiguration) {
        this._onDidChange = new Emitter();
        this._languageId = languageId;
        this.setDiagnosticsOptions(diagnosticsOptions);
        this.setModeConfiguration(modeConfiguration);
    }
    get onDidChange() {
        return this._onDidChange.event;
    }
    get languageId() {
        return this._languageId;
    }
    get modeConfiguration() {
        return this._modeConfiguration;
    }
    get diagnosticsOptions() {
        return this._diagnosticsOptions;
    }
    setDiagnosticsOptions(options) {
        this._diagnosticsOptions = options || Object.create(null);
        this._onDidChange.fire(this);
    }
    setModeConfiguration(modeConfiguration) {
        this._modeConfiguration = modeConfiguration || Object.create(null);
        this._onDidChange.fire(this);
    }
}
export const modeConfigurationDefault = {
    completionItems: true,
    hovers: true,
    documentSymbols: true,
    definitions: true,
    references: true,
    documentHighlights: true,
    rename: true,
    colors: true,
    foldingRanges: true,
    diagnostics: true,
    selectionRanges: true,
};
export const diagnosticDefault = {
    validate: true,
};
