// @gravity-ui/date-components/dist/cjs/components/RangeDateSelection/hooks/useRangeDateSelectionState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRangeDateSelectionState = useRangeDateSelectionState;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const uikit_1 = require("@gravity-ui/uikit");
const constants_1 = require("../../utils/constants.js");
const dates_1 = require("../../utils/dates.js");
const useDefaultTimeZone_1 = require("../../utils/useDefaultTimeZone.js");
const date_1 = require("../utils/date.js");
function getViewportIntervalFromDates(start, end, numberOfIntervals = 4, placeInViewport = 0.5) {
    const diff = end.diff(start);
    return {
        start: start.subtract(diff * (numberOfIntervals - 1) * placeInViewport),
        end: end.add(diff * (numberOfIntervals - 1) * (1 - placeInViewport)),
    };
}
function useRangeDateSelectionState(props) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const inputTimeZone = (0, useDefaultTimeZone_1.useDefaultTimeZone)(((_a = props.value) === null || _a === void 0 ? void 0 : _a.start) || ((_b = props.defaultValue) === null || _b === void 0 ? void 0 : _b.start) || props.placeholderValue);
    const timeZone = props.timeZone || inputTimeZone;
    const placeholderValue = (0, dates_1.createPlaceholderValue)({
        placeholderValue: props.placeholderValue,
        timeZone,
    });
    const [isDragging, setIsDragging] = react_1.default.useState(false);
    const [currentValue, setCurrentValue] = react_1.default.useState(null);
    const [value, setValue] = (0, uikit_1.useControlledState)(props.value, (_c = props.defaultValue) !== null && _c !== void 0 ? _c : {
        start: placeholderValue.startOf('day').timeZone(inputTimeZone),
        end: placeholderValue.endOf('day').timeZone(inputTimeZone),
    }, props.onUpdate);
    const setInterval = (newValue) => {
        setCurrentValue(null);
        setValue({
            start: (0, date_1.alignDateTime)(newValue.start, props.align).timeZone(inputTimeZone),
            end: (0, date_1.alignDateTime)(newValue.end, props.align).timeZone(inputTimeZone),
        });
    };
    const displayValue = {
        start: value.start.timeZone(timeZone),
        end: value.end.timeZone(timeZone),
    };
    const interval = isDragging ? displayValue : (currentValue !== null && currentValue !== void 0 ? currentValue : displayValue);
    const viewportInterval = getViewportIntervalFromDates(interval.start, interval.end, props.numberOfIntervals, props.placeOnRuler);
    const minValue = ((_d = props.minValue) === null || _d === void 0 ? void 0 : _d.isAfter(viewportInterval.start))
        ? props.minValue
        : viewportInterval.start;
    const maxValue = ((_e = props.maxValue) === null || _e === void 0 ? void 0 : _e.isBefore(viewportInterval.end))
        ? props.maxValue
        : viewportInterval.end;
    const align = (_f = props.align) !== null && _f !== void 0 ? _f : constants_1.SECOND;
    const minDuration = Math.max((_g = props.minDuration) !== null && _g !== void 0 ? _g : constants_1.SECOND, align);
    const maxDuration = Math.max(Math.min((_h = props.maxDuration) !== null && _h !== void 0 ? _h : 15 * constants_1.YEAR, maxValue.diff(minValue)), minDuration);
    const canResize = minDuration < maxDuration;
    const draggingState = react_1.default.useRef(null);
    const setDraggingValue = (v) => {
        draggingState.current = v;
        setCurrentValue(v);
    };
    const startDragging = () => {
        setIsDragging(true);
        setDraggingValue(displayValue);
    };
    const endDragging = () => {
        setIsDragging(false);
        if (draggingState.current) {
            setInterval(draggingState.current);
        }
    };
    const moveInterval = (delta) => {
        if (!draggingState.current) {
            return;
        }
        const { start, end } = draggingState.current;
        let newStart = start.add(delta, 'millisecond');
        let newEnd = end.add(delta, 'millisecond');
        const diff = newEnd.diff(newStart);
        if (newStart.valueOf() < minValue.valueOf()) {
            newStart = minValue;
            newEnd = (0, dates_1.constrainValue)(newStart.add(diff, 'millisecond'), newStart, maxValue);
        }
        else if (newEnd.valueOf() > maxValue.valueOf()) {
            newEnd = maxValue;
            newStart = (0, dates_1.constrainValue)(newEnd.subtract(diff, 'millisecond'), minValue, newEnd);
        }
        setDraggingValue({ start: newStart, end: newEnd });
    };
    const moveStart = (delta, { visualMinDuration = minDuration } = {}) => {
        if (!draggingState.current) {
            return;
        }
        const { start, end } = draggingState.current;
        let newStart = start.add(delta, 'millisecond');
        const deltaMin = Math.max(visualMinDuration, minDuration);
        const deltaMax = Math.min(maxDuration, end.diff(minValue));
        newStart = (0, dates_1.constrainValue)(newStart, end.subtract(deltaMax, 'millisecond'), end.subtract(deltaMin, 'millisecond'));
        setDraggingValue({ start: newStart, end });
    };
    const moveEnd = (delta, { visualMinDuration = minDuration } = {}) => {
        if (!draggingState.current) {
            return;
        }
        const { start, end } = draggingState.current;
        let newEnd = end.add(delta, 'millisecond');
        const deltaMin = Math.max(visualMinDuration, minDuration);
        const deltaMax = Math.min(maxDuration, maxValue.diff(start));
        newEnd = (0, dates_1.constrainValue)(end.add(delta, 'millisecond'), start.add(deltaMin, 'millisecond'), start.add(deltaMax, 'millisecond'));
        setDraggingValue({ start, end: newEnd });
    };
    const zoomInterval = (scale, { visualMinDuration = minDuration, fixedPoint = 0.5, } = {}) => {
        if (!draggingState.current) {
            return;
        }
        const { start, end } = draggingState.current;
        const len = end.diff(start);
        let newDiff = len * scale;
        newDiff = align * Math.round(newDiff / align);
        newDiff = Math.min(Math.max(newDiff, Math.max(visualMinDuration, minDuration)), maxDuration);
        let newStart = start.subtract(fixedPoint * (newDiff - len));
        let newEnd = newStart.add(newDiff, 'millisecond');
        const diff = newEnd.diff(newStart);
        if (newStart.valueOf() < minValue.valueOf()) {
            newStart = minValue;
            newEnd = (0, dates_1.constrainValue)(newStart.add(diff, 'millisecond'), newStart, maxValue);
        }
        else if (newEnd.valueOf() > maxValue.valueOf()) {
            newEnd = maxValue;
            newStart = (0, dates_1.constrainValue)(newEnd.subtract(diff, 'millisecond'), minValue, newEnd);
        }
        setDraggingValue({ start: newStart, end: newEnd });
    };
    return {
        viewportInterval,
        value: currentValue !== null && currentValue !== void 0 ? currentValue : displayValue,
        setValue: setInterval,
        setDraggingValue,
        canResize,
        startDragging,
        move: moveInterval,
        scale: zoomInterval,
        moveStart,
        moveEnd,
        endDragging,
        timeZone,
        align,
        isDragging,
    };
}
