// @gravity-ui/table/build/cjs/hooks/useColumnsAutoSize/utils/renderHeaderContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderHeaderContent = renderHeaderContent;
const react_table_1 = require("@tanstack/react-table");
const constants_1 = require("../constants.js");
async function renderHeaderContent(tableInstance, column) {
    try {
        if (!column.header) {
            return '';
        }
        if (typeof column.header === 'function') {
            const id = column.id || ('accessorKey' in column && String(column.accessorKey)) || '';
            const headerContext = {
                column: {
                    ...column,
                    id,
                    getSize: () => column.size || constants_1.cellDefaultWidth,
                },
                header: {
                    id,
                    column: column,
                    isPlaceholder: false,
                    placeholderId: undefined,
                    getContext: () => headerContext,
                    depth: 0,
                    index: 0,
                    colSpan: 1,
                    rowSpan: 1,
                    subHeaders: [],
                    getSize: () => column.size || constants_1.cellDefaultWidth,
                },
                table: tableInstance || {
                    getSortedRowModel: () => ({ rows: [] }),
                    getState: () => ({ sorting: [] }),
                    getHeaderGroups: () => [],
                    getAllColumns: () => [],
                    options: { enableSorting: false },
                },
                sortDescriptor: null,
                sortDirection: null,
            };
            const renderedHeader = (0, react_table_1.flexRender)(column.header, headerContext);
            if (renderedHeader instanceof Promise) {
                return await renderedHeader;
            }
            return renderedHeader;
        }
        return column.header;
    }
    catch {
        return column.id || ('accessorKey' in column && String(column.accessorKey)) || 'Header';
    }
}
