// @gravity-ui/date-components/dist/cjs/components/RelativeRangeDatePicker/RelativeRangeDatePicker.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelativeRangeDatePicker = RelativeRangeDatePicker;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = tslib_1.__importDefault(require("react"));
const uikit_1 = require("@gravity-ui/uikit");
const cn_1 = require("../../utils/cn.js");
const HiddenInput_1 = require("../HiddenInput/HiddenInput.js");
const filterDOMProps_1 = require("../utils/filterDOMProps.js");
const Control_1 = require("./components/Control/Control.js");
const PickerDialog_1 = require("./components/PickerDialog/PickerDialog.js");
const useRelativeRangeDatePickerState_1 = require("./hooks/useRelativeRangeDatePickerState.js");
require("./RelativeRangeDatePicker.css");
const b = (0, cn_1.block)('relative-range-date-picker');
function RelativeRangeDatePicker(props) {
    const state = (0, useRelativeRangeDatePickerState_1.useRelativeRangeDatePickerState)(props);
    const isMobile = (0, uikit_1.useMobile)();
    const [anchor, setAnchor] = react_1.default.useState(null);
    const dialogClosing = react_1.default.useRef(false);
    const [open, setOpen] = (0, uikit_1.useControlledState)(undefined, false, props.onOpenChange);
    const { focusWithinProps } = (0, uikit_1.useFocusWithin)({
        isDisabled: props.disabled,
        onFocusWithin: props.onFocus,
        onBlurWithin: props.onBlur,
        onFocusWithinChange: (isFocusWithin) => {
            var _a;
            if (isFocusWithin &&
                !dialogClosing.current &&
                !props.renderControl &&
                ((_a = document.activeElement) === null || _a === void 0 ? void 0 : _a.tagName) !== 'BUTTON') {
                setOpen(true);
            }
        },
    });
    const DOMProps = (0, filterDOMProps_1.filterDOMProps)(props);
    delete DOMProps.id;
    return ((0, jsx_runtime_1.jsxs)("div", Object.assign({}, DOMProps, { ref: setAnchor }, focusWithinProps, { className: b(null, props.className), style: props.style, children: [(0, jsx_runtime_1.jsx)(Control_1.Control, { props: props, state: state, open: open, setOpen: setOpen, isMobile: isMobile, onClick: () => {
                    if (props.disabled) {
                        return;
                    }
                    setOpen(true);
                }, onKeyDown: (e) => {
                    if (props.disabled) {
                        return;
                    }
                    if (e.altKey && (e.key === 'ArrowDown' || e.key === 'ArrowUp')) {
                        e.preventDefault();
                        setOpen(true);
                    }
                }, onClickCalendar: () => {
                    setOpen(!open);
                }, onUpdate: (v) => {
                    if (!props.readOnly && !v) {
                        state.setValue(null, 'default');
                    }
                } }), (0, jsx_runtime_1.jsx)(HiddenInput_1.HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a, _b; return (_b = (_a = v === null || v === void 0 ? void 0 : v.start) === null || _a === void 0 ? void 0 : _a.type) !== null && _b !== void 0 ? _b : ''; }, disabled: props.disabled }), (0, jsx_runtime_1.jsx)(HiddenInput_1.HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a; return getNativeValue((_a = v === null || v === void 0 ? void 0 : v.start) !== null && _a !== void 0 ? _a : null); }, disabled: props.disabled }), (0, jsx_runtime_1.jsx)(HiddenInput_1.HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a, _b; return (_b = (_a = v === null || v === void 0 ? void 0 : v.end) === null || _a === void 0 ? void 0 : _a.type) !== null && _b !== void 0 ? _b : ''; }, disabled: props.disabled }), (0, jsx_runtime_1.jsx)(HiddenInput_1.HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a; return getNativeValue((_a = v === null || v === void 0 ? void 0 : v.end) !== null && _a !== void 0 ? _a : null); }, disabled: props.disabled }), (0, jsx_runtime_1.jsx)(HiddenInput_1.HiddenInput, { name: props.name, form: props.form, onReset: (v) => {
                    state.setValue(v.value, v.timeZone);
                }, value: { value: state.value, timeZone: state.timeZone }, toStringValue: (v) => v.timeZone, disabled: props.disabled }), (0, jsx_runtime_1.jsx)(PickerDialog_1.PickerDialog, { value: state.value, onUpdate: state.setValue, timeZone: state.timeZone, open: open, onClose: () => {
                    setOpen(false);
                    dialogClosing.current = true;
                }, onRemove: () => {
                    setTimeout(() => {
                        dialogClosing.current = false;
                    });
                }, anchor: anchor, modal: true, returnFocus: true, popupClassName: props.popupClassName, popupStyle: props.popupStyle, popupPlacement: props.popupPlacement, popupOffset: props.popupOffset, isMobile: isMobile, size: props.size, format: props.format, readOnly: props.readOnly, minValue: props.minValue, maxValue: props.maxValue, allowNullableValues: props.allowNullableValues, isDateUnavailable: props.isDateUnavailable, placeholderValue: props.placeholderValue, withPresets: props.withPresets, applyPresetsImmediately: props.applyPresetsImmediately, presetTabs: props.presetTabs, docs: props.docs, withApplyButton: props.withApplyButton, withZonesList: props.withZonesList, withHeader: props.withHeader })] })));
}
function getNativeValue(value) {
    if (!value) {
        return '';
    }
    if (value.type === 'relative') {
        return value.value;
    }
    return value.value.toISOString();
}
