// @statoscope/config/dist/path.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveAliasPackage = exports.makeRequireFromPath = exports.normalizePath = exports.PackageAliasType = void 0;
const path_1 = __importDefault(require("path"));
const module_1 = __importDefault(require("module"));
const chalk_1 = __importDefault(require("chalk"));
var PackageAliasType;
(function (PackageAliasType) {
    PackageAliasType["PLUGIN"] = "PLUGIN";
    PackageAliasType["REPORTER"] = "REPORTER";
})(PackageAliasType || (exports.PackageAliasType = PackageAliasType = {}));
function normalizePath(source, rootDir) {
    return source.replace('<rootDir>', rootDir);
}
exports.normalizePath = normalizePath;
function makeRequireFromPath(pathname) {
    return module_1.default.createRequire(path_1.default.resolve(pathname, '_'));
}
exports.makeRequireFromPath = makeRequireFromPath;
const packageAliasPrefixes = new Map();
var PrefixType;
(function (PrefixType) {
    PrefixType["STATOSCOPE_NS"] = "@statoscope";
    PrefixType["NON_STATOSCOPE_NS"] = "@NON_STATOSCOPE_NS";
})(PrefixType || (PrefixType = {}));
packageAliasPrefixes.set(PackageAliasType.PLUGIN, {
    [PrefixType.STATOSCOPE_NS]: 'stats-validator-plugin',
    [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-plugin',
});
packageAliasPrefixes.set(PackageAliasType.REPORTER, {
    [PrefixType.STATOSCOPE_NS]: 'stats-validator-reporter',
    [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-reporter',
});
/**
 * Concatenate namespace (when provided) with the package name.
 */
function getFullPackageName([namespace, name]) {
    if (namespace) {
        return `${namespace}/${name}`;
    }
    return name;
}
/**
 * Return error message based on the list of package aliases (provided by the user and alternative).
 */
function getAliasPackageResolutionError(packageAliases) {
    const providedAlias = getFullPackageName(packageAliases[0]);
    let errorMessage = `Can't resolve package ${chalk_1.default.yellow.italic(providedAlias)}.\n`;
    const { italic: italicChalk } = chalk_1.default;
    if (packageAliases.length > 1) {
        const alternativeAlias = getFullPackageName(packageAliases[1]);
        errorMessage += `Also tried to resolve it with ${italicChalk.yellow(alternativeAlias)} alias which didn't work either.\n\n`;
    }
    else {
        errorMessage += '\n\n';
    }
    errorMessage += 'Try installing the package locally:\n';
    errorMessage += `- with ${italicChalk('npm')}: ${chalk_1.default.bgGrey('npm i -D ' + providedAlias)} (or corresponding package alias)\n`;
    errorMessage += `- with ${italicChalk('yarn')}: ${chalk_1.default.bgGrey('yarn add -D ' + providedAlias)} (or corresponding package alias)\n`;
    return errorMessage;
}
/**
 * Resolve full or short package name to it's absolute path.
 * @param packageAliasType - Plugin or reporter enum.
 * @param aliasName - Package name (alias). Can be passed in the "full" or "short" form.
 *   Will append prefixes based on packageAliasType value in order to resolve the short form alias.
 * @param fromDir - Directory used as "root" while resolving the package.
 */
function resolveAliasPackage(packageAliasType, aliasName, fromDir) {
    const localRequire = makeRequireFromPath(fromDir);
    aliasName = normalizePath(aliasName, fromDir);
    if (aliasName.startsWith('.') || path_1.default.isAbsolute(aliasName)) {
        localRequire(aliasName);
        return aliasName;
    }
    let packageNamespace = '';
    let packageName = aliasName;
    const packageNameRegex = /^(@.+?)\/(.+)/;
    if (packageNameRegex.test(aliasName)) {
        [, packageNamespace = '', packageName] = aliasName.match(packageNameRegex);
    }
    // can have from one up to two entries
    const packageAliases = [
        [packageNamespace, packageName], // original form
    ];
    const prefixes = packageAliasPrefixes.get(packageAliasType);
    const prefix = packageNamespace === PrefixType.STATOSCOPE_NS
        ? prefixes[PrefixType.STATOSCOPE_NS]
        : prefixes[PrefixType.NON_STATOSCOPE_NS];
    if (!packageName.startsWith(prefix)) {
        packageAliases.push([packageNamespace, `${prefix}-${packageName}`]);
    }
    const paths = packageAliases.map(([namespace, packageName]) => path_1.default.join(namespace, packageName));
    for (const path of paths) {
        try {
            localRequire(path);
            return path;
            // eslint-disable-next-line no-empty
        }
        catch (e) { }
    }
    throw new Error(getAliasPackageResolutionError(packageAliases));
}
exports.resolveAliasPackage = resolveAliasPackage;
