// @statoscope/stats-validator-reporter-console/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
class ConsoleReporter {
    constructor(options) {
        this.options = options;
    }
    async run(result) {
        var _a, _b, _c, _d, _e, _f, _g;
        var _h, _j, _k;
        const normalizedOptions = {
            useColors: (_b = (_a = this.options) === null || _a === void 0 ? void 0 : _a.useColors) !== null && _b !== void 0 ? _b : true,
        };
        const chalkCtx = new chalk_1.default.Instance(normalizedOptions.useColors ? {} : { level: 0 });
        // file -> compilation -> rule -> entry
        const groupedStorage = {};
        for (const rule of result.rules) {
            for (const item of rule.api.getStorage()) {
                const filename = (_c = item.filename) !== null && _c !== void 0 ? _c : 'unknown file';
                const compilation = (_d = item.compilation) !== null && _d !== void 0 ? _d : 'unknown compilation';
                (_e = groupedStorage[filename]) !== null && _e !== void 0 ? _e : (groupedStorage[filename] = {});
                (_f = (_h = groupedStorage[filename])[compilation]) !== null && _f !== void 0 ? _f : (_h[compilation] = {});
                (_g = (_j = groupedStorage[filename][compilation])[_k = rule.name]) !== null && _g !== void 0 ? _g : (_j[_k] = []);
                groupedStorage[filename][compilation][rule.name].push({
                    item,
                    execParams: rule.execParams,
                });
            }
        }
        let warn = 0;
        let error = 0;
        for (const [filename, compilations] of Object.entries(groupedStorage)) {
            console.log(chalkCtx.underline(filename));
            console.group();
            for (const [compilation, rules] of Object.entries(compilations)) {
                console.log(`${chalkCtx.cyan(compilation)}`);
                console.group();
                for (const [rule, items] of Object.entries(rules)) {
                    for (const result of items) {
                        const type = result.execParams.mode;
                        let decorate = chalkCtx.reset;
                        if (type === 'error') {
                            decorate = chalkCtx.red;
                        }
                        else {
                            decorate = chalkCtx.yellow;
                        }
                        console.log(`${decorate(type)}  ${result.item.message}  ${chalkCtx.cyan(rule)}`);
                        if (result.item.details) {
                            let detailDescriptor;
                            if (typeof result.item.details === 'string') {
                                detailDescriptor = {
                                    type: 'tty',
                                    content: result.item.details,
                                };
                            }
                            else {
                                // @ts-ignore
                                detailDescriptor =
                                    result.item.details.find((detail) => detail.type === 'tty') ||
                                        result.item.details.find((detail) => detail.type === 'text');
                            }
                            if (detailDescriptor) {
                                let content;
                                if (typeof detailDescriptor.content === 'string') {
                                    content = detailDescriptor.content;
                                }
                                else if (typeof detailDescriptor.content === 'function') {
                                    content = detailDescriptor.content.call(null);
                                }
                                else {
                                    content = detailDescriptor.content;
                                }
                                console.group();
                                if (typeof content === 'string') {
                                    console.log(content);
                                }
                                else {
                                    for (const line of content) {
                                        console.log(line);
                                    }
                                }
                                console.groupEnd();
                            }
                        }
                        if (type === 'error') {
                            error++;
                        }
                        else {
                            warn++;
                        }
                    }
                }
                console.groupEnd();
            }
            console.groupEnd();
            console.log('');
        }
        if (error || warn) {
            console.log(chalkCtx.yellow(`${error + warn} problems (${error} errors, ${warn} warnings)`));
        }
    }
}
exports.default = ConsoleReporter;
