// @gravity-ui/table/build/esm/components/TableSettings/TableSettings.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { DndContext, MouseSensor, TouchSensor, useSensor, useSensors } from '@dnd-kit/core';
import { SortableContext, verticalListSortingStrategy } from '@dnd-kit/sortable';
import { ChevronsDown, ChevronsRight, Gear } from '@gravity-ui/icons';
import { Button, Divider, Flex, Icon, Popup, Text, TextInput } from '@gravity-ui/uikit';
import { createColumn } from '@tanstack/react-table';
import debounce from 'lodash/debounce.js';
import { TableSettingsColumn } from "../TableSettingsColumn/TableSettingsColumn.js";
import { b } from "./TableSettings.classname.js";
import { getInitialOrderItems, getNestedColumnsCount, isDisplayedColumn, orderStateToColumnOrder, useOrderedItems, } from "./TableSettings.utils.js";
import { useFilterTableSettings } from "./hooks/index.js";
import i18n from "./i18n/index.js";
import "./TableSettings.css";
const POPUP_PLACEMENT = ['bottom-end', 'bottom', 'top-end', 'top', 'left', 'right'];
const NESTED_COLUMNS_PREFIX = 'nested_columns_count_';
export const TableSettings = ({ table, sortable = true, filterable = true, enableSearch = false, searchPlaceholder = '', onSettingsApply, }) => {
    const anchorRef = React.useRef(null);
    const [open, setOpen] = React.useState(false);
    const [search, setSearch] = React.useState('');
    const [expandNestedColumns, setExpandNestedColumns] = React.useState(false);
    const allColumns = React.useMemo(table.getAllColumns, [table]);
    const filteredColumns = React.useMemo(() => allColumns.filter(isDisplayedColumn), [allColumns]);
    const headers = table.getFlatHeaders();
    const headersById = React.useMemo(() => {
        return headers.reduce((acc, header) => {
            const result = { ...acc };
            result[header.column.id] = header;
            return result;
        }, {});
    }, [headers]);
    const { hiddenNodes, partiallyHiddenNodes } = useFilterTableSettings({
        filteredColumns,
        search,
        expandNestedColumns,
        headersById,
    });
    const [visibilityState, setVisibilityState] = React.useState(() => table.getState().columnVisibility);
    const [orderState, setOrderState] = React.useState(() => getInitialOrderItems(filteredColumns, table.getState().columnOrder));
    const sensors = useSensors(useSensor(MouseSensor), useSensor(TouchSensor));
    const { orderedItems, activeDepth, handleDragEnd, handleDragStart, handleDragCancel } = useOrderedItems(filteredColumns, orderState, setOrderState);
    const orderedItemIds = React.useMemo(() => orderedItems.map(({ id }) => id), [orderedItems]);
    const renderColumns = (renderedColumns, innerColumn) => {
        const columns = [...renderedColumns];
        const isSortableColumn = sortable && !search.length;
        const level = innerColumn?.depth ?? 0;
        const displayedColumns = columns.filter((column) => !hiddenNodes.has(column.id));
        if (innerColumn && !expandNestedColumns && partiallyHiddenNodes.has(innerColumn.id)) {
            let count = getNestedColumnsCount(innerColumn);
            if (displayedColumns.length) {
                count -= displayedColumns.length;
            }
            displayedColumns.push(createColumn(table, {
                id: `${NESTED_COLUMNS_PREFIX}${innerColumn.id}`,
                header: () => (_jsx(Text, { variant: "body-1", color: "secondary", children: i18n('nested_columns_count', {
                        count,
                    }) })),
                columns: [],
                enableHiding: false,
            }, level + 1, innerColumn));
        }
        return displayedColumns.map((innerColumn) => {
            const children = innerColumn.columns.length
                ? renderColumns(innerColumn.columns.filter(isDisplayedColumn), innerColumn)
                : null;
            const header = headersById[innerColumn.id] ?? innerColumn.columnDef.header;
            const isFilterableColumn = innerColumn.id.startsWith(NESTED_COLUMNS_PREFIX)
                ? false
                : filterable;
            return (_jsx(TableSettingsColumn, { column: innerColumn, header: header, visibilityState: visibilityState, sortable: isSortableColumn, filterable: isFilterableColumn, activeDepth: activeDepth, withOffset: true, onVisibilityToggle: setVisibilityState, children: children }, innerColumn.id));
        });
    };
    const resetSettings = () => {
        setVisibilityState(table.getState().columnVisibility);
        setOrderState(getInitialOrderItems(filteredColumns, table.getState().columnOrder));
        setSearch('');
    };
    const cancelEditing = () => {
        setOpen(false);
        resetSettings();
    };
    const togglePopup = () => {
        if (open)
            cancelEditing();
        else
            setOpen(true);
    };
    const applyNewSettings = () => {
        const columnOrder = orderStateToColumnOrder(orderState);
        if (onSettingsApply)
            onSettingsApply({ visibilityState, columnOrder });
        if (filterable)
            table.setColumnVisibility(visibilityState);
        if (sortable)
            table.setColumnOrder(columnOrder);
        setOpen(false);
        setSearch('');
    };
    const updateSearch = debounce((value) => {
        setSearch(value);
    }, 200);
    const toggleNestedColumns = () => setExpandNestedColumns(!expandNestedColumns);
    const emptyResult = search.length > 0 && !orderedItems.some((item) => !hiddenNodes.has(item.id));
    return (_jsxs(React.Fragment, { children: [_jsxs(Popup, { open: open, onClose: cancelEditing, anchorRef: anchorRef, placement: POPUP_PLACEMENT, className: b(), 
                // @ts-ignore
                contentClassName: b(), children: [_jsxs("div", { className: b('popover-content'), children: [enableSearch && (_jsx(TextInput, { placeholder: searchPlaceholder, className: b('search-input'), endContent: partiallyHiddenNodes.size > 0 && (_jsx(Button, { view: "flat-secondary", onClick: toggleNestedColumns, children: expandNestedColumns ? (_jsxs(Flex, { alignItems: "center", children: [_jsx(Icon, { data: ChevronsDown }), i18n('collapse_button')] })) : (_jsxs(Flex, { alignItems: "center", children: [_jsx(Icon, { data: ChevronsRight }), i18n('expand_button')] })) })), onUpdate: updateSearch })), emptyResult ? (_jsx(Text, { variant: "body-1", color: "secondary", className: b('empty-message'), children: i18n('not_found') })) : (_jsx(DndContext, { onDragEnd: handleDragEnd, onDragStart: handleDragStart, onDragCancel: handleDragCancel, sensors: sensors, children: _jsx(SortableContext, { items: orderedItemIds, strategy: verticalListSortingStrategy, children: renderColumns(orderedItems) }) }))] }), _jsx(Divider, {}), !emptyResult && (_jsx("div", { className: b('popover-actions'), children: _jsx(Button, { view: "action", size: "m", onClick: applyNewSettings, width: "max", children: i18n('button_apply') }) }))] }), _jsx(Button, { view: "flat-secondary", size: "m", ref: anchorRef, onClick: togglePopup, children: _jsx(Icon, { data: Gear }) })] }));
};
