// @statoscope/cli/dist/commands/create.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const templates_1 = require("./templates");
const SUPPORTED_ENTITIES = ['plugin', 'rule', 'reporter'];
const SUPPORTED_EXT = ["js" /* FileExt.js */, "ts" /* FileExt.ts */];
function default_1(yargs) {
    return yargs.command('create', `Generate custom validator plugin/rule/reporter
Examples:
create rule     # create example rule from the template

create plugin   # create example plugin from the template

create reporter # create example-reporter from the template`, (yargs) => {
        return yargs
            .option('entity', {
            describe: 'Entity to generate',
            alias: 'e',
            choices: ['plugin', 'rule', 'reporter'],
        })
            .option('output', {
            describe: 'Path to generated code',
            alias: 'o',
            type: 'string',
            default: './',
        })
            .option('type', {
            describe: `Output type`,
            choices: ["js" /* FileExt.js */, "ts" /* FileExt.ts */],
            alias: 't',
            default: "js" /* FileExt.js */,
        })
            .option('module', {
            describe: `Output modules type`,
            choices: ["commonjs" /* ModuleType.commonjs */, "esm" /* ModuleType.esm */],
            alias: 'm',
            default: "commonjs" /* ModuleType.commonjs */,
        })
            .demandOption('entity');
    }, async (argv) => {
        if (!SUPPORTED_ENTITIES.includes(argv.entity)) {
            throw new Error(`${argv.entity}: generation is not supported. You can generate: plugin, rule, reporter. `);
        }
        if (!SUPPORTED_EXT.includes(argv.type)) {
            console.log(`${argv.type} is not supported. File will be generated with type js.`);
            argv.type = "js" /* FileExt.js */;
        }
        if (argv.type === "ts" /* FileExt.ts */ && argv.module === "commonjs" /* ModuleType.commonjs */) {
            console.log(`The module type will be changed to esm. Only the esm module is used to extend the ts file.`);
            argv.module = "esm" /* ModuleType.esm */;
        }
        const file = path_1.default.resolve(argv.output, `${argv.entity}.statoscope.${argv.type}`);
        try {
            fs_1.default.writeFile(path_1.default.resolve(argv.output, file), (0, templates_1.getTemplate)(argv.entity, {
                output: {
                    fileExt: argv.type,
                    module: argv.module,
                },
            }), {
                encoding: 'utf-8',
            }, (err) => {
                if (err) {
                    throw err;
                }
            });
            console.log(`Template created: ${file}`);
        }
        catch (err) {
            console.error(err);
        }
    });
}
exports.default = default_1;
