// @statoscope/cli/dist/commands/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const templates_1 = require("./templates");
const CONFIG_FILE_NAME = 'statoscope.config.js';
function default_1(yargs) {
    return yargs.command('init', `Generate statoscope config file`, (yargs) => {
        return yargs
            .option('output', {
            describe: `config file name`,
            type: 'string',
            alias: 'o',
            default: path_1.default.resolve(CONFIG_FILE_NAME),
        })
            .demandOption('output');
    }, async (argv) => {
        try {
            fs_1.default.writeFile(path_1.default.resolve(argv.output), (0, templates_1.getTemplate)('config', {
                output: {
                    module: "commonjs" /* ModuleType.commonjs */,
                    fileExt: "js" /* FileExt.js */,
                },
            }), {
                encoding: 'utf-8',
            }, (err) => {
                if (err) {
                    throw err;
                }
            });
        }
        catch (err) {
            console.error(err);
        }
    });
}
exports.default = default_1;
