// @statoscope/cli/dist/commands/legacyWebpackValidator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// todo remove in 6.0
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const json_ext_1 = require("@discoveryjs/json-ext");
const webpack_model_1 = require("@statoscope/webpack-model");
const helpers_1 = require("@statoscope/helpers");
function makeQueryValidator(query) {
    function validate(type, message) {
        if (!(!type || type === 'error' || type === 'warn' || type === 'info')) {
            throw new Error(`Unknown message type [${type}]`);
        }
        if (!message) {
            throw new Error(`Message must be specified`);
        }
    }
    function callAPI(api, { type, filename, message }) {
        validate(type, message);
        if (typeof type === 'undefined' || type === 'error') {
            api.error(message, filename);
        }
        else if (type === 'warn') {
            api.warn(message, filename);
        }
        else if (type === 'info') {
            api.info(message, filename);
        }
        else {
            console.log('Unknown message type:', type);
            api.error(message, filename);
        }
    }
    return async (data, api) => {
        const result = data.query(query);
        for (const item of result) {
            let type = item.type;
            if (!item.assert) {
                if (type === undefined && !Object.prototype.hasOwnProperty.call(item, 'assert')) {
                    type = 'info';
                }
                callAPI(api, { type, message: item.message, filename: item.filename });
            }
        }
    };
}
function handleValidator(validator) {
    const validatorPath = path_1.default.resolve(validator);
    const ext = path_1.default.extname(validatorPath);
    let validatorFn;
    if (ext === '.js') {
        validatorFn = require(validatorPath);
    }
    else {
        validatorFn = fs_1.default.readFileSync(validatorPath, 'utf8');
    }
    if (typeof validatorFn === 'string') {
        validatorFn = makeQueryValidator(validatorFn);
    }
    return validatorFn;
}
async function default_1(argv) {
    const files = [];
    for (const file of argv.input) {
        const data = await (0, json_ext_1.parseChunked)(fs_1.default.createReadStream(file));
        files.push({
            name: path_1.default.basename(file),
            data,
        });
    }
    const prepared = (0, webpack_model_1.prepareWithJora)(files, { helpers: (0, helpers_1.jora)() });
    const validator = handleValidator(argv.validator);
    const storage = {};
    let hasErrors = false;
    let errors = 0;
    let warnings = 0;
    let infos = 0;
    const api = {
        warn(message, filename = 'unknown') {
            if (argv.warnAsError) {
                hasErrors = true;
            }
            storage[filename] = storage[filename] || [];
            storage[filename].push({ type: 'warn', filename, message });
            warnings++;
        },
        error(message, filename = 'unknown') {
            hasErrors = true;
            storage[filename] = storage[filename] || [];
            storage[filename].push({ type: 'error', filename, message });
            errors++;
        },
        info(message, filename = 'unknown') {
            storage[filename] = storage[filename] || [];
            storage[filename].push({ type: 'info', filename, message });
            infos++;
        },
    };
    await validator({
        files: prepared.files,
        compilations: prepared.compilations,
        query: prepared.query,
    }, api);
    for (const [filename, items] of Object.entries(storage)) {
        console.log(filename);
        for (const item of items) {
            console.log(`  ${item.type}: ${item.message}`);
        }
    }
    console.log('');
    if (errors) {
        console.log(`Errors: ${errors}`);
    }
    if (warnings) {
        console.log(`Warnings: ${warnings}`);
    }
    if (infos) {
        console.log(`Info: ${infos}`);
    }
    console.log('Done');
    if (hasErrors) {
        // eslint-disable-next-line no-process-exit
        process.exitCode = 1;
    }
}
exports.default = default_1;
