// @statoscope/cli/dist/commands/query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const webpack_model_1 = require("@statoscope/webpack-model");
const json_ext_1 = require("@discoveryjs/json-ext");
const utils_1 = require("@statoscope/report-writer/dist/utils");
function default_1(yargs) {
    return yargs.command('query', `Execute jora-query on stats file`, (yargs) => {
        return yargs
            .option('input', {
            describe: 'path to a stats.json',
            alias: 'i',
            type: 'string',
        })
            .option('query', {
            describe: 'jora query (could be passed as stdin)',
            alias: 'q',
            type: 'string',
        })
            .array('input')
            .demandOption('input');
    }, async (argv) => {
        const query = argv.query || fs_1.default.readFileSync(0, 'utf-8');
        const files = [];
        for (const input of argv.input) {
            files.push({
                name: path_1.default.basename(input),
                data: await (0, json_ext_1.parseChunked)(fs_1.default.createReadStream(input)),
            });
        }
        const prepared = (0, webpack_model_1.prepareWithJora)(files);
        const result = prepared.query(query);
        const outputStream = (0, json_ext_1.stringifyStream)(result);
        outputStream.pipe(process.stdout);
        await (0, utils_1.waitFinished)(outputStream);
    });
}
exports.default = default_1;
