// @statoscope/cli/dist/commands/validate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const stats_validator_1 = __importDefault(require("@statoscope/stats-validator"));
const stats_validator_reporter_console_1 = __importDefault(require("@statoscope/stats-validator-reporter-console"));
const make_reporter_instance_1 = require("@statoscope/config/dist/make-reporter-instance");
const config_1 = require("@statoscope/config");
const legacyWebpackValidator_1 = __importDefault(require("./legacyWebpackValidator"));
function default_1(yargs) {
    return yargs.command('validate [validator] [input]', `Validate or compare JSON stats
Examples:
Validate stats: validate --input path/to/stats.json
Compare stats: validate --input path/to/branch.json --reference path/to/master.json
Validate stats [DEPRECATED]: validate --validator path/to/validator.js --input path/to/stats.json
Validate multiple stats [DEPRECATED]: validate --validator path/to/validator.js --input path/to/stats.json --input path/to/another/stats.json`, (yargs) => {
        return (yargs
            // todo remove in 6.0
            .option('validator', {
            describe: '[DEPRECATED] path to validator script',
            alias: 'v',
            type: 'string',
        })
            .option('config', {
            describe: 'path to statoscope config',
            alias: 'c',
            type: 'string',
        })
            .option('input', {
            describe: 'path to a stats.json',
            alias: 'i',
            type: 'string',
        })
            .option('reference', {
            describe: 'path to a stats-file to compare with',
            alias: 'r',
            type: 'string',
        })
            .option('warn-as-error', {
            describe: 'treat warn as error',
            alias: 'w',
            type: 'boolean',
        })
            .demandOption(['input'])
            .array('input'));
    }, async (argv) => {
        var _a, _b;
        if (argv.validator) {
            await (0, legacyWebpackValidator_1.default)({
                validator: argv.validator,
                input: argv.input,
                warnAsError: argv['warn-as-error'],
            });
            return;
        }
        const { path: configPath, config } = (0, config_1.requireConfig)(argv.config);
        const rootPath = path_1.default.dirname(configPath);
        const validateConfig = (_a = config.validate) !== null && _a !== void 0 ? _a : { rules: {} };
        if (argv['warn-as-error']) {
            validateConfig.warnAsError = true;
        }
        const validator = new stats_validator_1.default(validateConfig, rootPath);
        const reporters = [];
        if ((config === null || config === void 0 ? void 0 : config.silent) !== true) {
            if ((_b = config === null || config === void 0 ? void 0 : config.validate) === null || _b === void 0 ? void 0 : _b.reporters) {
                for (const item of config.validate.reporters) {
                    reporters.push((0, make_reporter_instance_1.makeReporterInstance)(item, rootPath));
                }
            }
            else {
                reporters.push(new stats_validator_reporter_console_1.default());
            }
        }
        const result = await validator.validate(path_1.default.resolve(argv.input[0]), argv.reference ? path_1.default.resolve(argv.reference) : void 0);
        for (const rule of result.rules) {
            const storage = rule.api.getStorage();
            if (rule.execParams.mode === 'error' && storage.length) {
                process.exitCode = 1;
                break;
            }
        }
        for (const reporter of reporters) {
            await reporter.run(result);
        }
    });
}
exports.default = default_1;
