// @gravity-ui/date-components/dist/cjs/components/DatePicker/hooks/useDatePickerProps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useDatePickerProps = useDatePickerProps;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const date_utils_1 = require("@gravity-ui/date-utils");
const uikit_1 = require("@gravity-ui/uikit");
const DateField_1 = require("../../DateField/index.js");
const filterDOMProps_1 = require("../../utils/filterDOMProps.js");
const getButtonSizeForInput_1 = require("../../utils/getButtonSizeForInput.js");
const mergeProps_1 = require("../../utils/mergeProps.js");
const i18n_1 = require("../i18n/index.js");
const utils_1 = require("../utils/index.js");
function useDatePickerProps(state, _a) {
    var { onFocus, onBlur } = _a, props = tslib_1.__rest(_a, ["onFocus", "onBlur"]);
    const [isActive, setActive] = react_1.default.useState(false);
    const [focusedDate, setFocusedDate] = react_1.default.useState((0, date_utils_1.isDateTime)(state.dateFieldState.displayValue)
        ? state.dateFieldState.displayValue
        : state.dateFieldState.displayValue.start);
    const [prevDateValue, setPrevDateValue] = react_1.default.useState(state.dateFieldState.displayValue);
    if ((0, date_utils_1.isDateTime)(prevDateValue)) {
        if ((0, date_utils_1.isDateTime)(state.dateFieldState.displayValue) &&
            !state.dateFieldState.displayValue.isSame(prevDateValue, 'day')) {
            setPrevDateValue(state.dateFieldState.displayValue);
            setFocusedDate(state.dateFieldState.displayValue);
        }
    }
    else if (!(0, date_utils_1.isDateTime)(state.dateFieldState.displayValue)) {
        if (!state.dateFieldState.displayValue.start.isSame(prevDateValue.start, 'day')) {
            setPrevDateValue(state.dateFieldState.displayValue);
            setFocusedDate(state.dateFieldState.displayValue.start);
        }
        else if (!state.dateFieldState.displayValue.end.isSame(prevDateValue.end, 'day')) {
            setPrevDateValue(state.dateFieldState.displayValue);
            setFocusedDate(state.dateFieldState.displayValue.end);
        }
    }
    const { focusWithinProps } = (0, uikit_1.useFocusWithin)({
        onFocusWithin: onFocus,
        onBlurWithin: onBlur,
        onFocusWithinChange(isFocusWithin) {
            setActive(isFocusWithin);
            if (!isFocusWithin) {
                state.setOpen(false, 'FocusOut');
            }
        },
    });
    const { inputProps } = (0, DateField_1.useDateFieldProps)(state.dateFieldState, props);
    const inputRef = react_1.default.useRef(null);
    const handleRef = (0, uikit_1.useForkRef)(inputRef, inputProps.controlRef);
    const calendarRef = react_1.default.useRef(null);
    const calendarButtonRef = react_1.default.useRef(null);
    const groupRef = react_1.default.useRef(null);
    function focusInput() {
        setTimeout(() => {
            var _a;
            (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus({ preventScroll: true });
        });
    }
    const onlyTime = state.formatInfo.hasTime && !state.formatInfo.hasDate;
    const calendarModes = (0, utils_1.getCalendarModes)(state.formatInfo);
    const { t } = i18n_1.i18n.useTranslation();
    const DOMProps = (0, filterDOMProps_1.filterDOMProps)(props);
    delete DOMProps.id;
    return {
        groupProps: Object.assign(Object.assign(Object.assign(Object.assign({}, DOMProps), { ref: groupRef, tabIndex: -1, role: 'group' }), focusWithinProps), { style: props.style, 'aria-disabled': state.disabled || undefined, onKeyDown: (e) => {
                if (!onlyTime && e.altKey && (e.key === 'ArrowDown' || e.key === 'ArrowUp')) {
                    e.preventDefault();
                    e.stopPropagation();
                    state.setOpen(true, 'ShortcutKeyDown');
                }
            } }),
        fieldProps: (0, mergeProps_1.mergeProps)(inputProps, state.dateFieldState.isEmpty && !isActive && props.placeholder
            ? { value: '' }
            : undefined, {
            controlRef: handleRef,
            controlProps: { role: 'combobox', 'aria-expanded': state.isOpen },
        }),
        calendarButtonProps: {
            ref: calendarButtonRef,
            size: (0, getButtonSizeForInput_1.getButtonSizeForInput)(props.size),
            disabled: state.disabled,
            'aria-label': t('Calendar'),
            'aria-haspopup': 'dialog',
            'aria-expanded': state.isOpen,
            view: 'flat-secondary',
            onClick: () => {
                setActive(true);
                state.setOpen(!state.isOpen, 'TriggerButtonClick');
            },
        },
        popupProps: {
            open: state.isOpen,
            onOpenChange: (open, e, reason) => {
                var _a, _b;
                if (!open) {
                    if (reason === 'escape-key') {
                        state.setOpen(false, 'EscapeKeyDown');
                        focusInput();
                    }
                    else if (reason === 'outside-press') {
                        if (!(e === null || e === void 0 ? void 0 : e.target) || !((_a = calendarButtonRef.current) === null || _a === void 0 ? void 0 : _a.contains(e.target))) {
                            state.setOpen(false, 'ClickOutside');
                        }
                        if ((e === null || e === void 0 ? void 0 : e.target) && ((_b = groupRef.current) === null || _b === void 0 ? void 0 : _b.contains(e.target))) {
                            focusInput();
                        }
                    }
                }
            },
            onTransitionOutComplete: () => {
                setFocusedDate((0, date_utils_1.isDateTime)(state.dateFieldState.displayValue)
                    ? state.dateFieldState.displayValue
                    : state.dateFieldState.displayValue.start);
            },
            modal: !props.disableFocusTrap,
            returnFocus: !props.disableFocusTrap,
            disablePortal: props.disablePortal,
            placement: props.popupPlacement,
            offset: props.popupOffset,
            className: props.popupClassName,
            style: props.popupStyle,
        },
        calendarProps: {
            ref: calendarRef,
            autoFocus: true,
            size: props.size === 's' ? 'm' : props.size,
            disabled: props.disabled,
            readOnly: props.readOnly,
            onUpdate: (d) => {
                state.setDateValue(d);
                if (!state.formatInfo.hasTime) {
                    focusInput();
                }
            },
            value: state.dateFieldState.displayValue,
            minValue: props.minValue,
            maxValue: props.maxValue,
            isDateUnavailable: props.isDateUnavailable,
            timeZone: state.timeZone,
            focusedValue: focusedDate,
            onFocusUpdate: setFocusedDate,
            modes: calendarModes,
        },
        timeInputProps: {
            value: state.timeValue,
            placeholderValue: (0, utils_1.getDateTimeValue)(state.dateFieldState.displayValue),
            onUpdate: state.setTimeValue,
            format: state.timeFormat,
            readOnly: state.readOnly,
            disabled: state.disabled,
            timeZone: state.timeZone,
            hasClear: props.hasClear,
            size: props.size,
        },
    };
}
