// @gravity-ui/table/build/cjs/components/TableSettingsColumn/TableSettingsColumn.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableSettingsColumn = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const React = __importStar(require("react"));
const sortable_1 = require("@dnd-kit/sortable");
const icons_1 = require("@gravity-ui/icons");
const uikit_1 = require("@gravity-ui/uikit");
const TableSettingsColumn_classname_1 = require("./TableSettingsColumn.classname.js");
const TableSettingsColumn_utils_1 = require("./TableSettingsColumn.utils.js");
require("./TableSettingsColumn.css");
const TableSettingsColumn = ({ column, header, children, visibilityState, filterable, sortable, disabled = false, activeDepth, withOffset = false, onVisibilityToggle, }) => {
    const innerColumns = column.getLeafColumns();
    const isVisible = React.useMemo(() => innerColumns.some((innerColumn) => (0, TableSettingsColumn_utils_1.getIsVisible)(innerColumn, visibilityState)), [innerColumns, visibilityState]);
    const isIndeterminate = React.useMemo(() => isVisible &&
        innerColumns.some((innerColumn) => !(0, TableSettingsColumn_utils_1.getIsVisible)(innerColumn, visibilityState)), [isVisible, innerColumns, visibilityState]);
    const isParent = innerColumns.length > 1;
    const toggle = () => {
        if (isParent) {
            const newState = innerColumns.reduce((acc, innerColumn) => {
                const result = { ...acc };
                if ((0, TableSettingsColumn_utils_1.isEnabledHiding)(innerColumn))
                    result[innerColumn.id] = !(isVisible && !isIndeterminate);
                return result;
            }, {});
            onVisibilityToggle((prevState) => ({ ...prevState, ...newState }));
        }
        else {
            onVisibilityToggle((prevState) => ({ ...prevState, [column.id]: !isVisible }));
        }
    };
    const renderSpacers = () => {
        const spacers = [];
        for (let index = 0; index < column.depth; index++) {
            spacers.push((0, jsx_runtime_1.jsx)("div", { className: (0, TableSettingsColumn_classname_1.b)('vertical-spacer') }, index));
        }
        return spacers;
    };
    const isDisabledContext = typeof activeDepth === 'number' ? activeDepth !== column.depth : false;
    const isRoot = column.depth === 0;
    const { attributes, listeners, setNodeRef, transform, transition, isDragging } = (0, sortable_1.useSortable)({
        id: column.id,
        disabled: {
            draggable: isDisabledContext,
            droppable: isDisabledContext,
        },
    });
    const style = {
        transform: `translate3d(0px, ${transform?.y ?? 0}px, 0)`,
        transition,
    };
    return ((0, jsx_runtime_1.jsxs)("div", { style: style, ref: setNodeRef, className: (0, TableSettingsColumn_classname_1.b)({ dragging: isDragging, root: isRoot, sortable }), children: [(0, jsx_runtime_1.jsx)("div", { className: (0, TableSettingsColumn_classname_1.b)('background'), children: (0, jsx_runtime_1.jsxs)("div", { ...attributes, ...(sortable ? { ...listeners } : {}), className: (0, TableSettingsColumn_classname_1.b)('layout', { dragging: isDragging }), children: [(0, jsx_runtime_1.jsxs)("div", { className: (0, TableSettingsColumn_classname_1.b)('content', { disabled }), children: [filterable && (
                                // eslint-disable-next-line jsx-a11y/no-static-element-interactions
                                (0, jsx_runtime_1.jsx)("div", { onMouseDown: (e) => {
                                        e.stopPropagation();
                                    }, onTouchStart: (e) => {
                                        e.stopPropagation();
                                    }, children: (0, jsx_runtime_1.jsx)(uikit_1.Checkbox, { checked: isVisible, disabled: !(0, TableSettingsColumn_utils_1.isEnabledHiding)(column), onChange: toggle, indeterminate: isIndeterminate, className: (0, TableSettingsColumn_classname_1.b)('checkbox', { 'unset-coursor': isDragging }) }) })), !filterable && withOffset && (0, jsx_runtime_1.jsx)("div", { className: (0, TableSettingsColumn_classname_1.b)('checkbox-spacer') }), renderSpacers(), (0, jsx_runtime_1.jsx)(uikit_1.Text, { variant: "body-1", className: (0, TableSettingsColumn_classname_1.b)('name', { parent: isParent }), children: (0, TableSettingsColumn_utils_1.getColumnTitle)(column, header) }), sortable ? ((0, jsx_runtime_1.jsx)("span", { className: (0, TableSettingsColumn_classname_1.b)('drag-handle', { 'unset-coursor': isDragging }), ...listeners, children: (0, jsx_runtime_1.jsx)(uikit_1.Icon, { data: icons_1.Grip, size: 16 }) })) : null] }), (0, jsx_runtime_1.jsx)(sortable_1.SortableContext, { id: column.id, items: column.columns?.map(({ id }) => id), strategy: sortable_1.verticalListSortingStrategy, children: children })] }) }), isRoot && !isDragging ? ((0, jsx_runtime_1.jsx)("div", { className: (0, TableSettingsColumn_classname_1.b)('divider'), children: (0, jsx_runtime_1.jsx)(uikit_1.Divider, {}) })) : null] }));
};
exports.TableSettingsColumn = TableSettingsColumn;
