// @gravity-ui/date-components/dist/cjs/components/RelativeRangeDatePicker/components/PickerDialog/PickerDoc.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PickerDoc = PickerDoc;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = tslib_1.__importDefault(require("react"));
const icons_1 = require("@gravity-ui/icons");
const uikit_1 = require("@gravity-ui/uikit");
const cn_1 = require("../../../../utils/cn.js");
const getButtonSizeForInput_1 = require("../../../utils/getButtonSizeForInput.js");
const defaultPresets_1 = require("../Presets/defaultPresets.js");
const i18n_1 = require("../Presets/i18n/index.js");
require("./PickerDoc.css");
const b = (0, cn_1.block)('relative-range-date-picker-doc');
const data = [
    {
        title: (0, jsx_runtime_1.jsx)(defaultPresets_1.PresetTitle, { title: "Last 5 minutes" }),
        from: 'now - 5m',
        to: 'now',
    },
    {
        title: (0, jsx_runtime_1.jsx)(defaultPresets_1.PresetTitle, { title: "From start of day" }),
        from: 'now/d',
        to: 'now',
    },
    {
        title: (0, jsx_runtime_1.jsx)(defaultPresets_1.PresetTitle, { title: "This week" }),
        from: 'now/w',
        to: 'now/w',
    },
    {
        title: (0, jsx_runtime_1.jsx)(defaultPresets_1.PresetTitle, { title: "From start of week" }),
        from: 'now/w',
        to: 'now',
    },
    {
        title: (0, jsx_runtime_1.jsx)(defaultPresets_1.PresetTitle, { title: "Previous month" }),
        from: 'now - 1M/M',
        to: 'now - 1M/M',
    },
];
function DocContent({ size, docs, onStartUpdate, onEndUpdate }) {
    const isMobile = (0, uikit_1.useMobile)();
    const { t } = i18n_1.i18n.useTranslation();
    const columns = react_1.default.useMemo(() => [
        {
            id: 'title',
            name: () => {
                return t('Range');
            },
        },
        {
            id: 'from',
            name: () => {
                return t('From');
            },
            template: (item) => ((0, jsx_runtime_1.jsx)(uikit_1.Button, { size: isMobile ? 'l' : (0, getButtonSizeForInput_1.getButtonSizeForInput)(size), onClick: () => onStartUpdate(item.from), children: item.from })),
        },
        {
            id: 'to',
            name: () => {
                return t('To');
            },
            template: (item) => ((0, jsx_runtime_1.jsx)(uikit_1.Button, { size: isMobile ? 'l' : (0, getButtonSizeForInput_1.getButtonSizeForInput)(size), onClick: () => onEndUpdate(item.to), children: item.to })),
        },
    ], [isMobile, onEndUpdate, onStartUpdate, size, t]);
    return (0, jsx_runtime_1.jsx)(uikit_1.Table, { columns: columns, data: docs, className: b('table', { size }), wordWrap: true });
}
function DesktopDoc(_a) {
    var { className, size } = _a, props = tslib_1.__rest(_a, ["className", "size"]);
    return ((0, jsx_runtime_1.jsx)(uikit_1.Popover, { className: b(null), hasArrow: false, placement: ['right-start', 'left-start'], content: (0, jsx_runtime_1.jsx)(DocContent, Object.assign({ size: size }, props)), children: (0, jsx_runtime_1.jsx)(uikit_1.Button, { className: b('button', className), view: "flat-secondary", size: (0, getButtonSizeForInput_1.getButtonSizeForInput)(size), children: (0, jsx_runtime_1.jsx)(uikit_1.Icon, { data: icons_1.CircleQuestion }) }) }));
}
function MobileDoc(_a) {
    var { className } = _a, props = tslib_1.__rest(_a, ["className"]);
    const [open, setOpen] = react_1.default.useState(false);
    return ((0, jsx_runtime_1.jsxs)("div", { className: b(null, className), children: [(0, jsx_runtime_1.jsx)(uikit_1.Button, { className: b('button'), view: "flat-secondary", size: "l", onClick: () => {
                    setOpen(true);
                }, children: (0, jsx_runtime_1.jsx)(uikit_1.Icon, { data: icons_1.CircleQuestion }) }), (0, jsx_runtime_1.jsx)(uikit_1.Sheet, { visible: open, onClose: () => setOpen(false), children: (0, jsx_runtime_1.jsx)(DocContent, Object.assign({}, props)) })] }));
}
function PickerDoc(_a) {
    var { docs = data } = _a, props = tslib_1.__rest(_a, ["docs"]);
    const isMobile = (0, uikit_1.useMobile)();
    if (!Array.isArray(docs) || docs.length === 0) {
        return null;
    }
    if (isMobile) {
        return (0, jsx_runtime_1.jsx)(MobileDoc, Object.assign({}, props, { docs: docs }));
    }
    return (0, jsx_runtime_1.jsx)(DesktopDoc, Object.assign({}, props, { docs: docs }));
}
