// @gravity-ui/date-components/dist/cjs/components/RelativeRangeDatePicker/components/PickerDialog/useRelativeRangeDatePickerDialogState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRelativeRangeDatePickerDialogState = useRelativeRangeDatePickerDialogState;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const uikit_1 = require("@gravity-ui/uikit");
const i18n_1 = require("../../../utils/validation/i18n/index.js");
const relativeRangeDatePicker_1 = require("../../../utils/validation/relativeRangeDatePicker.js");
function useRelativeRangeDatePickerDialogState(props) {
    var _a, _b, _c, _d;
    const [userTimeZone, setUserTimeZone] = (0, uikit_1.useControlledState)(props.timeZone, (_a = props.defaultTimeZone) !== null && _a !== void 0 ? _a : 'default', props.onUpdateTimeZone);
    const [value, setValue] = (0, uikit_1.useControlledState)(props.value, (_b = props.defaultValue) !== null && _b !== void 0 ? _b : null, (v, tz) => {
        var _a;
        setUserTimeZone(tz);
        if (value !== v || (value && userTimeZone !== tz)) {
            (_a = props.onUpdate) === null || _a === void 0 ? void 0 : _a.call(props, v, tz);
        }
    });
    const { withApplyButton } = props;
    const [start, setStart] = react_1.default.useState((_c = value === null || value === void 0 ? void 0 : value.start) !== null && _c !== void 0 ? _c : null);
    const [end, setEnd] = react_1.default.useState((_d = value === null || value === void 0 ? void 0 : value.end) !== null && _d !== void 0 ? _d : null);
    const [innerTimeZone, setTimeZone] = react_1.default.useState(userTimeZone);
    const timeZone = withApplyButton ? innerTimeZone : userTimeZone;
    const allowNullableValues = props.allowNullableValues;
    function setStartValue(newValue) {
        if (props.readOnly) {
            return;
        }
        setStart(newValue);
        if (!withApplyButton) {
            setValue(getRangeValue(newValue, end, Object.assign(Object.assign({}, props), { timeZone, allowNullableValues })), timeZone);
        }
    }
    function setEndValue(newValue) {
        if (props.readOnly) {
            return;
        }
        setEnd(newValue);
        if (!withApplyButton) {
            setValue(getRangeValue(start, newValue, Object.assign(Object.assign({}, props), { timeZone, allowNullableValues })), timeZone);
        }
    }
    function setTimeZoneValue(newTimeZone) {
        if (props.readOnly) {
            return;
        }
        setTimeZone(newTimeZone);
        const newStart = start ? Object.assign({}, start) : null;
        if ((newStart === null || newStart === void 0 ? void 0 : newStart.type) === 'absolute') {
            newStart.value = newStart.value.timeZone(newTimeZone, true);
            setStart(newStart);
        }
        const newEnd = end ? Object.assign({}, end) : null;
        if ((newEnd === null || newEnd === void 0 ? void 0 : newEnd.type) === 'absolute') {
            newEnd.value = newEnd.value.timeZone(newTimeZone, true);
            setEnd(newEnd);
        }
        if (!withApplyButton) {
            setValue(getRangeValue(newStart, newEnd, Object.assign(Object.assign({}, props), { timeZone: newTimeZone, allowNullableValues })), newTimeZone);
        }
    }
    function setRange(newStart, newEnd, force) {
        if (props.readOnly) {
            return;
        }
        setStart(newStart);
        setEnd(newEnd);
        if (!withApplyButton || force) {
            setValue(getRangeValue(newStart, newEnd, Object.assign(Object.assign({}, props), { timeZone, allowNullableValues })), timeZone);
        }
    }
    function applyValue() {
        if (props.readOnly) {
            return;
        }
        setValue(getRangeValue(start, end, Object.assign(Object.assign({}, props), { timeZone, allowNullableValues })), timeZone);
    }
    const { t } = i18n_1.i18n.useTranslation();
    const validation = react_1.default.useMemo(() => (0, relativeRangeDatePicker_1.getRangeValidationResult)(start || end ? { start, end } : null, allowNullableValues, props.minValue, props.maxValue, props.isDateUnavailable, timeZone, t), [
        allowNullableValues,
        end,
        props.isDateUnavailable,
        props.maxValue,
        props.minValue,
        start,
        timeZone,
        t,
    ]);
    return {
        start,
        end,
        timeZone,
        setStart: setStartValue,
        setEnd: setEndValue,
        setRange,
        setTimeZone: setTimeZoneValue,
        applyValue,
        isInvalid: validation.isInvalid,
        startValidation: validation.startValidationResult,
        endValidation: validation.endValidationResult,
    };
}
function getRangeValue(start, end, options = {}) {
    var _a;
    if (!start && !end) {
        return null;
    }
    const { isInvalid } = (0, relativeRangeDatePicker_1.getRangeValidationResult)({ start, end }, options.allowNullableValues, options.minValue, options.maxValue, options.isDateUnavailable, (_a = options.timeZone) !== null && _a !== void 0 ? _a : 'default');
    if (isInvalid) {
        return null;
    }
    return { start, end };
}
