// @gravity-ui/date-components/dist/cjs/components/RelativeRangeDatePicker/components/Presets/defaultPresets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allPresets = exports.DEFAULT_OTHERS_PRESETS = exports.DEFAULT_TIME_PRESETS = exports.DEFAULT_DATE_PRESETS = void 0;
exports.PresetTitle = PresetTitle;
const jsx_runtime_1 = require("react/jsx-runtime");
const i18n_1 = require("./i18n/index.js");
function PresetTitle({ title, }) {
    const { t } = i18n_1.i18n.useTranslation();
    return t(title);
}
exports.DEFAULT_DATE_PRESETS = [
    {
        from: 'now-1d',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last day" }),
    },
    {
        from: 'now-3d',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 3 days" }),
    },
    {
        from: 'now-1w',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last week" }),
    },
    {
        from: 'now-1M',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last month" }),
    },
    {
        from: 'now-3M',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 3 months" }),
    },
    {
        from: 'now-6M',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 6 months" }),
    },
    {
        from: 'now-1y',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last year" }),
    },
    {
        from: 'now-3y',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 3 years" }),
    },
];
exports.DEFAULT_TIME_PRESETS = [
    {
        from: 'now-5m',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 5 minutes" }),
    },
    {
        from: 'now-15m',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 15 minutes" }),
    },
    {
        from: 'now-30m',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 30 minutes" }),
    },
    {
        from: 'now-1h',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last hour" }),
    },
    {
        from: 'now-3h',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 3 hours" }),
    },
    {
        from: 'now-6h',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 6 hours" }),
    },
    {
        from: 'now-12h',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Last 12 hours" }),
    },
];
exports.DEFAULT_OTHERS_PRESETS = [
    {
        from: 'now/d',
        to: 'now/d',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Today" }),
    },
    {
        from: 'now-1d/d',
        to: 'now-1d/d',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Yesterday" }),
    },
    {
        from: 'now-2d/d',
        to: 'now-2d/d',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "Day before yesterday" }),
    },
    {
        from: 'now/w',
        to: 'now/w',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "This week" }),
    },
    {
        from: 'now/M',
        to: 'now/M',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "This month" }),
    },
    {
        from: 'now/y',
        to: 'now/y',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "This year" }),
    },
    {
        from: 'now/d',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "From start of day" }),
    },
    {
        from: 'now/w',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "From start of week" }),
    },
    {
        from: 'now/M',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "From start of month" }),
    },
    {
        from: 'now/y',
        to: 'now',
        title: (0, jsx_runtime_1.jsx)(PresetTitle, { title: "From start of year" }),
    },
];
exports.allPresets = exports.DEFAULT_TIME_PRESETS.concat(exports.DEFAULT_DATE_PRESETS, exports.DEFAULT_OTHERS_PRESETS);
