// @gravity-ui/date-components/dist/cjs/components/RelativeRangeDatePicker/components/Presets/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPresetTitle = getPresetTitle;
exports.filterPresets = filterPresets;
exports.getDefaultPresetTabs = getDefaultPresetTabs;
exports.filterPresetTabs = filterPresetTabs;
const date_utils_1 = require("@gravity-ui/date-utils");
const defaultPresets_1 = require("./defaultPresets.js");
const i18n_1 = require("./i18n/index.js");
const lastRe = /^now-(\d+)([smhdwMy])$/;
const oneUnit = {
    s: 'Last second',
    m: 'Last minute',
    h: 'Last hour',
    d: 'Last day',
    w: 'Last week',
    M: 'Last month',
    y: 'Last year',
};
const countUnit = {
    s: 'Last {count} second',
    m: 'Last {count} minute',
    h: 'Last {count} hour',
    d: 'Last {count} day',
    w: 'Last {count} week',
    M: 'Last {count} month',
    y: 'Last {count} year',
};
function getPresetTitle(start, end, presets = defaultPresets_1.allPresets, t = i18n_1.i18n) {
    const startText = start.replace(/\s+/g, '');
    const endText = end.replace(/\s+/g, '');
    for (const preset of presets) {
        if (preset.from === startText && preset.to === endText) {
            return preset.title;
        }
    }
    if (end === 'now') {
        const match = lastRe.exec(startText);
        if (match) {
            const [, count, unit] = match;
            if (isDateUnit(unit)) {
                const template = Number(count) === 1 ? oneUnit[unit] : countUnit[unit];
                return t(template, { count });
            }
        }
    }
    return startText + ' — ' + endText;
}
function isDateUnit(value) {
    return ['s', 'm', 'h', 'd', 'w', 'M', 'y'].includes(value);
}
function filterPresets(presets, minValue) {
    return presets.filter((preset) => {
        const from = (0, date_utils_1.dateTimeParse)(preset.from);
        const to = (0, date_utils_1.dateTimeParse)(preset.to, { roundUp: true });
        if (!from || !to) {
            return false;
        }
        if (to.isBefore(from)) {
            return false;
        }
        if (minValue && from.isBefore(minValue)) {
            return false;
        }
        return true;
    });
}
function getDefaultPresetTabs({ withTime, minValue, t = i18n_1.i18n, }) {
    const tabs = [];
    const mainTab = {
        id: 'main',
        title: t('Main'),
        presets: [],
    };
    const mainPresets = defaultPresets_1.DEFAULT_DATE_PRESETS;
    if (withTime) {
        mainPresets.unshift(...defaultPresets_1.DEFAULT_TIME_PRESETS);
    }
    mainTab.presets = filterPresets(mainPresets, minValue);
    if (mainTab.presets.length > 0) {
        tabs.push(mainTab);
    }
    const otherTab = {
        id: 'other',
        title: t('Other'),
        presets: filterPresets(defaultPresets_1.DEFAULT_OTHERS_PRESETS, minValue),
    };
    if (otherTab.presets.length > 0) {
        tabs.push(otherTab);
    }
    return tabs;
}
function filterPresetTabs(tabs, { minValue } = {}) {
    return tabs.reduce((acc, tab) => {
        const presets = filterPresets(tab.presets, minValue);
        if (presets.length) {
            acc.push(Object.assign(Object.assign({}, tab), { presets }));
        }
        return acc;
    }, []);
}
