// @farfetched/core/dist/core.js
import { createStore as x, is as q, combine as O, createEffect as k, attach as T, sample as c, createEvent as $, createNode as Fe, step as dt, split as J, withRegion as pt, merge as ce, scopeBind as Ie, createApi as mt } from "effector";
const re = {
  isData: (e) => !0,
  getErrorMessages: () => []
};
function D({
  field: e
}) {
  let t;
  if (e === void 0)
    t = x((r) => null, {
      serialize: "ignore",
      name: "ff.$target/undefined",
      sid: "ff.$target/undefined"
    });
  else if (q.store(e))
    t = O(
      e,
      (n) => (a) => n ?? null
    );
  else if (e != null && e.source && (e != null && e.fn)) {
    const r = e;
    t = O(
      r.source,
      (n) => (a) => r.fn(a, n) ?? null
    );
  } else if (typeof e == "function") {
    const r = e;
    t = x(
      (n) => r(n) ?? null,
      {
        serialize: "ignore",
        name: "ff.$target/callbackField",
        sid: "ff.$target/$callbackField"
      }
    );
  } else {
    const r = e;
    t = x((n) => r ?? null, {
      serialize: "ignore",
      name: "ff.$target/valueField",
      sid: "ff.$target/$valueField"
    });
  }
  return t;
}
function ht(e) {
  let t;
  if (e === void 0)
    t = k(async (r) => null);
  else if (q.store(e))
    t = T({
      source: e,
      async effect(n, a) {
        return n;
      }
    });
  else if (e.source && e.fn) {
    const r = e;
    t = T({
      source: r.source,
      async effect(n, a) {
        return r.fn(a, n);
      }
    });
  } else if (typeof e == "function") {
    const r = e;
    t = k(async (n) => r(n));
  } else {
    const r = e;
    t = k(async (n) => r);
  }
  return t;
}
function R(e) {
  return e ? q.store(e) ? e : x(e, {
    serialize: "ignore",
    name: "ff.$target/valueField",
    sid: "ff.$target/$valueField"
  }) : x(null, {
    serialize: "ignore",
    name: "ff.$target/undefined",
    sid: "ff.$target/$undefined"
  });
}
function yt(e) {
  return q.store(e) ? e : e != null && e.source ? e.source : null;
}
function Sr(e, t) {
  const r = {}, n = {};
  for (const [o, s] of Object.entries(e))
    q.store(s) ? r[o] = s : s != null && s.source && (s != null && s.fn) ? (r[o] = s.source, n[o] = s.fn) : typeof s == "function" ? n[o] = s : r[o] = s;
  return {
    source: O(r),
    fn: (o, s) => {
      const i = {};
      for (const l of Object.keys(e))
        l in s && (i[l] = s[l]), l in n && (i[l] = n[l](o, s[l]));
      return t ? t(i) : i;
    }
  };
}
function Te({
  clock: e,
  timeout: t,
  target: r = $()
}) {
  const n = k(
    ({ payload: a, milliseconds: o }) => new Promise((s) => {
      setTimeout(s, o, a);
    })
  );
  return c({
    source: R(t),
    clock: e,
    fn: (a, o) => ({
      payload: o,
      milliseconds: a
    }),
    target: n
  }), c({
    clock: n.doneData,
    target: r
  }), r;
}
function le(e, t) {
  let r = [], n = () => !1;
  Array.isArray(e) ? (r = e, n = t) : Array.isArray(e.stores) && (r = e.stores, n = e.predicate);
  let a;
  gt(n) ? a = n : q.store(n) ? a = n.map((i) => (l) => i === l) : a = (i) => i === n;
  const o = O(r);
  return O(a, o, (i, l) => l.every(i));
}
function gt(e) {
  return typeof e == "function";
}
function H(e) {
  return e.map((t) => !t);
}
function De(...e) {
  return le({ predicate: !0, stores: e });
}
function je({
  clock: e,
  until: t
}) {
  const r = $(), n = x(!1, {
    serialize: "ignore",
    name: "ff.$fired",
    sid: "ff.$fired"
  }).on(r, () => !0).on(e, () => !1);
  return c({
    clock: [e, t],
    source: e,
    filter: De(t, H(n)),
    target: r
  }), r;
}
function se(e) {
  if (e === "ignore")
    return "ignore";
}
const ke = k(() => Date.now());
function $t({
  clock: e
}) {
  const t = x(Date.now(), {
    name: "ff.$time",
    sid: "ff.$time"
  });
  return c({
    clock: e,
    fn: () => {
    },
    target: ke
  }), c({ clock: ke.doneData, target: t }), t;
}
function E(e) {
  return e.map((t) => t, { skipVoid: !1 });
}
function _t(e) {
  const t = $(), r = x(null, {
    serialize: "ignore",
    name: "ff.$timeoutId",
    sid: "ff.$timeoutId"
  }).on(t, (g, f) => f), n = $(), a = x(null, {
    serialize: "ignore",
    name: "ff.$rejecter",
    sid: "ff.$rejecter"
  }).on(n, (g, f) => f), o = $(), s = T({
    source: {
      timeoutId: r,
      rejectPromise: a
    },
    effect: ({ timeoutId: g, rejectPromise: f }) => (g && clearTimeout(g), f && f(), new Promise((_, u) => {
      n(u), t(setTimeout(_, 0));
    }))
  });
  a.reset(s.done), r.reset(s.done);
  const i = x([], {
    serialize: "ignore",
    name: "ff.$payload",
    sid: "ff.$payload"
  }).on(e, (g, f) => [f]), l = x(!0, {
    serialize: "ignore",
    sid: "ff.$canTick",
    name: "ff.$canTick"
  }), y = $();
  return l.on(y, () => !1).on(
    [
      o,
      // debounce timeout can be restarted in later ticks
      s
    ],
    () => !0
  ), c({
    clock: e,
    filter: l,
    target: y
  }), c({
    clock: y,
    target: s
  }), c({
    source: i,
    clock: s.done,
    fn: ([g]) => g,
    target: o
  }), o;
}
function ve(e, t) {
  const r = Object.entries(e).map(
    ([n, a]) => [n, t(a)]
  );
  return Object.fromEntries(r);
}
function be(e) {
  const t = {};
  for (const [r, n] of Object.entries(e))
    for (const [a, o] of Object.entries(n))
      t[a] = { ...t[a], [r]: o };
  return t;
}
function kt({ min: e, max: t }) {
  return Math.random() * (t - e) + e;
}
function vt() {
  const e = {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    resolve: () => {
    },
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    reject: () => {
    },
    // @ts-expect-error it will be set later
    promise: null
  };
  return e.promise = new Promise((t, r) => {
    e.resolve = t, e.reject = r;
  }), e.promise.catch(() => {
  }), e;
}
function bt(e) {
  return e == null;
}
function we(e) {
  return !bt(e);
}
function Z(e, t) {
  try {
    if (e === t || Number.isNaN(e) && Number.isNaN(t))
      return !0;
    const r = Ae(e), n = Ae(t);
    if (r !== n)
      return !1;
    if (r === "pure-object") {
      if (e === t)
        return !0;
      const a = Object.keys(e), o = Object.keys(t).length;
      if (a.length !== o)
        return !1;
      for (let s = 0, i = a.length; s < i; s++) {
        const l = a[s];
        if (!Object.prototype.hasOwnProperty.call(t, a[s]))
          return !1;
        const y = e[l], g = t[l];
        if (y === e || g === t || y === t || g === e)
          return y === g;
        if (!Z(y, g))
          return !1;
      }
      return !0;
    } else if (r === "array") {
      if (e.length === t.length)
        for (let a = 0; a < e.length; a++) {
          const o = e[a], s = t[a];
          if (o === e || s === t || o === t || s === e)
            return o === s;
          if (!Z(o, s))
            return !1;
        }
      else
        return !1;
      return !0;
    } else if (r === "object" && e.valueOf && t.valueOf)
      return e.valueOf() === t.valueOf();
  } catch {
  }
  return !1;
}
function wt(e) {
  if (Array.isArray(e))
    return !0;
  const t = e.length;
  return typeof t == "number" && t > -1 ? t ? 0 in e && t - 1 in e : !0 : !1;
}
function Ae(e) {
  const t = typeof e;
  return t === "object" ? e === null ? "null" : wt(e) ? "array" : e.constructor === Object || Object.getPrototypeOf(e) === null ? "pure-object" : "object" : t;
}
function At(e, t) {
  const r = [], n = [];
  for (const a of e)
    t(a) ? r.push(a) : n.push(a);
  return [r, n];
}
function U(e) {
  return (t) => t[e];
}
function xt(e) {
  return Array.from(new Set(e));
}
class Et {
  constructor(t = !1) {
    this._resolve = null, this._promise = Promise.resolve(), t && this.acquire();
  }
  get isLocked() {
    return !!this._resolve;
  }
  acquire() {
    this.isLocked || (this._promise = new Promise((t) => {
      this._resolve = t;
    }));
  }
  waitForUnlock() {
    return this._promise;
  }
  release() {
    this._resolve && (this._resolve(), this._resolve = null);
  }
}
function Pt({
  events: e,
  reset: t
}) {
  const r = $(), n = x([], { serialize: "ignore" });
  return t && c({ clock: t, target: n.reinit }), e.forEach((a, o) => {
    n.on(a, (s) => xt([...s, o]));
  }), c({
    clock: e,
    source: n,
    filter: (a) => a.length === e.length,
    target: r
  }), r;
}
function Oe() {
  var e, t;
  return ((t = (e = Fe({ regional: !0 }).family.owners[0]) == null ? void 0 : e.meta) == null ? void 0 : t.name) ?? void 0;
}
const Re = "INVALID_DATA", qe = "TIMEOUT", Ne = "ABORT", Le = "PREPARATION", Me = "HTTP", ze = "NETWORK", St = "CONFIGURATION";
function ee(e) {
  return {
    ...e,
    errorType: Re,
    explanation: "Response was considered as invalid against a given contract"
  };
}
function Ft(e) {
  return {
    ...e,
    errorType: qe,
    explanation: "Request was cancelled due to timeout"
  };
}
function Ce() {
  return {
    errorType: Ne,
    explanation: "Request was cancelled due to concurrency policy"
  };
}
function It(e) {
  return {
    ...e,
    errorType: Le,
    explanation: "Extraction of data from the response was failed"
  };
}
function Be(e) {
  return {
    ...e,
    errorType: Me,
    explanation: "Request was finished with unsuccessful HTTP code"
  };
}
function Tt(e) {
  return {
    ...e,
    errorType: ze,
    explanation: "Request was failed due to network problems"
  };
}
function ie(e) {
  return {
    ...e,
    errorType: St,
    explanation: "Operation is misconfigured"
  };
}
function Dt(e) {
  return k({
    handler: ({ result: r }) => {
      if (!e.isData(r))
        throw ee({
          validationErrors: e.getErrorMessages(r),
          response: r
        });
      return r;
    },
    sid: "ff.applyContractFx"
  });
}
function jt(e) {
  return !!(e === !0 || Array.isArray(e) && e.length === 0 || typeof e == "string" && e.length === 0);
}
function xe(e) {
  return e === !0 ? [] : e === !1 ? ["Invalid data"] : Array.isArray(e) ? e.length === 0 ? [] : e : [e];
}
const Ot = () => !0;
function Fr(e) {
  var t;
  return ((t = e.error) == null ? void 0 : t.errorType) === Re;
}
function Ir(e) {
  var t;
  return ((t = e.error) == null ? void 0 : t.errorType) === qe;
}
function Rt(e) {
  var t;
  return ((t = e.error) == null ? void 0 : t.errorType) === Ne;
}
function Tr(e) {
  var t;
  return ((t = e.error) == null ? void 0 : t.errorType) === Le;
}
function He(e) {
  var t;
  return ((t = e.error) == null ? void 0 : t.errorType) === Me;
}
function Dr(e) {
  return function(r) {
    return He(r) ? (Array.isArray(e) ? e : [e]).includes(r.error.status) : !1;
  };
}
function jr(e) {
  var t;
  return ((t = e.error) == null ? void 0 : t.errorType) === ze;
}
let te = null, ue = !1;
function qt() {
  ue = !0;
}
function Nt() {
  ue = !1;
}
function Ue(e) {
  if (!ue)
    throw ie({
      reason: "onAbort call is not allowed",
      validationErrors: [
        "onAbort can be called only in the context of a handler before any async operation is performed"
      ]
    });
  if (te)
    throw ie({
      reason: "onAbort call is not allowed",
      validationErrors: ["onAbort can be called only once per operation"]
    });
  te = e;
}
function Lt() {
  const e = te;
  return te = null, e;
}
function Mt(e) {
  const t = $(e.shortName + ".internalCall"), r = dt.compute({
    fn: (a) => {
      const o = a.handler, s = zt(o, t);
      return a.handler = s, a;
    }
  });
  return Ct(e).seq.splice(1, 0, r), t;
}
function zt(e, t) {
  function r(...n) {
    const { result: a, abortCallback: o } = Ut(e, ...n);
    if (a instanceof Promise) {
      const s = vt(), i = Ee(s, o);
      return t(i), a.then(s.resolve, s.reject), s.promise;
    } else {
      const s = Ee(void 0, o);
      return t(s), a;
    }
  }
  return r;
}
function Ee(e, t) {
  let r = e ? "pending" : "finished";
  function n() {
    r = "finished", a.status = r;
  }
  e && e.promise.then(n, n);
  const a = {
    id: Ht(),
    status: r,
    abort: (o = Ce()) => {
      t == null || t(), r !== "finished" && e && e.reject(o);
    },
    promise: e == null ? void 0 : e.promise
  };
  return a;
}
function Ct(e) {
  return e.graphite.scope.runner;
}
let Bt = 0;
function Ht() {
  return `${Bt++}`;
}
function Ut(e, ...t) {
  Vt(), qt();
  const r = e(...t), n = Lt();
  return Nt(), {
    result: r,
    abortCallback: n
  };
}
const Vt = $();
function Ve(e, {
  strategy: t,
  abortAll: r
}) {
  e.__.meta.flags.concurrencyFieldUsed && console.error(
    `Both concurrency-operator and concurrency-field are used on operation ${e.__.meta.name}.`,
    "Please use only concurrency-operator, because field concurrency-field in createJsonQuery and createJsonMutation is deprecated and will be deleted soon."
  ), e.__.meta.flags.concurrencyOperatorUsed = !0;
  const n = Qe(e);
  if (t)
    switch (t) {
      case "TAKE_EVERY":
        break;
      case "TAKE_FIRST": {
        const a = x(!1, { serialize: "ignore" }), o = $();
        c({
          clock: o,
          fn: () => !0,
          target: a
        });
        const s = {
          name: "concurrency",
          get: T({
            source: a,
            async effect(i, l) {
              if (i)
                throw Ce();
              return o(), null;
            }
          })
        };
        e.__.lowLevelAPI.dataSources.unshift(s), c({
          clock: e.finished.finally,
          fn: () => !1,
          target: a
        });
        break;
      }
      case "TAKE_LATEST": {
        c({
          clock: e.__.lowLevelAPI.callObjectCreated,
          source: n,
          fn: (a, o) => a.filter((s) => s !== o),
          target: fe
        });
        break;
      }
    }
  return r && Ke(e, { clock: r }), e;
}
function Ke(e, { clock: t }) {
  c({ clock: t, source: Qe(e), target: fe });
}
function Qe(e) {
  if (!e.__.meta.$callObjects) {
    const t = x([], { serialize: "ignore" });
    c({
      clock: e.__.lowLevelAPI.callObjectCreated,
      source: t,
      fn: (r, n) => r.filter((a) => a.status === "pending").concat(n),
      target: t
    }), c({
      clock: fe.done,
      source: t,
      fn: (r, { params: n }) => r.filter((a) => !n.includes(a)),
      target: t
    }), e.__.meta.$callObjects = t;
  }
  return e.__.meta.$callObjects;
}
const fe = k(
  (e) => e.forEach((t) => t.abort())
), ne = Symbol("Meta"), ae = Symbol("Result");
function Kt(e) {
  return typeof e == "object" && e !== null && ne in e && ae in e;
}
function Je({
  name: e,
  meta: t,
  kind: r,
  serialize: n,
  enabled: a,
  contract: o,
  validate: s,
  mapData: i,
  sourced: l,
  paramsAreMeaningless: y
}) {
  const g = $(), f = $(), _ = $(), u = $(), p = Dt(
    o
  ), m = e ?? "unnamed", d = k({
    handler: () => {
      throw new Error("Not implemented");
    },
    sid: `ff.${m}.executeFx`,
    name: `${m}.executeFx`
  }), w = Mt(d), F = [{
    name: "remote_source",
    get: k(async ({ params: h }) => {
      const v = await d(h);
      return Kt(v) ? {
        result: v[ae],
        stale: !1,
        meta: v[ne]
      } : { result: v, stale: !1 };
    })
  }], {
    retrieveDataFx: I,
    notifyAboutNewValidDataFx: N,
    notifyAboutDataInvalidationFx: C
  } = Qt(F), B = $(), b = $(), S = $();
  c({
    clock: B,
    fn: (h) => ({
      params: h,
      meta: { stopErrorPropagation: !1, stale: !0 }
    }),
    target: u
  });
  const A = {
    success: $(),
    failure: $(),
    skip: $(),
    finally: $()
  }, M = $(), V = $(), K = $();
  J({
    source: M,
    match: {
      aborted: ({ error: h }) => Rt({ error: h })
    },
    cases: {
      aborted: K,
      __: A.failure
    }
  });
  const L = x("initial", {
    sid: `ff.${m}.$status`,
    name: `ff.${m}.$status`,
    serialize: n
  });
  c({ clock: b, target: L.reinit });
  const G = x([], {
    serialize: "ignore",
    name: `ff.${m}.$statusHistory`,
    sid: `ff.${m}.$statusHistory`
  });
  c({
    clock: L.updates,
    source: G,
    fn: (h, v) => [...h, v],
    target: G
  });
  const Q = R(a ?? !0).map(Boolean), ye = x(void 0, {
    serialize: "ignore",
    name: `ff.${m}.$latestParams`,
    sid: `ff.${m}.$latestParams`,
    skipVoid: !1
  }), ct = L.map((h) => h === "initial"), it = L.map((h) => h === "pending"), lt = L.map((h) => h === "fail"), ut = L.map((h) => h === "done"), ft = L.map((h) => ["fail", "done"].includes(h));
  c({
    clock: [
      I.map(() => "pending"),
      A.success.map(() => "done"),
      A.failure.map(() => "fail"),
      c({
        clock: K,
        source: {
          history: G,
          retrieveDataPengind: I.pending
        },
        fn: ({ history: h, retrieveDataPengind: v }) => v ? "pending" : h[h.length - 2] ?? "initial"
      })
    ],
    target: L
  }), c({
    clock: u,
    filter: Q,
    fn: ({ params: h }) => h,
    target: ye
  }), c({
    clock: B,
    filter: H(Q),
    fn(h) {
      return {
        params: h,
        meta: { stopErrorPropagation: !1, stale: !1 }
      };
    },
    target: A.skip
  }), c({
    clock: g,
    fn: ({ params: h }) => ({
      params: h,
      meta: { stopErrorPropagation: !1, stale: !0 }
    }),
    target: C
  }), c({
    clock: C.finally,
    source: g,
    filter: ({ refresh: h }) => h,
    fn: ({ params: h }) => ({
      params: h,
      meta: { stopErrorPropagation: !1, stale: !0 }
    }),
    target: u
  }), c({
    clock: u,
    filter: Q,
    target: I
  }), c({ clock: I, target: S }), c({
    clock: I.done,
    fn: ({ params: h, result: v }) => ({
      params: h.params,
      result: v.result,
      meta: {
        stopErrorPropagation: v.stopErrorPropagation ?? !1,
        stale: v.stale,
        responseMeta: v.meta
      }
    }),
    filter: Q,
    target: p
  }), c({
    clock: I.fail,
    source: Q,
    filter: (h, { error: v }) => h && !v.stopErrorPropagation,
    fn: (h, { error: v, params: j }) => ({
      error: v.error,
      params: j.params,
      meta: { stopErrorPropagation: v.stopErrorPropagation, stale: !1 }
    }),
    target: M
  });
  const { validDataRecieved: ge, __: $e } = J(
    c({
      clock: p.done,
      source: {
        partialValidator: D({
          field: s ?? Ot
        })
      },
      fn: ({ partialValidator: h }, {
        params: {
          /* Extract original params, it is params of params */
          params: v,
          meta: j
        },
        result: z
      }) => ({
        result: z,
        params: v,
        validation: h({
          result: z,
          params: v
        }),
        meta: j
      })
    }),
    {
      validDataRecieved: ({ validation: h }) => jt(h)
    }
  );
  c({
    clock: ge,
    source: {
      partialMapper: D({
        field: i
      })
    },
    fn: ({ partialMapper: h }, { params: v, result: j, meta: z }) => ({
      result: h({
        params: v,
        result: j,
        ...Jt(z) ? z.responseMeta : {}
      }),
      params: v,
      meta: z
    }),
    target: A.success
  }), c({
    clock: A.success,
    filter: ({ meta: h }) => h.stale,
    fn: ({ params: h, meta: v }) => ({ params: h, meta: v, skipStale: !0 }),
    target: I
  }), c({
    clock: ge,
    filter: ({ meta: h }) => !h.stale,
    target: N
  }), c({
    clock: p.fail,
    filter: ({ params: h }) => !h.meta.stopErrorPropagation,
    fn: ({ error: h, params: v }) => ({
      error: h,
      // Extract original params, it is params of params
      params: v.params,
      meta: v.meta
    }),
    target: M
  }), c({
    clock: p.fail,
    fn: ({ error: h, params: v }) => ({
      error: h,
      // Extract original params, it is params of params
      params: v.params,
      meta: v.meta
    }),
    target: V
  }), c({
    clock: $e,
    filter: ({ meta: h }) => !h.stopErrorPropagation,
    fn: ({ params: h, validation: v, meta: j, result: z }) => ({
      params: h,
      error: ee({
        validationErrors: xe(v),
        response: z
      }),
      meta: j
    }),
    target: M
  }), c({
    clock: $e,
    fn: ({ params: h, validation: v, meta: j, result: z }) => ({
      params: h,
      error: ee({
        validationErrors: xe(v),
        response: z
      }),
      meta: j
    }),
    target: V
  }), c({
    clock: Q.updates,
    source: B,
    filter: H(Q),
    fn: (h) => ({
      params: h,
      meta: {
        stopErrorPropagation: !1,
        stale: !0
      }
    }),
    target: A.skip
  }), c({
    clock: A.success,
    fn: ({ params: h, result: v, meta: j }) => ({
      status: "done",
      params: h,
      result: v,
      meta: j
    }),
    target: A.finally
  }), c({
    clock: A.failure,
    fn: ({ params: h, error: v, meta: j }) => ({
      status: "fail",
      params: h,
      error: v,
      meta: j
    }),
    target: A.finally
  }), c({
    clock: A.skip,
    fn: ({ params: h, meta: v }) => ({
      status: "skip",
      params: h,
      meta: v
    }),
    target: A.finally
  });
  const _e = {
    start: B,
    finished: A,
    started: S,
    aborted: K,
    reset: b,
    $status: L,
    $idle: ct,
    $pending: it,
    $failed: lt,
    $succeeded: ut,
    $finished: ft,
    $enabled: Q,
    __: {
      executeFx: d,
      meta: { ...t, name: m, flags: {} },
      kind: r,
      $latestParams: E(ye),
      lowLevelAPI: {
        dataSources: F,
        dataSourceRetrieverFx: I,
        sourced: l ?? [],
        paramsAreMeaningless: y ?? !1,
        revalidate: g,
        pushError: _,
        pushData: f,
        startWithMeta: u,
        callObjectCreated: w,
        failedIgnoreSuppression: V
      }
    }
  };
  return Ke(_e, { clock: b }), _e;
}
function Qt(e) {
  const t = k({
    handler: async ({ params: a, skipStale: o }) => {
      for (const s of e)
        try {
          const i = await s.get({ params: a });
          if (o && (i != null && i.stale))
            continue;
          if (i)
            return i;
        } catch (i) {
          throw {
            stopErrorPropagation: !1,
            error: i
          };
        }
      throw {
        stopErrorPropagation: !1,
        error: new Error("No data source returned data")
      };
    }
  }), r = k({
    handler: async ({ params: a, result: o }) => {
      await Promise.all(
        e.map(U("set")).filter(Boolean).map((s) => s({ params: a, result: o }))
      );
    }
  }), n = k({
    handler: async ({ params: a }) => {
      await Promise.all(
        e.map(U("unset")).filter(Boolean).map((o) => o({ params: a }))
      );
    }
  });
  return {
    retrieveDataFx: t,
    notifyAboutNewValidDataFx: r,
    notifyAboutDataInvalidationFx: n
  };
}
function Jt(e) {
  return "responseMeta" in e && typeof e.responseMeta == "object";
}
const We = Symbol("Query");
function Ge(e) {
  var t;
  return ((t = e == null ? void 0 : e.__) == null ? void 0 : t.kind) === We;
}
const Wt = Symbol("Farfetched node meta"), Gt = Symbol("Farfetched links meta");
function Yt(e, t) {
  return Fe({
    meta: { [Wt]: e, [Gt]: t },
    regional: !0
  });
}
function de(e) {
  const {
    initialData: t,
    contract: r,
    mapData: n,
    enabled: a,
    validate: o,
    name: s,
    serialize: i,
    sourced: l,
    paramsAreMeaningless: y
  } = e, g = t ?? null, f = Je({
    name: s ?? Oe(),
    kind: We,
    serialize: se(i),
    enabled: a,
    meta: {
      serialize: i,
      initialData: g,
      sid: Xt(x(null, { sid: "dummy" }))
    },
    contract: r,
    validate: o,
    mapData: n,
    sourced: l,
    paramsAreMeaningless: y
  }), _ = $(), u = x(g, {
    sid: `ff.${f.__.meta.name}.$data`,
    name: `${f.__.meta.name}.$data`,
    serialize: i,
    skipVoid: !1
  }), p = x(null, {
    sid: `ff.${f.__.meta.name}.$error`,
    name: `${f.__.meta.name}.$error`,
    serialize: se(i),
    skipVoid: !1
  }), m = x(!0, {
    sid: `ff.${f.__.meta.name}.$stale`,
    name: `${f.__.meta.name}.$stale`,
    serialize: se(i),
    skipVoid: !1
  });
  c({
    clock: f.finished.success,
    fn: () => null,
    target: p
  }), c({
    clock: f.finished.success,
    fn: ({ result: b }) => b,
    target: u
  }), u.reset(f.finished.failure), c({
    clock: f.finished.failure,
    fn: ({ error: b }) => b,
    target: p
  }), c({
    clock: f.finished.finally,
    fn: ({ meta: b }) => b.stale,
    target: m
  }), c({
    clock: f.__.lowLevelAPI.pushData,
    target: [u, p.reinit]
  }), c({
    clock: f.__.lowLevelAPI.pushError,
    target: [p, u.reinit]
  });
  const d = je({
    clock: _,
    until: f.$enabled
  }), w = $(), { haveToStart: P, __: F } = J(
    c({
      clock: d,
      source: { stale: m, latestParams: f.__.$latestParams },
      fn: ({ stale: b, latestParams: S }, A) => ({
        haveToStart: b || !Z(A, S),
        params: A
      })
    }),
    {
      haveToStart: ({ haveToStart: b }) => b
    }
  );
  c({
    clock: F,
    fn: () => null,
    target: w
  }), c({
    clock: P,
    fn: ({ params: b }) => b,
    target: f.start
  }), c({
    clock: f.reset,
    target: [u.reinit, p.reinit, m.reinit]
  });
  const I = {
    data: u,
    error: p,
    stale: m,
    pending: f.$pending,
    start: f.start
  }, N = () => I, C = ({
    source: b,
    mapParams: S
  }) => {
    const A = de(e);
    return A.__.lowLevelAPI.dataSourceRetrieverFx.use(
      T({
        source: b,
        mapParams: ({ params: M, ...V }, K) => ({
          params: S ? S(M, K) : M,
          ...V
        }),
        effect: f.__.lowLevelAPI.dataSourceRetrieverFx
      })
    ), A;
  }, B = Yt(
    { type: "query", name: e.name },
    { $status: f.$status, $data: u, $error: p }
  );
  return pt(B, () => ({
    refresh: _,
    start: f.start,
    reset: f.reset,
    started: E(f.started),
    $data: E(u),
    $error: E(p),
    $status: E(f.$status),
    $idle: E(f.$idle),
    $pending: E(f.$pending),
    $succeeded: E(f.$succeeded),
    $failed: E(f.$failed),
    $finished: E(f.$finished),
    $enabled: E(f.$enabled),
    $stale: m,
    aborted: E(f.aborted),
    finished: {
      success: E(f.finished.success),
      failure: E(f.finished.failure),
      finally: E(f.finished.finally),
      skip: E(f.finished.skip)
    },
    __: {
      ...f.__,
      lowLevelAPI: { ...f.__.lowLevelAPI, refreshSkipDueToFreshness: w },
      experimentalAPI: { attach: C }
    },
    "@@unitShape": N
  }));
}
function Xt(e) {
  const t = e.sid;
  return t != null && t.includes("|") ? t : null;
}
function Ye(e) {
  const t = e;
  if (q.effect(t.effect))
    return t.effect;
  if (typeof t.handler == "function")
    return k(t.handler);
  throw new Zt(
    "handler or effect must be passed to the config"
  );
}
class Zt extends Error {
  constructor(t) {
    super(t);
  }
}
function Or(e) {
  const t = de({
    initialData: e.initialData ?? null,
    contract: e.contract ?? re,
    mapData: e.mapData ?? (({ result: r }) => r),
    enabled: e.enabled,
    validate: e.validate,
    name: e.name,
    serialize: e.serialize
  });
  return t.__.executeFx.use(Ye(e)), t;
}
function Rr(e) {
  const { source: t, target: r } = e, n = Ge(t), a = Array.isArray(r) ? r : [r], o = n ? [t] : Object.values(t), s = e == null ? void 0 : e.fn, i = e == null ? void 0 : e.filter, l = ce(o.map((u) => u.start)), y = ce(
    o.map((u) => u.finished.success)
  ), g = le({
    stores: o.map((u) => u.$data),
    predicate: (u) => u !== null
  }), f = n ? t.$data : O(ve(t, (u) => u.$data)), _ = x(null, {
    serialize: "ignore",
    name: "ff.$allParentParamsDictionary",
    sid: "ff.$allParentParamsDictionary",
    skipVoid: !1
  });
  if (n)
    c({
      clock: t.finished.success,
      fn: ({ params: u }) => u,
      target: _
    });
  else
    for (const [u, p] of Object.entries(
      ve(t, (m) => m.finished.success)
    ))
      c({
        clock: p,
        source: _,
        fn: (m, { params: d }) => ({
          ...m,
          [u]: d
        }),
        target: _
      });
  c({
    clock: l,
    fn() {
      return !0;
    },
    target: a.map((u) => u.$stale)
  }), c({
    clock: je({
      clock: y,
      until: g
    }),
    source: {
      data: f,
      params: _
    },
    filter({ data: u, params: p }) {
      return (i == null ? void 0 : i(
        n ? { result: u, params: p } : be({ result: u, params: p })
      )) ?? !0;
    },
    fn({ data: u, params: p }) {
      const m = s == null ? void 0 : s(
        n ? { result: u, params: p } : be({ result: u, params: p })
      );
      return m == null ? void 0 : m.params;
    },
    target: a.map((u) => u.start)
  });
}
function Xe(...e) {
  const t = {};
  for (const r of e)
    if (typeof r == "object")
      for (const [n, a] of Object.entries(r || {})) {
        const o = pe(a);
        o !== null && (t[n] ? t[n] = [t[n], o].flat() : t[n] = o);
      }
  return t;
}
function er(...e) {
  const t = [];
  for (const r of e) {
    if (!r)
      continue;
    let n;
    typeof r != "string" ? n = Ze(r).toString() : n = r, t.push(n);
  }
  return t.join("&");
}
function tr(e) {
  const t = new Headers();
  for (const [r, n] of Object.entries(e)) {
    const a = pe(n);
    if (Array.isArray(a))
      for (const o of a)
        t.append(r, o);
    else
      a !== null && t.append(r, a);
  }
  return t;
}
function rr(e, t) {
  let r;
  e.startsWith("/") && (r = window.location.origin);
  let n, a;
  typeof t == "string" ? a = t : a = Ze(t).toString(), a ? n = `${e}?${a}` : n = e;
  const o = [n, r].filter(Boolean);
  try {
    return new URL(...o);
  } catch {
    throw ie({
      reason: "Invalid URL",
      validationErrors: [`"${n}" is not valid URL`]
    });
  }
}
function Ze(e) {
  const t = new URLSearchParams();
  for (const [r, n] of Object.entries(e)) {
    const a = pe(n);
    if (Array.isArray(a))
      for (const o of a)
        t.append(r, o);
    else
      a !== null && t.append(r, a);
  }
  return t;
}
function pe(e) {
  return typeof e == "number" || typeof e == "boolean" ? e.toString() : e ?? null;
}
const nr = k({
  sid: "ff.fetchFx",
  handler: globalThis.fetch
}), ar = k({
  handler: async (e) => {
    const t = await nr(e).catch((r) => {
      throw Tt({
        reason: (r == null ? void 0 : r.message) ?? null,
        cause: r
      });
    });
    if (!t.ok)
      throw Be({
        status: t.status,
        statusText: t.statusText,
        response: await t.text().catch(() => null) ?? null
      });
    return t;
  },
  sid: "ff.requestFx"
});
function sr(e) {
  const t = k(e.response.extract), r = k(
    async ({
      url: n,
      method: a,
      query: o,
      headers: s,
      credentials: i,
      body: l,
      abortController: y
    }) => {
      const g = l ? e.request.mapBody(l) : null, f = new Request(rr(n, o), {
        method: a,
        headers: tr(s),
        credentials: i,
        body: g,
        signal: y == null ? void 0 : y.signal
      }), _ = await ar(f).catch((m) => {
        throw e.response.transformError ? e.response.transformError(m) : m;
      }), u = _.clone(), p = await t(_).catch(async (m) => {
        throw It({
          response: await u.text(),
          reason: (m == null ? void 0 : m.message) ?? null
        });
      });
      if (e.response.status) {
        const m = Array.isArray(e.response.status.expected) ? e.response.status.expected : [e.response.status.expected];
        if (!m.includes(_.status))
          throw ee({
            validationErrors: [
              `Expected response status has to be one of [${m.join(
                ", "
              )}], got ${_.status}`
            ],
            response: p
          });
      }
      return { result: p, meta: { headers: _.headers } };
    }
  );
  return T({
    source: {
      url: R(e.request.url),
      method: R(e.request.method),
      query: R(e.request.query),
      headers: R(e.request.headers),
      credentials: R(e.request.credentials),
      body: R(e.request.body)
    },
    mapParams(n, a) {
      const o = a.url ?? // @ts-expect-error TS cannot infer type correctly, but there is always field in staticConfig or dynamicConfig
      n.url, s = a.credentials ?? // @ts-expect-error TS cannot infer type correctly, but there is always field in staticConfig or dynamicConfig
      n.credentials, i = a.body ?? // @ts-expect-error TS cannot infer type correctly, but there is always field in staticConfig or dynamicConfig
      n.body, l = er(a.query, n.query), y = Xe(a.headers, n.headers), { method: g } = a, { abortController: f } = n;
      return {
        url: o,
        method: g,
        // TODO: fix type inference here
        query: l,
        headers: y,
        credentials: s,
        body: i,
        abortController: f
      };
    },
    effect: r
  });
}
function et(e) {
  var n;
  const t = O(
    {
      method: R(e.request.method),
      headers: R(e.request.headers)
    },
    ({ method: a, headers: o }) => (
      // reversed merge order to allow any modifications in the user code
      Xe(
        {
          Accept: "application/json",
          "Content-Type": ["GET", "HEAD"].includes(a) ? void 0 : "application/json"
        },
        o
      )
    )
  );
  return sr({
    ...e,
    request: {
      ...e.request,
      headers: t,
      // Serialize body to JSON-string
      mapBody: (a) => JSON.stringify(a)
    },
    response: {
      extract: async (a) => await or(a) ? null : a.json(),
      transformError: (a) => {
        if (!He({ error: a }))
          return a;
        const o = a;
        if (typeof o.response != "string")
          return o;
        try {
          const s = JSON.parse(o.response);
          return Be({
            status: o.status,
            statusText: o.statusText,
            response: s
          });
        } catch {
          return o;
        }
      },
      status: (n = e.response) == null ? void 0 : n.status
    }
  });
}
async function or(e) {
  return !e.body || e.headers.get("Content-Length") === "0" || (await e.clone().text()).length === 0;
}
function qr(e) {
  const t = e.request.credentials, r = et({
    request: {
      method: e.request.method,
      credentials: t
    }
  }), n = de({
    initialData: e.initialData,
    contract: e.response.contract ?? re,
    mapData: e.response.mapData ?? (({ result: s }) => s),
    validate: e.response.validate,
    enabled: e.enabled,
    name: e.name,
    serialize: e.serialize,
    sourced: [
      e.request.url,
      e.request.body,
      e.request.headers,
      e.request.query
    ],
    paramsAreMeaningless: !0
  }), a = k(async (s) => {
    const i = new AbortController();
    Ue(() => i.abort());
    const { result: l, meta: y } = await r({ ...s, abortController: i });
    return { [ae]: l, [ne]: y };
  });
  n.__.executeFx.use(
    T({
      source: {
        partialUrl: D({
          field: e.request.url
        }),
        partialBody: D({
          field: e.request.body
        }),
        partialHeaders: D({
          field: e.request.headers
        }),
        partialQuery: D({
          field: e.request.query
        })
      },
      mapParams(s, { partialUrl: i, partialBody: l, partialHeaders: y, partialQuery: g }) {
        return {
          url: i(s),
          body: l(s),
          headers: y(s),
          query: g(s)
        };
      },
      effect: a
    })
  );
  const o = {
    ...n,
    __: { ...n.__, executeFx: a }
  };
  return e.concurrency && (console.error(
    "concurrency field in createJsonQuery is deprecated, please use concurrency operator instead: https://ff.effector.dev/adr/concurrency.html"
  ), o.__.meta.flags.concurrencyFieldUsed = !0), setTimeout(() => {
    var s, i;
    o.__.meta.flags.concurrencyOperatorUsed || (console.error(
      "Please apply concurrency operator to the query, read more: https://ff.effector.dev/adr/concurrency.html"
    ), Ve(o, {
      strategy: ((s = e.concurrency) == null ? void 0 : s.strategy) ?? "TAKE_LATEST",
      abortAll: (i = e.concurrency) == null ? void 0 : i.abort
    }));
  }), o;
}
const tt = Symbol("Mutation");
function cr(e) {
  var t;
  return ((t = e == null ? void 0 : e.__) == null ? void 0 : t.kind) === tt;
}
function me(e) {
  const { name: t, enabled: r, contract: n, validate: a, mapData: o } = e, s = Je({
    name: t ?? Oe(),
    serialize: "ignore",
    enabled: r,
    kind: tt,
    meta: null,
    contract: n,
    validate: a,
    mapData: o
  }), i = {
    pending: s.$pending,
    start: s.start,
    reset: s.reset
  }, l = () => i, y = ({
    source: g,
    mapParams: f
  }) => {
    const _ = me(e);
    return _.__.lowLevelAPI.dataSourceRetrieverFx.use(
      T({
        source: g,
        mapParams: ({ params: u, ...p }, m) => ({
          params: f ? f(u, m) : u,
          ...p
        }),
        effect: s.__.lowLevelAPI.dataSourceRetrieverFx
      })
    ), _;
  };
  return {
    start: s.start,
    reset: s.reset,
    started: E(s.started),
    aborted: E(s.aborted),
    $status: E(s.$status),
    $idle: E(s.$idle),
    $pending: E(s.$pending),
    $succeeded: E(s.$succeeded),
    $failed: E(s.$failed),
    $finished: E(s.$finished),
    $enabled: E(s.$enabled),
    finished: {
      success: E(s.finished.success),
      failure: E(s.finished.failure),
      finally: E(s.finished.finally),
      skip: E(s.finished.skip)
    },
    __: { ...s.__, experimentalAPI: { attach: y } },
    "@@unitShape": l
  };
}
function Nr(e) {
  const t = me({
    name: e.name,
    enabled: e.enabled,
    contract: e.contract ?? re,
    mapData: ({ result: r }) => r
  });
  return t.__.executeFx.use(Ye(e)), t;
}
function Lr(e) {
  const t = e.request.credentials, r = et({
    request: { method: e.request.method, credentials: t },
    response: { status: e.response.status }
  }), n = me({
    contract: e.response.contract ?? re,
    mapData: e.response.mapData ?? (({ result: s }) => s),
    validate: e.response.validate,
    enabled: e.enabled,
    name: e.name
  }), a = k(async (s) => {
    const i = new AbortController();
    Ue(() => i.abort());
    const { result: l, meta: y } = await r({ ...s, abortController: i });
    return { [ae]: l, [ne]: y };
  });
  n.__.executeFx.use(
    T({
      source: {
        partialUrl: D({
          field: e.request.url
        }),
        partialBody: D({
          field: e.request.body
        }),
        partialHeaders: D({
          field: e.request.headers
        }),
        partialQuery: D({
          field: e.request.query
        })
      },
      mapParams(s, { partialUrl: i, partialBody: l, partialHeaders: y, partialQuery: g }) {
        return {
          url: i(s),
          body: l(s),
          headers: y(s),
          query: g(s)
        };
      },
      effect: a
    })
  );
  const o = {
    ...n,
    __: { ...n.__, executeFx: a }
  };
  return e.concurrency && (console.error(
    "concurrency field in createJsonMutation is deprecated, please use concurrency operator instead: https://ff.effector.dev/adr/concurrency.html"
  ), o.__.meta.flags.concurrencyFieldUsed = !0), e.concurrency && setTimeout(() => {
    var s, i;
    o.__.meta.flags.concurrencyOperatorUsed || Ve(o, {
      strategy: ((s = e.concurrency) == null ? void 0 : s.strategy) ?? "TAKE_EVERY",
      abortAll: (i = e.concurrency) == null ? void 0 : i.abort
    });
  }), o;
}
const ir = [
  "ms",
  "milli",
  "millisecond",
  "milliseconds"
], lr = ["s", "sec", "secs", "second", "seconds"], ur = ["m", "min", "mins", "minute", "minutes"], fr = ["h", "hr", "hrs", "hour", "hours"];
function W(e) {
  if (typeof e == "number")
    return e;
  let t = 0;
  for (const r of e.split(" "))
    switch (!0) {
      case Y(r, ir):
        t += X(r);
        break;
      case Y(r, lr):
        t += X(r) * 1e3;
        break;
      case Y(r, ur):
        t += X(r) * 6e4;
        break;
      case Y(r, fr):
        t += X(r) * 36e5;
        break;
    }
  return t;
}
function Y(e, t) {
  return t.includes(dr(e));
}
function dr(e) {
  return e.replace(/[0-9.]/g, "");
}
function X(e) {
  return e.includes(".") ? parseFloat(e) : parseInt(e);
}
function Mr(e, {
  times: t,
  delay: r,
  filter: n,
  mapParams: a,
  ...o
}) {
  const s = o.supressIntermediateErrors ?? !0, i = R(t), l = x(1, {
    serialize: "ignore",
    name: "ff.$attempt",
    sid: "ff.$attempt"
  }), y = O({
    attempt: l
  }), g = O(
    l,
    i,
    (d, w) => s && d <= w
  ), f = $(), _ = $(), u = D({
    field: n ?? !0
  }), { planNextAttempt: p, __: m } = J(
    c({
      clock: f,
      source: {
        maxAttempts: i,
        attempt: l,
        partialFilter: u
      },
      filter: ({ partialFilter: d }, w) => d(w),
      fn: ({ attempt: d, maxAttempts: w }, { params: P, error: F, meta: I }) => ({
        params: P,
        error: F,
        meta: { ...I, attempt: d, maxAttempts: w }
      })
    }),
    { planNextAttempt: ({ meta: d }) => d.attempt <= d.maxAttempts }
  );
  if (c({
    clock: Te({
      clock: c({
        clock: p,
        source: {
          partialMapper: D({
            field: a ?? (({ params: d }) => d)
          })
        },
        fn: ({ partialMapper: d }, w) => d(w)
      }),
      timeout: O(
        {
          partialTimeout: D({
            field: r
          }),
          meta: y
        },
        ({ partialTimeout: d, meta: w }) => W(d(w))
      )
    }),
    fn: (d) => ({
      params: d,
      meta: { stopErrorPropagation: !1, stale: !0 }
    }),
    target: [_, e.__.lowLevelAPI.startWithMeta]
  }), l.on(_, (d) => d + 1).reset([e.finished.success, e.start]), o.otherwise && c({ clock: m, target: o.otherwise }), s) {
    const d = e.__.lowLevelAPI.dataSourceRetrieverFx.use.getCurrent();
    c({
      clock: e.__.lowLevelAPI.failedIgnoreSuppression,
      target: f
    }), e.__.lowLevelAPI.dataSourceRetrieverFx.use(
      T({
        source: { supressError: g, partialFilter: u },
        mapParams: (w, { supressError: P, partialFilter: F }) => ({
          ...w,
          supressError: P,
          partialFilter: F
        }),
        effect: k(async ({ supressError: w, partialFilter: P, ...F }) => {
          const I = Ie(f, { safe: !0 });
          try {
            return { ...await d(F), stopErrorPropagation: w };
          } catch (N) {
            const C = {
              params: F.params,
              error: N.error,
              meta: F.meta
            };
            throw (
              /*
               * If filter returns false, this fail is not supposed to be retried
               * so we should not suppress this error in any case.
               *
               * If filter returns is true, we should suppress this error only if
               * supressError is true.
               */
              P(C) && w ? (I(C), { error: N.error, stopErrorPropagation: !0 }) : N
            );
          }
        })
      })
    );
  }
  c({ clock: e.finished.failure, target: f });
}
const rt = {
  randomize: { spread: 0 }
};
function zr(e, t = rt) {
  return function({ attempt: r }) {
    return e * r + nt(t);
  };
}
function Cr(e, t = rt) {
  return function({ attempt: r }) {
    return e ** r + nt(t);
  };
}
function nt({ randomize: e }) {
  const { spread: t } = e;
  return kt({ min: -t, max: t });
}
function Br(e, t) {
  const r = T({
    source: R(t.after).map(W),
    effect(n, a) {
      return new Promise((o) => {
        var l;
        const s = setTimeout(() => {
          a.abort(Ft({ timeout: n })), o();
        }, n), i = () => {
          clearTimeout(s), o();
        };
        (l = a.promise) == null || l.then(i, i);
      });
    }
  });
  c({
    clock: e.__.lowLevelAPI.callObjectCreated,
    target: r
  });
}
function Hr(e, {
  on: t,
  by: r
}) {
  const n = pr(e), a = $(), o = $();
  J({
    source: c({
      clock: t.finished.success,
      source: {
        partialRule: D({
          field: r.success
        }),
        queryState: n
      },
      fn: ({ partialRule: l, queryState: y }, g) => l({
        query: y,
        mutation: g
      })
    }),
    match: {
      fillData: (l) => we(l.result)
    },
    cases: { fillData: a, __: o }
  }), r.failure && J({
    source: c({
      clock: t.finished.failure,
      source: {
        partialRule: D({
          field: r.failure
        }),
        queryState: n
      },
      fn: ({ partialRule: l, queryState: y }, g) => l({
        query: y,
        mutation: g
      })
    }),
    match: {
      fillData: (l) => we(l.result)
    },
    cases: { fillData: a, __: o }
  }), c({
    clock: a,
    fn: ({ result: l }) => l,
    target: e.__.lowLevelAPI.pushData
  }), c({
    clock: o,
    fn: ({ error: l }) => l,
    target: e.__.lowLevelAPI.pushError
  });
  const { shouldRefetch: s, __: i } = J(
    ce([a, o]).map(({ refetch: l }) => l),
    {
      shouldRefetch: (l) => !!l
    }
  );
  c({
    clock: s,
    source: n,
    filter: (l, y) => typeof y == "object" && "params" in y || l && "params" in l,
    fn: (l, y) => typeof y == "object" && "params" in y ? { params: y.params, refresh: !0 } : { params: l == null ? void 0 : l.params, refresh: !0 },
    target: [e.__.lowLevelAPI.revalidate, e.$stale.reinit]
  }), c({
    clock: i,
    source: n,
    filter: (l) => l && "params" in l,
    fn: (l) => ({
      params: l.params,
      refresh: !1
    }),
    target: e.__.lowLevelAPI.revalidate
  });
}
function pr(e) {
  return O(
    {
      idle: e.$idle,
      result: e.$data,
      params: e.__.$latestParams,
      error: e.$error,
      failed: e.$failed
    },
    ({ idle: t, result: r, params: n, error: a, failed: o }) => r == null && a == null ? null : t ? { result: r } : o ? { error: a, params: n } : { result: r, params: n }
  );
}
function Ur(e, t) {
  var a;
  console.error(
    "attachOperation is deprecated since 0.12, please read the migration guide: https://ff.effector.dev/adr/attach_operation_deprecation.html"
  );
  const { source: r, mapParams: n } = t ?? {};
  return (a = e.__.experimentalAPI) == null ? void 0 : a.attach({
    source: r ?? x(null, {
      serialize: "ignore"
    }),
    mapParams: n ?? ((o) => o)
  });
}
function he(e) {
  const t = x(e, {
    serialize: "ignore",
    name: "ff.$cacheInstance",
    sid: "ff.$cacheInstance"
  });
  return { ...e, __: { $instance: t } };
}
function at({
  adapter: e,
  options: t,
  events: r
}) {
  t != null && t.hit && c({
    clock: e.get.done,
    filter: ({ result: n }) => n !== null,
    fn: ({ params: n }) => ({ key: n.key }),
    target: t.hit
  }), t != null && t.miss && c({
    clock: e.get.done,
    filter: ({ result: n }) => n === null,
    fn: ({ params: n }) => ({ key: n.key }),
    target: t.miss
  }), t != null && t.expired && (r != null && r.itemExpired) && c({
    clock: r.itemExpired,
    fn: ({ key: n }) => ({ key: n }),
    target: t.expired
  }), t != null && t.evicted && (r != null && r.itemEvicted) && c({
    clock: r.itemEvicted,
    target: t.evicted
  });
}
function mr(e) {
  const { maxEntries: t, maxAge: r, observability: n } = e ?? {};
  let a = {};
  const o = $(), s = $(), i = $(), l = $(), y = $();
  y.watch(() => {
    a = {};
  });
  const g = $t({ clock: o }), f = c({
    clock: o,
    source: { now: g },
    fn: ({ now: u }, { key: p, value: m }) => hr(
      a,
      { key: p, entry: { value: m, cachedAt: u } },
      t
    )
  });
  if (f.watch(({ next: u }) => {
    a = u;
  }), c({
    clock: f,
    filter: ({ evicted: u }) => !!u,
    fn: ({ evicted: u }) => ({ key: u }),
    target: l
  }), s.watch(({ key: u }) => {
    const { [u]: p, ...m } = a;
    a = m;
  }), r) {
    const u = W(r);
    o.watch((p) => {
      const m = Ie(i, { safe: !0 });
      setTimeout(() => m(p), u);
    }), c({
      clock: i,
      fn: ({ key: p }) => ({ key: p }),
      target: s
    });
  }
  const _ = {
    get: k(({ key: u }) => {
      const p = a[u] ?? null;
      if (!p)
        return null;
      if (r) {
        const m = (p == null ? void 0 : p.cachedAt) + W(r);
        if (Date.now() >= m)
          return s({ key: u }), null;
      }
      return p;
    }),
    set: k(o),
    unset: k(s),
    purge: y
  };
  return at({
    adapter: _,
    options: n,
    events: { itemExpired: i, itemEvicted: l }
  }), he(_);
}
function hr(e, { key: t, entry: r }, n) {
  if (n === void 0)
    return { next: { ...e, [t]: r }, evicted: null };
  const a = Object.keys(e);
  if (a.length < n)
    return { next: { ...e, [t]: r }, evicted: null };
  const [o] = a, { [o]: s, ...i } = e;
  return { next: { ...i, [t]: r }, evicted: o };
}
function yr(e) {
  let t = 0, r = 0;
  for (; r < e.length; )
    t = (t << 5) - t + e.charCodeAt(r++) | 0;
  return t.toString(36);
}
function st(e) {
  const t = /* @__PURE__ */ new Set();
  function r(n) {
    if (n === void 0)
      return;
    if (n === null)
      return "null";
    if (typeof n == "number")
      return isFinite(n) ? `${n}` : "null";
    if (typeof n == "function")
      throw new TypeError("Can't serialize function");
    if (typeof n != "object")
      return JSON.stringify(n);
    if (t.has(n))
      throw new TypeError("Can't serialize cyclic structure");
    if (t.add(n), Array.isArray(n)) {
      const o = n.map((s) => r(s) || "null").join(",");
      return t.delete(n), `[${o}]`;
    }
    const a = Object.keys(n).sort().map((o) => {
      const s = r(n[o]);
      return s ? `${r(o)}:${s}` : "";
    }).filter(Boolean).join(",");
    return t.delete(n), `{${a}}`;
  }
  return r(e);
}
function gr({
  sid: e,
  params: t = null,
  sources: r
}) {
  try {
    return st({ params: t, sources: r, sid: e }) ?? null;
  } catch {
    return null;
  }
}
function $r({
  sid: e,
  params: t = null,
  sources: r
}) {
  try {
    const n = st({ params: t, sources: r, sid: e });
    return yr(n);
  } catch {
    return null;
  }
}
function _r(e) {
  const t = kr(e);
  if (t)
    return t;
  const r = vr(e);
  if (r)
    return r;
  throw new Error(
    "Query does not have sid or uniq name, which is required for caching, read more https://effector.dev/en/explanation/sids/"
  );
}
function kr(e) {
  return e.__.meta.sid ?? null;
}
const Pe = /* @__PURE__ */ new Set();
function vr(e) {
  const t = e.__.meta.name;
  return Pe.has(t) ? null : (Pe.add(t), t);
}
function Vr(e, t) {
  const {
    adapter: r,
    staleAfter: n,
    purge: a,
    humanReadableKeys: o
  } = {
    adapter: (t == null ? void 0 : t.adapter) ?? mr(),
    humanReadableKeys: !1,
    ...t
  }, s = o ? gr : $r, i = _r(e), l = e.__.lowLevelAPI.sourced.map(ht), y = k(async (p) => Promise.all(l.map((m) => m(p)))), g = k(async ({ instance: p, params: m }) => {
    const d = await y(m), w = s({
      sid: i,
      params: e.__.lowLevelAPI.paramsAreMeaningless ? null : m,
      sources: d
    });
    w && await p.unset({ key: w });
  }), f = k(async ({ instance: p, params: m, result: d }) => {
    const w = await y(m), P = s({
      sid: i,
      params: e.__.lowLevelAPI.paramsAreMeaningless ? null : m,
      sources: w
    });
    P && await p.set({ key: P, value: d });
  }), _ = k(async ({ params: p, instance: m }) => {
    const d = await y(p), w = s({
      sid: i,
      params: e.__.lowLevelAPI.paramsAreMeaningless ? null : p,
      sources: d
    });
    if (!w)
      return null;
    const P = await m.get({ key: w });
    if (!P)
      return null;
    const F = n ? P.cachedAt + W(n) <= Date.now() : !0;
    return { result: P.value, stale: F };
  }), u = {
    name: "cache",
    get: T({
      source: {
        instance: r.__.$instance
      },
      mapParams: ({ params: p }, { instance: m }) => ({
        params: p,
        instance: m
      }),
      effect: _
    }),
    set: T({
      source: {
        instance: r.__.$instance
      },
      mapParams: ({ params: p, result: m }, { instance: d }) => ({
        params: p,
        result: m,
        instance: d
      }),
      effect: f
    }),
    unset: T({
      source: {
        instance: r.__.$instance
      },
      mapParams: ({ params: p }, { instance: m }) => ({
        instance: m,
        params: p
      }),
      effect: g
    })
  };
  e.__.lowLevelAPI.dataSources.unshift(u), a && c({
    clock: a,
    source: { instance: r.__.$instance },
    target: k(
      ({ instance: p }) => p.purge()
    )
  });
}
const oe = "__farfetched_meta__";
function ot(e) {
  const { storage: t, observability: r, maxAge: n, maxEntries: a, serialize: o } = e;
  function s() {
    const d = k(async (b) => {
      const S = await p(b);
      if (!S)
        return null;
      try {
        const A = JSON.parse(S);
        return {
          ...A,
          value: o != null && o.read ? o.read(A.value) : A.value
        };
      } catch {
        return null;
      }
    }), w = k(
      async ({ key: b, value: S }) => {
        const A = JSON.stringify({
          value: o != null && o.write ? o.write(S) : S,
          timestamp: Date.now()
        });
        _.addKey({ key: b }), await u({ key: b, value: A });
      }
    ), P = k(async (b) => {
      _.removeKey({ key: b }), await m(b);
    }), F = $(), I = $(), N = $(), C = k(
      async (b) => Promise.all(b.map(P))
    );
    c({
      clock: N,
      source: _.$meta,
      fn: (b) => (b == null ? void 0 : b.keys) ?? [],
      target: C
    }), n && c({
      clock: Te({
        clock: c({
          clock: u.done,
          filter: ({ params: b }) => b.key !== oe
        }),
        timeout: W(n)
      }),
      fn: U("params"),
      target: [F, P]
    });
    const B = {
      get: k(async ({ key: b }) => {
        const S = await d(b);
        if (!S)
          return null;
        if (n) {
          const A = (S == null ? void 0 : S.timestamp) + W(n);
          if (Date.now() >= A)
            return F({ key: b, value: S.value }), await P(b), null;
        }
        return { value: S.value, cachedAt: S.timestamp };
      }),
      set: k(
        async ({ key: b, value: S }) => {
          var V, K;
          const A = await l(), M = ((V = A == null ? void 0 : A.keys) == null ? void 0 : V.length) ?? 0;
          if (a && M >= a) {
            const L = (K = A == null ? void 0 : A.keys) == null ? void 0 : K.slice(0, M - a + 1);
            for (const G of L ?? [])
              I({ key: G }), await P(G);
          }
          await w({ key: b, value: S });
        }
      ),
      unset: k(async ({ key: b }) => {
        await P(b);
      }),
      purge: N
    };
    return at({
      adapter: B,
      options: r,
      events: { itemExpired: F, itemEvicted: I }
    }), he(B);
  }
  const i = x(null, {
    serialize: "ignore",
    name: "ff.browserStorage.$meta",
    sid: "ff.browserStorage.$meta"
  }), l = k(async () => {
    const d = await p(oe);
    if (!d)
      return null;
    try {
      return JSON.parse(d);
    } catch {
      return null;
    }
  }), y = k(
    (d) => u({ key: oe, value: JSON.stringify(d) })
  ), g = $(), f = $(), _ = {
    $meta: i,
    addKey: g,
    removeKey: f
  };
  c({ clock: l.doneData, target: i }), c({ clock: i, filter: Boolean, target: y }), c({
    clock: g,
    source: i,
    fn: (d, { key: w }) => {
      const P = (d == null ? void 0 : d.keys) ?? [];
      return P.includes(w) ? d : { ...d, keys: [...P, w] };
    },
    target: i
  }), c({
    clock: f,
    source: i,
    fn: (d, { key: w }) => {
      var P;
      return {
        ...d,
        keys: ((P = d == null ? void 0 : d.keys) == null ? void 0 : P.filter((F) => F !== w)) ?? []
      };
    },
    target: i
  });
  const u = k((d) => {
    t().setItem(d.key, d.value);
  }), p = k((d) => t().getItem(d)), m = k((d) => t().removeItem(d));
  return s();
}
function Kr(e) {
  return ot({ storage: () => localStorage, ...e });
}
function Qr(e) {
  return ot({
    storage: () => sessionStorage,
    ...e
  });
}
function Jr() {
  return he({
    get: k(() => null),
    set: k(() => {
    }),
    purge: $(),
    unset: k()
  });
}
function Wr() {
  return $();
}
function Gr(e, t) {
  const r = [], [n, a] = At(t.triggers ?? [], q.event);
  r.push(...n);
  const o = [e.$enabled];
  t.enabled !== void 0 && o.push(t.enabled);
  const s = le({
    predicate: Boolean,
    stores: o
  });
  if (a.length > 0) {
    const l = a.map(
      (_) => _["@@trigger"]()
    ), y = x(!1, {
      serialize: "ignore",
      name: "ff.$alreadySetup",
      sid: "ff.$alreadySetup"
    }), { setup: g, teardown: f } = mt(y, {
      setup: () => !0,
      teardown: () => !1
    });
    c({
      clock: [
        e.finished.success,
        c({ clock: s.updates, filter: s })
      ],
      filter: H(y),
      target: [...l.map(U("setup")), g]
    }), c({
      clock: s.updates,
      filter: De(y, H(s)),
      target: [...l.map(U("teardown")), f]
    }), r.push(...l.map(U("fired")));
  }
  if (t.automatically) {
    const l = e.finished.finally.map(U("params")), y = x([], {
      serialize: "ignore",
      name: "ff.$previousSources",
      sid: "ff.$previousSources"
    }), g = O(
      e.__.lowLevelAPI.sourced.map(
        (_) => D({ field: _ })
      )
    );
    c({
      clock: l,
      source: g,
      fn: (_, u) => _.map((p) => p(u)),
      filter: s,
      target: y
    });
    const f = x(null, {
      serialize: "ignore",
      name: "ff.$nextSources",
      sid: "ff.$nextSources"
    });
    c({
      // @ts-expect-error I have no idea
      clock: e.__.lowLevelAPI.sourced.map(yt).filter(q.store),
      source: {
        latestParams: e.__.$latestParams,
        partialSources: g
      },
      filter: H(e.$idle),
      fn: ({ latestParams: _, partialSources: u }) => u.map((p) => p(_)),
      target: f
    }), r.push(
      c({
        clock: [f.updates, s.updates.filter({ fn: Boolean })],
        source: [f, y],
        filter: ([_, u]) => !Z(_, u)
      })
    );
  }
  const i = c({ clock: r, filter: s });
  c({
    clock: i,
    filter: H(e.$idle),
    fn: () => !0,
    target: e.$stale
  }), c({
    clock: _t(i),
    source: e.__.$latestParams,
    filter: H(e.$idle),
    target: e.refresh
  });
}
function br(e, { barrier: t }) {
  if (Array.isArray(e))
    for (const r of e)
      br(r, { barrier: t });
  else {
    c({ clock: e.started, target: t.__.touch }), c({
      clock: e.finished.failure,
      target: t.__.operationFailed
    }), c({
      clock: t.activated,
      filter: e.$failed,
      source: e.__.$latestParams,
      fn: (n) => ({
        params: n,
        meta: { stopErrorPropagation: !1, stale: !1 }
      }),
      target: e.__.lowLevelAPI.startWithMeta
    });
    const r = T({
      source: { mutex: t.__.$mutex, active: t.$active },
      async effect({ mutex: n }) {
        return await (n == null ? void 0 : n.waitForUnlock()), null;
      }
    });
    e.__.lowLevelAPI.dataSources.unshift({
      name: "barrier_blocker",
      get: r
    });
  }
}
function Yr({
  active: e,
  perform: t,
  activateOn: r,
  deactivateOn: n
}) {
  const a = $(), o = $(), s = $(), i = $(), l = $(), y = $(), g = Ar(t ?? []), f = $(), _ = x(!1, {
    sid: "barrier.$forceDeactivated",
    name: "barrier.$forceDeactivated",
    serialize: "ignore"
  }).on(f, () => !0);
  let u;
  if (e)
    u = e;
  else if (q.event(r) && q.event(n))
    u = x(!1, {
      sid: "barrier.$active",
      name: "barrier.$active"
    }).on(r, () => !0).on(n, () => !1);
  else if (r) {
    if (u = x(!1, {
      sid: "barrier.$active",
      name: "barrier.$active"
    }), "failure" in r && r.failure) {
      const d = r.failure;
      c({
        clock: l,
        filter: ({ error: w, params: P }) => d({ error: w, params: P }),
        fn: () => !0,
        target: [u, i]
      }), c({
        clock: Pt({
          events: g.map(U("end")),
          reset: l
        }),
        fn: () => !1,
        target: u
      });
    }
  } else
    throw new Error("Invalid configuration of createBarrier");
  const p = O(
    u,
    _,
    (d, w) => w ? !1 : d
  );
  c({
    clock: g.map(U("end")),
    target: s
  }), J({
    clock: [p, i],
    source: p,
    match: { activated: Boolean },
    cases: { activated: a, __: o }
  }), c({
    clock: i,
    filter: p,
    target: wr(g)
  });
  const m = x(null, { serialize: "ignore" });
  return c({
    clock: i,
    source: { mutex: m, active: p },
    filter: ({ mutex: d }) => d === null,
    fn: ({ active: d }) => new Et(d),
    target: m
  }), c({
    clock: a,
    target: T({
      source: m,
      effect(d) {
        d == null || d.acquire();
      }
    })
  }), c({
    clock: o,
    target: T({
      source: m,
      effect(d) {
        d == null || d.release();
      }
    })
  }), {
    $active: E(p),
    activated: E(a),
    deactivated: E(o),
    performed: E(s),
    forceDeactivate: f,
    __: { touch: i, operationFailed: l, operationDone: y, $mutex: m }
  };
}
function wr(e) {
  const t = $();
  for (const { start: r, $pending: n } of e)
    c({
      clock: t,
      filter: H(n),
      target: r
    });
  return t;
}
function Ar(e) {
  return e.map((t) => {
    if (xr(t))
      return {
        start: t.start,
        end: Se(
          c({
            clock: [t.finished.success, t.finished.skip]
          })
        ),
        $pending: t.$pending
      };
    if (Er(t))
      return {
        start: t,
        end: Se(t.done),
        $pending: t.pending
      };
    {
      const r = x(!1, { serialize: "ignore" }).on(t.start, () => !0).on(t.end, () => !1);
      return { ...t, $pending: r };
    }
  });
}
function xr(e) {
  return Ge(e) || cr(e);
}
function Er(e) {
  return q.effect(e);
}
function Se(e) {
  return c({
    clock: e,
    fn: () => {
    }
  });
}
export {
  ne as Meta,
  Gt as NodeLinksSumbol,
  Wt as NodeMetaSumbol,
  ae as Result,
  Ce as abortError,
  br as applyBarrier,
  Ur as attachOperation,
  Vr as cache,
  Sr as combineSourced,
  Ve as concurrency,
  ie as configurationError,
  Rr as connectQuery,
  Yr as createBarrier,
  he as createCacheAdapter,
  me as createHeadlessMutation,
  de as createHeadlessQuery,
  Lr as createJsonMutation,
  qr as createJsonQuery,
  Nr as createMutation,
  Or as createQuery,
  Wr as declareParams,
  Cr as exponentialDelay,
  nr as fetchFx,
  Be as httpError,
  mr as inMemoryCache,
  ee as invalidDataError,
  He as isHttpError,
  Dr as isHttpErrorCode,
  Fr as isInvalidDataError,
  jr as isNetworkError,
  Tr as isPreparationError,
  Ir as isTimeoutError,
  Gr as keepFresh,
  zr as linearDelay,
  Kr as localStorageCache,
  Tt as networkError,
  D as normalizeSourced,
  Ue as onAbort,
  It as preparationError,
  Mr as retry,
  Qr as sessionStorageCache,
  Br as timeout,
  Ft as timeoutError,
  re as unknownContract,
  Hr as update,
  Jr as voidCache
};
