// @gravity-ui/table/build/cjs/components/BaseTable/BaseTable.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseTable = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const React = __importStar(require("react"));
const utils_1 = require("../../utils/index.js");
const BaseDraggableRow_1 = require("../BaseDraggableRow/index.js");
const BaseFooterRow_1 = require("../BaseFooterRow/index.js");
const BaseHeaderRow_1 = require("../BaseHeaderRow/index.js");
const BaseRow_1 = require("../BaseRow/index.js");
const LastSelectedRowContext_1 = require("../LastSelectedRowContext/index.js");
const SortableListContext_1 = require("../SortableListContext/index.js");
const BaseTable_classname_1 = require("./BaseTable.classname.js");
require("./BaseTable.css");
exports.BaseTable = React.forwardRef(({ table, attributes, bodyAttributes, bodyClassName, bodyRef, cellAttributes, cellClassName, className, customFooterRowCount, emptyContent, footerAttributes, footerCellAttributes, footerCellClassName, footerClassName, footerRowAttributes, footerRowClassName, getGroupTitle, getIsCustomRow, getIsGroupHeaderRow, groupHeaderClassName, headerAttributes, headerCellAttributes, headerCellClassName, headerClassName, headerRowAttributes, headerRowClassName, onRowClick, renderCustomFooterContent, renderCustomRowContent, renderGroupHeader, renderGroupHeaderRowContent, renderHeaderCellContent, renderResizeHandle, renderSortIndicator, resizeHandleClassName, rowAttributes, rowClassName, rowVirtualizer, sortIndicatorClassName, stickyFooter = false, stickyHeader = false, withFooter = false, withHeader = true, }, ref) => {
    const draggableContext = React.useContext(SortableListContext_1.SortableListContext);
    const draggingRowIndex = draggableContext?.activeItemIndex ?? -1;
    const { rows, rowsById } = table.getRowModel();
    const ariaRowIndexMap = React.useMemo(() => {
        return (0, utils_1.getAriaRowIndexMap)(rows);
    }, [rows]);
    const headerGroups = withHeader ? table.getHeaderGroups() : [];
    const footerGroups = withFooter ? table.getFooterGroups() : [];
    const colCount = table.getVisibleLeafColumns().length;
    const headerRowCount = headerGroups.length;
    const bodyRowCount = Object.keys(rowsById).length;
    const footerRowCount = (withFooter &&
        ((renderCustomFooterContent && customFooterRowCount) || footerGroups.length)) ||
        0;
    const rowCount = bodyRowCount + headerRowCount + footerRowCount;
    const bodyRows = rowVirtualizer?.getVirtualItems() || rows;
    const renderEmptyContent = () => {
        if (!emptyContent) {
            return null;
        }
        const emptyRowClassName = typeof rowClassName === 'function' ? rowClassName() : rowClassName;
        const emptyCellClassName = typeof cellClassName === 'function' ? cellClassName() : cellClassName;
        return ((0, jsx_runtime_1.jsx)("tr", { className: (0, BaseTable_classname_1.b)('row', { empty: true }, emptyRowClassName), children: (0, jsx_runtime_1.jsx)("td", { className: (0, BaseTable_classname_1.b)('cell', {}, emptyCellClassName), colSpan: colCount, style: {
                    width: rowVirtualizer ? table.getTotalSize() : undefined,
                }, children: typeof emptyContent === 'function' ? emptyContent() : emptyContent }) }));
    };
    const renderBodyRows = () => {
        return bodyRows.map((virtualItemOrRow, index) => {
            const row = rowVirtualizer
                ? rows[virtualItemOrRow.index]
                : virtualItemOrRow;
            const rowIndex = rowVirtualizer ? virtualItemOrRow.index : index;
            const style = row.depth > 0
                ? {
                    '--_--tree-depth': row.depth,
                    '--_--last-nested': rows[rowIndex + 1]?.depth === 0 ? 1 : 0,
                }
                : undefined;
            const virtualItem = rowVirtualizer ? virtualItemOrRow : undefined;
            const key = virtualItem?.key ?? row.id;
            const rowProps = {
                cellClassName,
                className: rowClassName,
                getGroupTitle,
                getIsCustomRow,
                getIsGroupHeaderRow,
                groupHeaderClassName,
                attributes: rowAttributes,
                cellAttributes,
                onClick: onRowClick,
                renderCustomRowContent,
                renderGroupHeader,
                renderGroupHeaderRowContent,
                row,
                rowVirtualizer,
                table,
                virtualItem,
                style,
                'aria-rowindex': headerRowCount + ariaRowIndexMap[row.id],
                'aria-selected': table.options.enableRowSelection
                    ? row.getIsSelected()
                    : undefined,
            };
            if (draggableContext) {
                return (0, jsx_runtime_1.jsx)(BaseDraggableRow_1.BaseDraggableRow, { ...rowProps }, key);
            }
            return (0, jsx_runtime_1.jsx)(BaseRow_1.BaseRow, { ...rowProps }, key);
        });
    };
    return ((0, jsx_runtime_1.jsx)(LastSelectedRowContext_1.LastSelectedRowContextProvider, { children: (0, jsx_runtime_1.jsxs)("table", { ref: ref, className: (0, BaseTable_classname_1.b)({ 'with-row-virtualization': Boolean(rowVirtualizer) }, className), "data-dragging-row-index": draggingRowIndex > -1 ? draggingRowIndex : undefined, "aria-colcount": colCount > 0 ? colCount : undefined, "aria-rowcount": rowCount > 0 ? rowCount : undefined, "aria-multiselectable": (0, utils_1.getAriaMultiselectable)(table), ...attributes, children: [withHeader && ((0, jsx_runtime_1.jsx)("thead", { className: (0, BaseTable_classname_1.b)('header', { sticky: stickyHeader }, headerClassName), ...headerAttributes, children: headerGroups.map((headerGroup, index) => ((0, jsx_runtime_1.jsx)(BaseHeaderRow_1.BaseHeaderRow, { cellClassName: headerCellClassName, className: headerRowClassName, headerGroup: headerGroup, parentHeaderGroup: headerGroups[index - 1], renderHeaderCellContent: renderHeaderCellContent, renderResizeHandle: renderResizeHandle, renderSortIndicator: renderSortIndicator, resizeHandleClassName: resizeHandleClassName, sortIndicatorClassName: sortIndicatorClassName, attributes: headerRowAttributes, cellAttributes: headerCellAttributes, "aria-rowindex": index + 1 }, headerGroup.id))) })), (0, jsx_runtime_1.jsx)("tbody", { ref: bodyRef, className: (0, BaseTable_classname_1.b)('body', bodyClassName), ...bodyAttributes, style: {
                        height: bodyRows.length ? rowVirtualizer?.getTotalSize() : undefined,
                        ...bodyAttributes?.style,
                    }, children: bodyRows.length ? renderBodyRows() : renderEmptyContent() }), withFooter && ((0, jsx_runtime_1.jsx)("tfoot", { className: (0, BaseTable_classname_1.b)('footer', { sticky: stickyFooter }, footerClassName), ...footerAttributes, children: renderCustomFooterContent
                        ? renderCustomFooterContent({
                            cellClassName: (0, BaseTable_classname_1.b)('footer-cell'),
                            footerGroups,
                            rowClassName: (0, BaseTable_classname_1.b)('footer-row'),
                            rowIndex: headerRowCount + bodyRowCount + 1,
                        })
                        : footerGroups.map((footerGroup, index) => (0, utils_1.shouldRenderFooterRow)(footerGroup) ? ((0, jsx_runtime_1.jsx)(BaseFooterRow_1.BaseFooterRow, { footerGroup: footerGroup, attributes: footerRowAttributes, cellAttributes: footerCellAttributes, cellClassName: footerCellClassName, className: footerRowClassName, "aria-rowindex": headerRowCount + bodyRowCount + index + 1 }, footerGroup.id)) : null) }))] }) }));
});
exports.BaseTable.displayName = 'BaseTable';
