// monaco-yql-languages/build/yql/autocomplete/generateSuggestions.js
import * as monaco from '../../fillers/monaco-editor-core';
import { getSuggestionIndex, isVariable, removeBackticks, removeStringDuplicates, wrapStringToBackticks, } from './utils';
import { suggestionIndexToWeight } from '../../utils';
export async function generateSimpleTypesSuggestion(rangeToInsertSuggestion, simpleTypes = []) {
    return simpleTypes.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.TypeParameter,
        detail: 'Type',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestSimpleTypes')),
    }));
}
export async function generateEntitiesSuggestion(rangeToInsertSuggestion, fetchedEntities, prefix = '') {
    const withBackticks = prefix === null || prefix === void 0 ? void 0 : prefix.startsWith('`');
    return fetchedEntities.reduce((acc, { value, detail, isDir }) => {
        const label = isDir ? `${value}/` : value;
        let labelAsSnippet;
        if (isDir && !withBackticks) {
            labelAsSnippet = `\`${label}$0\``;
        }
        const suggestionIndex = acc.length;
        acc.push({
            label,
            insertText: labelAsSnippet !== null && labelAsSnippet !== void 0 ? labelAsSnippet : label,
            kind: isDir
                ? monaco.languages.CompletionItemKind.Folder
                : monaco.languages.CompletionItemKind.Text,
            insertTextRules: labelAsSnippet
                ? monaco.languages.CompletionItemInsertTextRule.InsertAsSnippet
                : monaco.languages.CompletionItemInsertTextRule.None,
            detail: detail,
            range: rangeToInsertSuggestion,
            command: label.endsWith('/')
                ? { id: 'editor.action.triggerSuggest', title: '' }
                : undefined,
            // first argument is responsible for sorting groups of suggestions, the second - to preserve suggestions order returned from backend
            sortText: suggestionIndexToWeight(getSuggestionIndex('suggestEntity')) +
                suggestionIndexToWeight(suggestionIndex),
        });
        return acc;
    }, []);
}
export function generateKeywordsSuggestion(rangeToInsertSuggestion, suggestKeywords = []) {
    return suggestKeywords.map((keywordSuggestion) => ({
        label: keywordSuggestion.value,
        insertText: keywordSuggestion.value,
        kind: monaco.languages.CompletionItemKind.Keyword,
        detail: 'Keyword',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestKeywords')),
    }));
}
export function generateVariableSuggestion(rangeToInsertSuggestion, suggestVariables = []) {
    return suggestVariables.map(({ name }) => {
        const variable = '$' + name;
        return {
            label: variable,
            insertText: variable,
            kind: monaco.languages.CompletionItemKind.Variable,
            detail: 'Variable',
            range: rangeToInsertSuggestion,
            sortText: suggestionIndexToWeight(getSuggestionIndex('suggestVariables')),
        };
    });
}
export async function generateSimpleFunctionsSuggestion(rangeToInsertSuggestion, functions = []) {
    return functions.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.Function,
        detail: 'Function',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestFunctions')),
    }));
}
export async function generateAggregateFunctionsSuggestion(rangeToInsertSuggestion, aggreagteFunctions = []) {
    return aggreagteFunctions.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.Function,
        detail: 'Aggregate function',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestAggregateFunctions')),
    }));
}
export async function generateWindowFunctionsSuggestion(rangeToInsertSuggestion, windowFunctions = []) {
    return windowFunctions.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.Function,
        detail: 'Window function',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestWindowFunctions')),
    }));
}
export async function generateTableFunctionsSuggestion(rangeToInsertSuggestion, tableFunctions = []) {
    return tableFunctions.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.Function,
        detail: 'Table function',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestTableFunctions')),
    }));
}
export async function generateUdfSuggestion(rangeToInsertSuggestion, udfs = []) {
    return udfs.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.Function,
        detail: 'UDF',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestUdfs')),
    }));
}
export async function generatePragmasSuggestion(rangeToInsertSuggestion, pragmas = []) {
    return pragmas.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.Module,
        detail: 'Pragma',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestPragmas')),
    }));
}
export async function generateEntitySettingsSuggestion(rangeToInsertSuggestion, entitySettings = []) {
    return entitySettings.map((el) => ({
        label: el,
        insertText: el,
        kind: monaco.languages.CompletionItemKind.Property,
        detail: 'Setting',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestEntitySettings')),
    }));
}
export function generateColumnAliasesSuggestion(rangeToInsertSuggestion, suggestColumnAliases = []) {
    return suggestColumnAliases.map((columnAliasSuggestion) => ({
        label: columnAliasSuggestion.name,
        insertText: columnAliasSuggestion.name,
        kind: monaco.languages.CompletionItemKind.Variable,
        detail: 'Column alias',
        range: rangeToInsertSuggestion,
        sortText: suggestionIndexToWeight(getSuggestionIndex('suggestColumnAliases')),
    }));
}
export async function generateColumnsSuggestion(rangeToInsertSuggestion, suggestColumns, suggestVariables, fetchedColumns) {
    var _a;
    if (!(suggestColumns === null || suggestColumns === void 0 ? void 0 : suggestColumns.tables)) {
        return [];
    }
    const suggestions = [];
    const allColumnsSuggestion = suggestColumns.all ? [] : undefined;
    const multi = suggestColumns.tables.length > 1;
    const tableNames = suggestColumns.tables.map((entity) => entity.name);
    const filteredTableNames = removeStringDuplicates(tableNames);
    const variableSources = filteredTableNames.filter(isVariable);
    const columnsFromVariable = [];
    if (variableSources.length) {
        variableSources.forEach((source) => {
            var _a, _b, _c, _d;
            const newColumns = (_d = (_c = (_b = (_a = suggestVariables === null || suggestVariables === void 0 ? void 0 : suggestVariables
            // Variable name from suggestions doesn't include $ sign
            .find((variable) => source.slice(1) === variable.name)) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.columns) === null || _c === void 0 ? void 0 : _c.map((col) => ({
                name: col,
                parent: source,
            }))) !== null && _d !== void 0 ? _d : [];
            columnsFromVariable.push(...newColumns);
        });
    }
    const predefinedColumns = suggestColumns.tables.reduce((acc, entity) => {
        const columns = entity.columns;
        if (columns) {
            acc.push(...columns.map((col) => ({
                name: col,
                parent: entity.name,
            })));
        }
        return acc;
    }, []);
    const tableNameToAliasMap = (_a = suggestColumns.tables) === null || _a === void 0 ? void 0 : _a.reduce((acc, entity) => {
        var _a;
        const name = removeBackticks(entity.name);
        const aliases = (_a = acc[name]) !== null && _a !== void 0 ? _a : [];
        if (entity.alias) {
            aliases.push(entity.alias);
        }
        acc[name] = aliases;
        return acc;
    }, {});
    [...fetchedColumns, ...columnsFromVariable, ...predefinedColumns].forEach((col) => {
        const normalizedName = wrapStringToBackticks(col.name);
        const aliases = tableNameToAliasMap[removeBackticks(col.parent)];
        const currentSuggestionIndex = suggestions.length;
        if (aliases === null || aliases === void 0 ? void 0 : aliases.length) {
            aliases.forEach((a) => {
                const columnNameSuggestion = `${a}.${normalizedName}`;
                suggestions.push({
                    label: { label: columnNameSuggestion, description: col.detail },
                    insertText: columnNameSuggestion,
                    kind: monaco.languages.CompletionItemKind.Variable,
                    detail: 'Column',
                    range: rangeToInsertSuggestion,
                    sortText: suggestionIndexToWeight(getSuggestionIndex('suggestColumns')) +
                        suggestionIndexToWeight(currentSuggestionIndex),
                });
                allColumnsSuggestion === null || allColumnsSuggestion === void 0 ? void 0 : allColumnsSuggestion.push(columnNameSuggestion);
            });
        }
        else {
            let columnNameSuggestion = normalizedName;
            if (multi) {
                columnNameSuggestion = `${wrapStringToBackticks(col.parent)}.${normalizedName}`;
            }
            suggestions.push({
                label: {
                    label: columnNameSuggestion,
                    description: col.detail,
                },
                insertText: columnNameSuggestion,
                kind: monaco.languages.CompletionItemKind.Variable,
                detail: 'Column',
                range: rangeToInsertSuggestion,
                sortText: suggestionIndexToWeight(getSuggestionIndex('suggestColumns')) +
                    suggestionIndexToWeight(currentSuggestionIndex),
            });
            allColumnsSuggestion === null || allColumnsSuggestion === void 0 ? void 0 : allColumnsSuggestion.push(columnNameSuggestion);
        }
    });
    if (allColumnsSuggestion && allColumnsSuggestion.length > 1) {
        const allColumns = allColumnsSuggestion.join(', ');
        suggestions.push({
            label: allColumns,
            insertText: allColumns,
            kind: monaco.languages.CompletionItemKind.Variable,
            range: rangeToInsertSuggestion,
            sortText: suggestionIndexToWeight(getSuggestionIndex('suggestAllColumns')),
        });
    }
    return suggestions;
}
