// monaco-yql-languages/build/yql/autocomplete/utils.js
const SuggestionsWeight = {
    suggestTemplates: 0,
    suggestPragmas: 1,
    suggestEntity: 2,
    suggestAllColumns: 3,
    suggestColumns: 4,
    suggestColumnAliases: 5,
    suggestVariables: 6,
    suggestTableIndexes: 7,
    suggestTableHints: 8,
    suggestEntitySettings: 9,
    suggestKeywords: 10,
    suggestAggregateFunctions: 11,
    suggestTableFunctions: 12,
    suggestWindowFunctions: 13,
    suggestFunctions: 14,
    suggestSimpleTypes: 15,
    suggestUdfs: 16,
};
export function getSuggestionIndex(suggestionType) {
    return SuggestionsWeight[suggestionType];
}
export function isVariable(value) {
    return value.startsWith('$');
}
export function removeStringDuplicates(value = []) {
    return Array.from(new Set(value));
}
export function getEntitiesToFetchColumns(suggestColumns) {
    var _a;
    const names = (_a = suggestColumns === null || suggestColumns === void 0 ? void 0 : suggestColumns.tables) === null || _a === void 0 ? void 0 : _a.map((entity) => entity.name);
    // remove duplicates if any
    const filteredTableNames = removeStringDuplicates(names);
    return filteredTableNames.filter((name) => !isVariable(name));
}
const specialSymbols = /[\s'"-/@]/;
export function wrapStringToBackticks(value) {
    if (value.startsWith('`') && value.endsWith('`')) {
        return value;
    }
    let result = value;
    if (value.match(specialSymbols)) {
        result = `\`${value}\``;
    }
    return result;
}
export function removeBackticks(value) {
    let sliceStart = 0;
    let sliceEnd = value.length;
    if (value.startsWith('`')) {
        sliceStart = 1;
    }
    if (value.endsWith('`')) {
        sliceEnd = -1;
    }
    return value.slice(sliceStart, sliceEnd);
}
