// @gravity-ui/date-components/dist/esm/components/RangeDateSelection/components/SelectionControl/SelectionControl.js
import { __rest } from "tslib";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { useMove } from "../../../../hooks/useMove.js";
import { block } from "../../../../utils/cn.js";
import { i18n } from "../../i18n/index.js";
import { lerp } from "../../utils/span.js";
import { useViewportDimensions, useViewportInterval } from "../Ruler/Ruler.js";
import "./SelectionControl.css";
const b = block('range-date-selection-control');
const MIN_SELECTION_WIDTH = 12; // px;
const MIN_DRAG_SCALE = 0.01;
const MAX_DRAG_SCALE = 100;
const DRAG_SCALE_STRENGTH = 20;
export function SelectionControl(_a) {
    var _b;
    var { state, style, className } = _a, props = __rest(_a, ["state", "style", "className"]);
    const { start: startDate, end: endDate } = useViewportInterval();
    const viewport = useViewportDimensions();
    const [isDragging, setIsDragging] = React.useState(null);
    const moveDataRef = React.useRef({
        isZooming: false,
    });
    const visualMinDuration = (MIN_SELECTION_WIDTH * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
    const handleMoveEnd = () => {
        state.endDragging();
        setIsDragging(null);
    };
    const { moveProps } = useMove({
        onMoveStart: (e) => {
            setIsDragging('dragger');
            state.startDragging();
            moveDataRef.current = {
                isZooming: state.canResize && (e.ctrlKey || e.metaKey),
            };
        },
        onMoveEnd: handleMoveEnd,
        onMove: (e) => {
            if (e.deltaX === 0) {
                return;
            }
            if (moveDataRef.current.isZooming) {
                let scale = Math.exp((e.deltaX * DRAG_SCALE_STRENGTH) / viewport.width);
                scale = Math.min(Math.max(scale, MIN_DRAG_SCALE), MAX_DRAG_SCALE);
                state.scale(scale, { visualMinDuration });
            }
            else {
                let delta = (e.deltaX * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
                if (e.pointerType === 'keyboard') {
                    delta =
                        delta > 0
                            ? Math.max(delta, state.align)
                            : Math.min(delta, -1 * state.align);
                }
                state.move(delta, { visualMinDuration });
            }
        },
    });
    const { moveProps: moveStartProps } = useMove({
        onMoveStart: () => {
            setIsDragging('start');
            state.startDragging();
        },
        onMoveEnd: handleMoveEnd,
        onMove: (e) => {
            if (e.deltaX === 0) {
                return;
            }
            let delta = (e.deltaX * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
            if (e.pointerType === 'keyboard') {
                delta =
                    delta > 0 ? Math.max(delta, state.align) : Math.min(delta, -1 * state.align);
            }
            state.moveStart(delta, { visualMinDuration });
        },
    });
    const { moveProps: moveEndProps } = useMove({
        onMoveStart: () => {
            setIsDragging('end');
            state.startDragging();
        },
        onMoveEnd: handleMoveEnd,
        onMove: (e) => {
            if (e.deltaX === 0) {
                return;
            }
            let delta = (e.deltaX * (endDate.valueOf() - startDate.valueOf())) / viewport.width;
            if (e.pointerType === 'keyboard') {
                delta =
                    delta > 0 ? Math.max(delta, state.align) : Math.min(delta, -1 * state.align);
            }
            state.moveEnd(delta, { visualMinDuration });
        },
    });
    const { start: startValue, end: endValue } = state.value;
    const startX = lerp({ start: startDate.valueOf(), end: endDate.valueOf() }, { start: 0, end: viewport.width }, startValue.valueOf());
    const endX = lerp({ start: startDate.valueOf(), end: endDate.valueOf() }, { start: 0, end: viewport.width }, endValue.valueOf());
    const draggerInputRef = React.useRef(null);
    const startInputRef = React.useRef(null);
    const endInputRef = React.useRef(null);
    let id = React.useId();
    id = (_b = props.id) !== null && _b !== void 0 ? _b : id;
    const { t } = i18n.useTranslation();
    return (_jsxs("div", { className: b(null, className), style: Object.assign(Object.assign({}, style), { left: startX, width: endX - startX }), children: [_jsx("div", Object.assign({ className: b('dragger', {
                    dragging: isDragging === 'dragger',
                    disabled: Boolean(isDragging && isDragging !== 'dragger'),
                }) }, moveProps, { onClick: () => {
                    var _a;
                    (_a = draggerInputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
                }, children: _jsx("div", { className: b('slider-input-wrapper'), children: _jsx("input", { id: `${id}-range`, ref: draggerInputRef, className: b('slider-input'), type: "range", step: 1, "aria-label": t('Range'), "aria-labelledby": [`${id}-range`, props['aria-labelledby']]
                            .filter(Boolean)
                            .join(' '), "aria-orientation": "horizontal", "aria-valuetext": `${startValue.format()} - ${endValue.format()}`, value: 0, onChange: () => { } }) }) })), state.canResize ? (
            /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
            _jsx("div", Object.assign({ className: b('start', {
                    dragging: isDragging === 'start',
                    disabled: Boolean(isDragging && isDragging !== 'start'),
                }) }, moveStartProps, { onClick: () => {
                    var _a;
                    (_a = startInputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
                }, children: _jsx("div", { className: b('slider-input-wrapper'), children: _jsx("input", { id: `${id}-start`, ref: startInputRef, className: b('slider-input'), type: "range", step: 1, "aria-label": t('Start of range'), "aria-labelledby": [`${id}-start`, props['aria-labelledby']]
                            .filter(Boolean)
                            .join(' '), "aria-orientation": "horizontal", "aria-valuetext": `${startValue.format()}`, value: 0, onChange: () => { } }) }) }))) : null, state.canResize ? (
            /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
            _jsx("div", Object.assign({ className: b('end', {
                    dragging: isDragging === 'end',
                    disabled: Boolean(isDragging && isDragging !== 'end'),
                }) }, moveEndProps, { onClick: () => {
                    var _a;
                    (_a = endInputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
                }, children: _jsx("div", { className: b('slider-input-wrapper'), children: _jsx("input", { id: `${id}-end`, ref: endInputRef, className: b('slider-input'), type: "range", step: 1, "aria-label": t('End of range'), "aria-labelledby": [`${id}-end`, props['aria-labelledby']]
                            .filter(Boolean)
                            .join(' '), "aria-orientation": "horizontal", "aria-valuetext": `${endValue.format()}`, value: 0, onChange: () => { } }) }) }))) : null] }));
}
