// @gravity-ui/date-components/dist/cjs/components/RelativeDatePicker/hooks/useRelativeDatePickerProps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRelativeDatePickerProps = useRelativeDatePickerProps;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const uikit_1 = require("@gravity-ui/uikit");
const DateField_1 = require("../../DateField/index.js");
const utils_1 = require("../../DatePicker/utils/index.js");
const RelativeDateField_1 = require("../../RelativeDateField/index.js");
const filterDOMProps_1 = require("../../utils/filterDOMProps.js");
const getButtonSizeForInput_1 = require("../../utils/getButtonSizeForInput.js");
const mergeProps_1 = require("../../utils/mergeProps.js");
const i18n_1 = require("../i18n/index.js");
function useRelativeDatePickerProps(state, _a) {
    var { onFocus, onBlur } = _a, props = tslib_1.__rest(_a, ["onFocus", "onBlur"]);
    const { mode, setMode, datePickerState, relativeDateState } = state;
    const [focusedDate, setFocusedDate] = react_1.default.useState(mode === 'relative'
        ? relativeDateState.lastCorrectDate
        : datePickerState.dateFieldState.displayValue);
    const [prevCorrectDate, setPrevCorrectDate] = react_1.default.useState(relativeDateState.lastCorrectDate);
    if (prevCorrectDate !== relativeDateState.lastCorrectDate) {
        setPrevCorrectDate(relativeDateState.lastCorrectDate);
        setFocusedDate(relativeDateState.lastCorrectDate);
    }
    const [prevDateValue, setPrevDateValue] = react_1.default.useState(datePickerState.dateFieldState.displayValue);
    if (!datePickerState.dateFieldState.displayValue.isSame(prevDateValue, 'day')) {
        setPrevDateValue(datePickerState.dateFieldState.displayValue);
        setFocusedDate(datePickerState.dateFieldState.displayValue);
    }
    const { focusWithinProps } = (0, uikit_1.useFocusWithin)({
        onFocusWithin: onFocus,
        onBlurWithin: onBlur,
        onFocusWithinChange(isFocusWithin) {
            if (!isFocusWithin) {
                state.setActive(false);
            }
        },
    });
    const [isOpen, setOpen] = (0, uikit_1.useControlledState)(undefined, false, props.onOpenChange);
    if (!state.isActive && isOpen) {
        setOpen(false);
    }
    const commonInputProps = {
        onFocus: () => {
            if (!state.isActive) {
                state.setActive(true);
                setOpen(true);
            }
        },
        errorMessage: props.errorMessage,
        errorPlacement: props.errorPlacement,
        controlProps: {
            onClick: () => {
                if (state.disabled) {
                    return;
                }
                if (!isOpen) {
                    state.setActive(true);
                    setOpen(true);
                }
            },
            role: 'combobox',
            'aria-expanded': isOpen,
        },
    };
    const { inputProps } = (0, DateField_1.useDateFieldProps)(datePickerState.dateFieldState, Object.assign(Object.assign({}, props), { value: undefined, defaultValue: undefined, onUpdate: undefined }));
    const { inputProps: relativeDateProps } = (0, RelativeDateField_1.useRelativeDateFieldProps)(relativeDateState, Object.assign(Object.assign({}, props), { value: undefined, defaultValue: undefined, onUpdate: undefined }));
    let validationState = props.validationState;
    if (!validationState) {
        validationState =
            mode === 'relative'
                ? relativeDateState.validationState
                : datePickerState.dateFieldState.validationState;
    }
    const inputRef = react_1.default.useRef(null);
    const handleRef = (0, uikit_1.useForkRef)(inputRef, mode === 'relative' ? relativeDateProps.controlRef : inputProps.controlRef);
    const calendarRef = react_1.default.useRef(null);
    function focusCalendar() {
        setTimeout(() => {
            var _a;
            (_a = calendarRef.current) === null || _a === void 0 ? void 0 : _a.focus();
        });
    }
    function focusInput() {
        setTimeout(() => {
            var _a;
            (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus({ preventScroll: true });
        });
    }
    const groupRef = react_1.default.useRef(null);
    const calendarModes = (0, utils_1.getCalendarModes)(datePickerState.formatInfo);
    const { t } = i18n_1.i18n.useTranslation();
    const DOMProps = (0, filterDOMProps_1.filterDOMProps)(props);
    delete DOMProps.id;
    return {
        groupProps: Object.assign(Object.assign(Object.assign(Object.assign({}, DOMProps), { ref: groupRef, tabIndex: -1, role: 'group' }), focusWithinProps), { onKeyDown: (e) => {
                if (e.altKey && (e.key === 'ArrowDown' || e.key === 'ArrowUp')) {
                    e.preventDefault();
                    e.stopPropagation();
                    setOpen(true);
                    focusCalendar();
                }
            } }),
        fieldProps: (0, mergeProps_1.mergeProps)(commonInputProps, mode === 'relative' ? relativeDateProps : inputProps, mode === 'absolute' &&
            datePickerState.dateFieldState.isEmpty &&
            !state.isActive &&
            props.placeholder
            ? { value: '' }
            : undefined, { controlRef: handleRef, validationState }),
        modeSwitcherProps: {
            size: (0, getButtonSizeForInput_1.getButtonSizeForInput)(props.size),
            disabled: state.readOnly || state.disabled,
            view: 'flat-secondary',
            style: { zIndex: 2, marginInlineEnd: 2 },
            selected: mode === 'relative',
            'aria-label': t('Formula input mode'),
            onClick: () => {
                setMode(mode === 'relative' ? 'absolute' : 'relative');
                if (mode === 'relative') {
                    const valueDate = datePickerState.value;
                    if (valueDate) {
                        setFocusedDate(valueDate);
                    }
                }
                else if (relativeDateState.parsedDate) {
                    setFocusedDate(relativeDateState.parsedDate);
                }
                focusInput();
            },
        },
        calendarButtonProps: {
            size: (0, getButtonSizeForInput_1.getButtonSizeForInput)(props.size),
            disabled: state.disabled,
            'aria-label': t('Calendar'),
            'aria-haspopup': 'dialog',
            'aria-expanded': isOpen,
            view: 'flat-secondary',
            onClick: () => {
                state.setActive(true);
                setOpen(!isOpen);
                if (!isOpen) {
                    focusCalendar();
                }
            },
        },
        popupProps: {
            open: isOpen,
            onEscapeKeyDown: () => {
                setOpen(false);
                focusInput();
            },
            onOutsideClick: (e) => {
                var _a;
                if (e.target && !((_a = groupRef.current) === null || _a === void 0 ? void 0 : _a.contains(e.target))) {
                    setOpen(false);
                }
            },
            onTransitionOutComplete: () => {
                setFocusedDate(mode === 'relative'
                    ? relativeDateState.lastCorrectDate
                    : datePickerState.dateFieldState.displayValue);
            },
            placement: props.popupPlacement,
            offset: props.popupOffset,
            className: props.popupClassName,
            style: props.popupStyle,
            returnFocus: false,
        },
        calendarProps: {
            ref: calendarRef,
            size: props.size === 's' ? 'm' : props.size,
            readOnly: props.readOnly,
            value: state.selectedDate,
            onUpdate: (v) => {
                datePickerState.setDateValue(v);
                if (!state.datePickerState.formatInfo.hasTime) {
                    setOpen(false);
                    focusInput();
                }
            },
            focusedValue: focusedDate,
            onFocusUpdate: setFocusedDate,
            minValue: props.minValue,
            maxValue: props.maxValue,
            modes: mode === 'absolute' ? calendarModes : undefined,
        },
        timeInputProps: {
            value: datePickerState.timeValue,
            onUpdate: datePickerState.setTimeValue,
            format: datePickerState.timeFormat,
            readOnly: state.readOnly,
            disabled: state.disabled,
            timeZone: props.timeZone,
            hasClear: props.hasClear,
            size: props.size,
        },
    };
}
