// @gravity-ui/table/build/esm/components/BaseRow/BaseRow.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { useForkRef } from '@gravity-ui/uikit';
import { BaseCell } from "../BaseCell/index.js";
import { BaseGroupHeader } from "../BaseGroupHeader/index.js";
import { b } from "../BaseTable/BaseTable.classname.js";
export const BaseRow = React.forwardRef(({ cellClassName, className: classNameProp, getGroupTitle, getIsCustomRow, getIsGroupHeaderRow, groupHeaderClassName, onClick, renderCustomRowContent, renderGroupHeader, renderGroupHeaderRowContent, row, rowVirtualizer, style, virtualItem, attributes: attributesProp, cellAttributes, table: _, ...restProps }, ref) => {
    const rowRef = useForkRef(rowVirtualizer?.measureElement, ref);
    const attributes = typeof attributesProp === 'function' ? attributesProp(row) : attributesProp;
    const className = typeof classNameProp === 'function' ? classNameProp(row) : classNameProp;
    const handleClick = React.useCallback((event) => {
        onClick?.(row, event);
    }, [onClick, row]);
    const renderRowContent = () => {
        if (getIsGroupHeaderRow?.(row)) {
            return renderGroupHeaderRowContent ? (renderGroupHeaderRowContent({
                row,
                Cell: BaseCell,
                cellClassName,
                getGroupTitle,
            })) : (_jsx(BaseCell, { className: cellClassName, colSpan: row.getVisibleCells().length, attributes: cellAttributes, "aria-colindex": 1, children: renderGroupHeader ? (renderGroupHeader({
                    row,
                    className: b('group-header', groupHeaderClassName),
                    getGroupTitle,
                })) : (_jsx(BaseGroupHeader, { row: row, className: b('group-header', groupHeaderClassName), getGroupTitle: getGroupTitle })) }));
        }
        if (getIsCustomRow?.(row) && renderCustomRowContent) {
            return renderCustomRowContent({ row, Cell: BaseCell, cellClassName });
        }
        return row
            .getVisibleCells()
            .map((cell) => (_jsx(BaseCell, { cell: cell, className: cellClassName, attributes: cellAttributes, "aria-colindex": cell.column.getIndex() + 1 }, cell.id)));
    };
    return (_jsx("tr", { ref: rowRef, className: b('row', {
            selected: row.getIsSelected(),
            interactive: Boolean(onClick),
        }, className), onClick: handleClick, "data-index": virtualItem?.index, ...restProps, ...attributes, style: {
            top: rowVirtualizer && virtualItem
                ? virtualItem.start - rowVirtualizer.options.scrollMargin
                : undefined,
            ...style,
            ...attributes?.style,
        }, children: renderRowContent() }));
});
BaseRow.displayName = 'BaseRow';
