// @gravity-ui/date-components/dist/cjs/components/utils/dates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isWeekend = isWeekend;
exports.createPlaceholderValue = createPlaceholderValue;
exports.isInvalid = isInvalid;
exports.constrainValue = constrainValue;
exports.mergeDateTime = mergeDateTime;
exports.formatDateTime = formatDateTime;
const date_utils_1 = require("@gravity-ui/date-utils");
function isWeekend(date) {
    return [0, 6].includes(date.day());
}
function createPlaceholderValue({ placeholderValue, timeZone }) {
    return placeholderValue !== null && placeholderValue !== void 0 ? placeholderValue : (0, date_utils_1.dateTime)({ timeZone }).startOf('day');
}
function isInvalid(value, minValue, maxValue) {
    if (!value) {
        return false;
    }
    if (minValue && value.isBefore(minValue)) {
        return true;
    }
    if (maxValue && maxValue.isBefore(value)) {
        return true;
    }
    return false;
}
function constrainValue(value, minValue, maxValue) {
    if (minValue && value.isBefore(minValue)) {
        return minValue;
    }
    if (maxValue && maxValue.isBefore(value)) {
        return maxValue;
    }
    return value;
}
function mergeDateTime(date, time) {
    return date
        .set('hours', time.hour())
        .set('minutes', time.minute())
        .set('seconds', time.second());
}
function formatDateTime(date, format, timezone, lang) {
    return (0, date_utils_1.dateTime)({ input: date, timeZone: timezone, lang }).format(format);
}
