// @gravity-ui/date-components/dist/esm/components/DateField/hooks/useBaseDateFieldState.js
import React from 'react';
import { createPlaceholderValue } from "../../utils/dates.js";
import { EDITABLE_SEGMENTS, formatSections, getCurrentEditableSectionIndex, getDurationUnitFromSectionType, } from "../utils.js";
const PAGE_STEP = {
    year: 5,
    quarter: 2,
    month: 2,
    weekday: 3,
    day: 7,
    hour: 2,
    minute: 15,
    second: 15,
};
export function useBaseDateFieldState(props) {
    const { value, validationState, displayValue, editableSections, formatInfo, selectedSectionIndexes, selectedSections, isEmpty, flushAllValidSections, flushValidSection, setSelectedSections, setValue, setDate, adjustSection, setSection, getSectionValue, setSectionValue, createPlaceholder, setValueFromString, } = props;
    const enteredKeys = React.useRef('');
    return {
        value,
        isEmpty,
        displayValue,
        setValue,
        setDate,
        text: formatSections(editableSections),
        readOnly: props.readOnly,
        disabled: props.disabled,
        sections: editableSections,
        formatInfo,
        hasDate: formatInfo.hasDate,
        hasTime: formatInfo.hasTime,
        selectedSectionIndexes,
        validationState,
        setSelectedSections(position) {
            enteredKeys.current = '';
            setSelectedSections(position);
        },
        focusSectionInPosition(position) {
            const nextSectionIndex = this.sections.findIndex((s) => s.end >= position);
            const index = nextSectionIndex === -1 ? 0 : nextSectionIndex;
            const nextSection = this.sections[index];
            if (nextSection) {
                this.setSelectedSections(EDITABLE_SEGMENTS[nextSection.type] ? index : nextSection.nextEditableSection);
            }
        },
        focusNextSection() {
            var _a, _b;
            const currentIndex = selectedSections === 'all' ? 0 : selectedSections;
            const newIndex = (_b = (_a = this.sections[currentIndex]) === null || _a === void 0 ? void 0 : _a.nextEditableSection) !== null && _b !== void 0 ? _b : -1;
            if (newIndex !== -1) {
                this.setSelectedSections(newIndex);
            }
        },
        focusPreviousSection() {
            var _a, _b;
            const currentIndex = selectedSections === 'all' ? 0 : selectedSections;
            const newIndex = (_b = (_a = this.sections[currentIndex]) === null || _a === void 0 ? void 0 : _a.previousEditableSection) !== null && _b !== void 0 ? _b : -1;
            if (newIndex !== -1) {
                this.setSelectedSections(newIndex);
            }
        },
        focusFirstSection() {
            var _a, _b;
            const newIndex = (_b = (_a = this.sections[0]) === null || _a === void 0 ? void 0 : _a.previousEditableSection) !== null && _b !== void 0 ? _b : -1;
            if (newIndex !== -1) {
                setSelectedSections(newIndex);
            }
        },
        focusLastSection() {
            var _a, _b;
            const newIndex = (_b = (_a = this.sections[this.sections.length - 1]) === null || _a === void 0 ? void 0 : _a.nextEditableSection) !== null && _b !== void 0 ? _b : -1;
            if (newIndex !== -1) {
                this.setSelectedSections(newIndex);
            }
        },
        increment() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex !== -1) {
                adjustSection(sectionIndex, 1);
            }
        },
        decrement() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex !== -1) {
                adjustSection(sectionIndex, -1);
            }
        },
        incrementPage() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex !== -1) {
                adjustSection(sectionIndex, PAGE_STEP[this.sections[sectionIndex].type] || 1);
            }
        },
        decrementPage() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex !== -1) {
                adjustSection(sectionIndex, -(PAGE_STEP[this.sections[sectionIndex].type] || 1));
            }
        },
        incrementToMax() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex !== -1) {
                const section = this.sections[sectionIndex];
                if (typeof section.maxValue === 'number') {
                    setSection(sectionIndex, section.maxValue);
                }
            }
        },
        decrementToMin() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex !== -1) {
                const section = this.sections[sectionIndex];
                if (typeof section.minValue === 'number') {
                    setSection(sectionIndex, section.minValue);
                }
            }
        },
        clearSection() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            if (selectedSections === 'all') {
                this.clearAll();
                return;
            }
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex === -1) {
                return;
            }
            flushValidSection(sectionIndex);
            const section = this.sections[sectionIndex];
            const placeholder = createPlaceholderValue({
                placeholderValue: props.placeholderValue,
                timeZone: props.timeZone,
            }).timeZone(props.timeZone);
            const displayPortion = getSectionValue(sectionIndex);
            let currentValue = displayPortion;
            // Reset day period to default without changing the hour.
            if (section.type === 'dayPeriod') {
                const isPM = displayPortion.hour() >= 12;
                const shouldBePM = placeholder.hour() >= 12;
                if (isPM && !shouldBePM) {
                    currentValue = displayPortion.set('hour', displayPortion.hour() - 12);
                }
                else if (!isPM && shouldBePM) {
                    currentValue = displayPortion.set('hour', displayPortion.hour() + 12);
                }
            }
            else {
                const type = getDurationUnitFromSectionType(section.type);
                currentValue = displayPortion.set(type, placeholder[type]());
            }
            setSectionValue(sectionIndex, currentValue);
        },
        clearAll() {
            if (this.readOnly || this.disabled) {
                return;
            }
            enteredKeys.current = '';
            flushAllValidSections();
            if (value !== null) {
                setDate(null);
            }
            const date = createPlaceholder();
            setValue(date);
        },
        onInput(key) {
            if (this.readOnly || this.disabled) {
                return;
            }
            const sectionIndex = getCurrentEditableSectionIndex(this.sections, selectedSections);
            if (sectionIndex === -1) {
                return;
            }
            const section = this.sections[sectionIndex];
            const isLastSection = section.nextEditableSection === sectionIndex;
            let newValue = enteredKeys.current + key;
            const onInputNumber = (numberValue) => {
                var _a;
                let sectionValue = section.type === 'month' ? numberValue - 1 : numberValue;
                const sectionMaxValue = (_a = section.maxValue) !== null && _a !== void 0 ? _a : 0;
                const allowsZero = section.minValue === 0;
                let shouldResetUserInput;
                if (
                // 12-hour clock format with AM/PM
                section.type === 'hour' &&
                    (sectionMaxValue === 11 || section.minValue === 12)) {
                    if (sectionValue > 12) {
                        sectionValue = Number(key);
                        newValue = key;
                    }
                    if (sectionValue === 0) {
                        sectionValue = -Infinity;
                    }
                    if (sectionValue === 12) {
                        sectionValue = 0;
                    }
                    if (section.minValue === 12) {
                        sectionValue += 12;
                    }
                    shouldResetUserInput = Number(newValue + '0') > 12;
                }
                else if (sectionValue > sectionMaxValue) {
                    sectionValue = Number(key) - (section.type === 'month' ? 1 : 0);
                    newValue = key;
                    if (sectionValue > sectionMaxValue) {
                        enteredKeys.current = '';
                        return;
                    }
                }
                const shouldSetValue = sectionValue > 0 || (sectionValue === 0 && allowsZero);
                if (shouldSetValue) {
                    setSection(sectionIndex, sectionValue);
                }
                if (shouldResetUserInput === undefined) {
                    shouldResetUserInput = Number(newValue + '0') > sectionMaxValue;
                }
                const isMaxLength = newValue.length >= String(sectionMaxValue).length;
                if (shouldResetUserInput) {
                    enteredKeys.current = '';
                    if (shouldSetValue) {
                        this.focusNextSection();
                    }
                }
                else if (isMaxLength && shouldSetValue && !isLastSection) {
                    this.focusNextSection();
                }
                else {
                    enteredKeys.current = newValue;
                }
            };
            const onInputString = (stringValue) => {
                var _a;
                const options = (_a = section.options) !== null && _a !== void 0 ? _a : [];
                let sectionValue = stringValue.toLocaleUpperCase();
                let foundOptions = options.filter((v) => v.startsWith(sectionValue));
                if (foundOptions.length === 0) {
                    if (stringValue !== key) {
                        sectionValue = key.toLocaleUpperCase();
                        foundOptions = options.filter((v) => v.startsWith(sectionValue));
                    }
                    if (foundOptions.length === 0) {
                        enteredKeys.current = '';
                        return;
                    }
                }
                const foundValue = foundOptions[0];
                const index = options.indexOf(foundValue);
                if (section.type === 'dayPeriod') {
                    setSection(sectionIndex, index === 1 ? 12 : 0);
                }
                else {
                    setSection(sectionIndex, index);
                }
                if (foundOptions.length > 1) {
                    enteredKeys.current = newValue;
                }
                else {
                    enteredKeys.current = '';
                    this.focusNextSection();
                }
            };
            switch (section.type) {
                case 'day':
                case 'hour':
                case 'minute':
                case 'second':
                case 'quarter':
                case 'year': {
                    if (!Number.isInteger(Number(newValue))) {
                        return;
                    }
                    const numberValue = Number(newValue);
                    onInputNumber(numberValue);
                    break;
                }
                case 'dayPeriod': {
                    onInputString(newValue);
                    break;
                }
                case 'weekday':
                case 'month': {
                    if (Number.isInteger(Number(newValue))) {
                        const numberValue = Number(newValue);
                        onInputNumber(numberValue);
                    }
                    else {
                        onInputString(newValue);
                    }
                    break;
                }
            }
        },
        setValueFromString(str) {
            enteredKeys.current = '';
            return setValueFromString(str);
        },
    };
}
