// @gravity-ui/date-components/dist/cjs/components/RangeDateSelection/components/Ruler/Ruler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateTimeRuler = DateTimeRuler;
exports.useViewportDimensions = useViewportDimensions;
exports.useViewportInterval = useViewportInterval;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = tslib_1.__importDefault(require("react"));
const react_dom_1 = tslib_1.__importDefault(require("react-dom"));
const useResizeObserver_1 = require("../../../../hooks/useResizeObserver.js");
const cn_1 = require("../../../../utils/cn.js");
const NowLine_1 = require("../NowLine/NowLine.js");
const RulerViewport_1 = require("../RulerViewport/RulerViewport.js");
const MiddleTicks_1 = require("../Ticks/MiddleTicks.js");
const SlitTicks_1 = require("../Ticks/SlitTicks.js");
const UnavialableTicks_1 = require("../Ticks/UnavialableTicks.js");
const utils_1 = require("../Ticks/utils.js");
const b = (0, cn_1.block)('ruler');
const viewportDimensionsContext = react_1.default.createContext(null);
const viewportIntervalContext = react_1.default.createContext(null);
function DateTimeRuler({ className, children, start, end, minValue, maxValue, onMove, onMoveStart, onMoveEnd, displayNow, formatTime, timeZone, dragDisabled, renderAdditionalRulerContent, }) {
    const [container, setContainer] = react_1.default.useState(null);
    const viewportInterval = react_1.default.useMemo(() => ({ start, end }), [start, end]);
    const [viewportDimensions, setViewportDimensions] = react_1.default.useState({
        width: 0,
        height: 0,
    });
    const onResize = react_1.default.useCallback(() => {
        if (!container) {
            return;
        }
        const { width, height } = container.getBoundingClientRect();
        setViewportDimensions({ width, height });
    }, [container]);
    const containerRef = react_1.default.useMemo(() => ({ current: container }), [container]);
    (0, useResizeObserver_1.useResizeObserver)({ ref: containerRef, onResize });
    const hasConstrains = Boolean(minValue || maxValue);
    return ((0, jsx_runtime_1.jsx)("div", { className: b(null, className), children: (0, jsx_runtime_1.jsx)(viewportIntervalContext.Provider, { value: viewportInterval, children: (0, jsx_runtime_1.jsxs)(viewportDimensionsContext.Provider, { value: viewportDimensions, children: [(0, jsx_runtime_1.jsxs)(RulerViewport_1.RulerViewport, { ref: setContainer, onMove: onMove, onMoveStart: onMoveStart, onMoveEnd: onMoveEnd, dragDisabled: dragDisabled, children: [(0, jsx_runtime_1.jsx)(MiddleTicks_1.MiddleTicks, { timeZone: timeZone }), (0, jsx_runtime_1.jsx)(SlitTicks_1.SlitTicks, { formatTime: formatTime, timeZone: timeZone }), hasConstrains ? ((0, jsx_runtime_1.jsx)(UnavialableTicks_1.UnavailableTicks, { theme: "dim", minValue: minValue, maxValue: maxValue, geometry: (0, utils_1.makeUnavailableTicksGeometry)({
                                    tickHeight: 10,
                                    viewportHeight: viewportDimensions.height,
                                }), tickWidth: viewportDimensions.height + 10 })) : null, displayNow ? (0, jsx_runtime_1.jsx)(NowLine_1.NowLine, {}) : null, renderAdditionalRulerContent === null || renderAdditionalRulerContent === void 0 ? void 0 : renderAdditionalRulerContent({
                                interval: viewportInterval,
                                dimensions: viewportDimensions,
                            })] }), container ? react_dom_1.default.createPortal(children, container) : null] }) }) }));
}
function useViewportDimensions() {
    const viewport = react_1.default.useContext(viewportDimensionsContext);
    if (!viewport) {
        throw new Error('useViewportDimensions must be used within a RulerViewport');
    }
    return viewport;
}
function useViewportInterval() {
    const viewport = react_1.default.useContext(viewportIntervalContext);
    if (!viewport) {
        throw new Error('useViewportInterval must be used within a RulerViewport');
    }
    return viewport;
}
