// @gravity-ui/table/build/esm/components/BaseTable/BaseTable.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { getAriaMultiselectable, getAriaRowIndexMap, shouldRenderFooterRow } from "../../utils/index.js";
import { BaseDraggableRow } from "../BaseDraggableRow/index.js";
import { BaseFooterRow } from "../BaseFooterRow/index.js";
import { BaseHeaderRow } from "../BaseHeaderRow/index.js";
import { BaseRow } from "../BaseRow/index.js";
import { LastSelectedRowContextProvider } from "../LastSelectedRowContext/index.js";
import { SortableListContext } from "../SortableListContext/index.js";
import { b } from "./BaseTable.classname.js";
import "./BaseTable.css";
export const BaseTable = React.forwardRef(({ table, attributes, bodyAttributes, bodyClassName, bodyRef, cellAttributes, cellClassName, className, customFooterRowCount, emptyContent, footerAttributes, footerCellAttributes, footerCellClassName, footerClassName, footerRowAttributes, footerRowClassName, getGroupTitle, getIsCustomRow, getIsGroupHeaderRow, groupHeaderClassName, headerAttributes, headerCellAttributes, headerCellClassName, headerClassName, headerRowAttributes, headerRowClassName, onRowClick, renderCustomFooterContent, renderCustomRowContent, renderGroupHeader, renderGroupHeaderRowContent, renderHeaderCellContent, renderResizeHandle, renderSortIndicator, resizeHandleClassName, rowAttributes, rowClassName, rowVirtualizer, sortIndicatorClassName, stickyFooter = false, stickyHeader = false, withFooter = false, withHeader = true, }, ref) => {
    const draggableContext = React.useContext(SortableListContext);
    const draggingRowIndex = draggableContext?.activeItemIndex ?? -1;
    const { rows, rowsById } = table.getRowModel();
    const ariaRowIndexMap = React.useMemo(() => {
        return getAriaRowIndexMap(rows);
    }, [rows]);
    const headerGroups = withHeader ? table.getHeaderGroups() : [];
    const footerGroups = withFooter ? table.getFooterGroups() : [];
    const colCount = table.getVisibleLeafColumns().length;
    const headerRowCount = headerGroups.length;
    const bodyRowCount = Object.keys(rowsById).length;
    const footerRowCount = (withFooter &&
        ((renderCustomFooterContent && customFooterRowCount) || footerGroups.length)) ||
        0;
    const rowCount = bodyRowCount + headerRowCount + footerRowCount;
    const bodyRows = rowVirtualizer?.getVirtualItems() || rows;
    const renderEmptyContent = () => {
        if (!emptyContent) {
            return null;
        }
        const emptyRowClassName = typeof rowClassName === 'function' ? rowClassName() : rowClassName;
        const emptyCellClassName = typeof cellClassName === 'function' ? cellClassName() : cellClassName;
        return (_jsx("tr", { className: b('row', { empty: true }, emptyRowClassName), children: _jsx("td", { className: b('cell', {}, emptyCellClassName), colSpan: colCount, style: {
                    width: rowVirtualizer ? table.getTotalSize() : undefined,
                }, children: typeof emptyContent === 'function' ? emptyContent() : emptyContent }) }));
    };
    const renderBodyRows = () => {
        return bodyRows.map((virtualItemOrRow, index) => {
            const row = rowVirtualizer
                ? rows[virtualItemOrRow.index]
                : virtualItemOrRow;
            const rowIndex = rowVirtualizer ? virtualItemOrRow.index : index;
            const style = row.depth > 0
                ? {
                    '--_--tree-depth': row.depth,
                    '--_--last-nested': rows[rowIndex + 1]?.depth === 0 ? 1 : 0,
                }
                : undefined;
            const virtualItem = rowVirtualizer ? virtualItemOrRow : undefined;
            const key = virtualItem?.key ?? row.id;
            const rowProps = {
                cellClassName,
                className: rowClassName,
                getGroupTitle,
                getIsCustomRow,
                getIsGroupHeaderRow,
                groupHeaderClassName,
                attributes: rowAttributes,
                cellAttributes,
                onClick: onRowClick,
                renderCustomRowContent,
                renderGroupHeader,
                renderGroupHeaderRowContent,
                row,
                rowVirtualizer,
                table,
                virtualItem,
                style,
                'aria-rowindex': headerRowCount + ariaRowIndexMap[row.id],
                'aria-selected': table.options.enableRowSelection
                    ? row.getIsSelected()
                    : undefined,
            };
            if (draggableContext) {
                return _jsx(BaseDraggableRow, { ...rowProps }, key);
            }
            return _jsx(BaseRow, { ...rowProps }, key);
        });
    };
    return (_jsx(LastSelectedRowContextProvider, { children: _jsxs("table", { ref: ref, className: b({ 'with-row-virtualization': Boolean(rowVirtualizer) }, className), "data-dragging-row-index": draggingRowIndex > -1 ? draggingRowIndex : undefined, "aria-colcount": colCount > 0 ? colCount : undefined, "aria-rowcount": rowCount > 0 ? rowCount : undefined, "aria-multiselectable": getAriaMultiselectable(table), ...attributes, children: [withHeader && (_jsx("thead", { className: b('header', { sticky: stickyHeader }, headerClassName), ...headerAttributes, children: headerGroups.map((headerGroup, index) => (_jsx(BaseHeaderRow, { cellClassName: headerCellClassName, className: headerRowClassName, headerGroup: headerGroup, parentHeaderGroup: headerGroups[index - 1], renderHeaderCellContent: renderHeaderCellContent, renderResizeHandle: renderResizeHandle, renderSortIndicator: renderSortIndicator, resizeHandleClassName: resizeHandleClassName, sortIndicatorClassName: sortIndicatorClassName, attributes: headerRowAttributes, cellAttributes: headerCellAttributes, "aria-rowindex": index + 1 }, headerGroup.id))) })), _jsx("tbody", { ref: bodyRef, className: b('body', bodyClassName), ...bodyAttributes, style: {
                        height: bodyRows.length ? rowVirtualizer?.getTotalSize() : undefined,
                        ...bodyAttributes?.style,
                    }, children: bodyRows.length ? renderBodyRows() : renderEmptyContent() }), withFooter && (_jsx("tfoot", { className: b('footer', { sticky: stickyFooter }, footerClassName), ...footerAttributes, children: renderCustomFooterContent
                        ? renderCustomFooterContent({
                            cellClassName: b('footer-cell'),
                            footerGroups,
                            rowClassName: b('footer-row'),
                            rowIndex: headerRowCount + bodyRowCount + 1,
                        })
                        : footerGroups.map((footerGroup, index) => shouldRenderFooterRow(footerGroup) ? (_jsx(BaseFooterRow, { footerGroup: footerGroup, attributes: footerRowAttributes, cellAttributes: footerCellAttributes, cellClassName: footerCellClassName, className: footerRowClassName, "aria-rowindex": headerRowCount + bodyRowCount + index + 1 }, footerGroup.id)) : null) }))] }) }));
});
BaseTable.displayName = 'BaseTable';
