// @gravity-ui/date-components/dist/esm/components/RangeDatePicker/hooks/useRangeDatePickerState.js
import { adjustDateToFormat } from "../../DateField/utils.js";
import { datePickerStateFactory } from "../../DatePicker/hooks/datePickerStateFactory.js";
import { getDateTimeValue } from "../../DatePicker/utils/index.js";
import { useRangeDateFieldState } from "../../RangeDateField/index.js";
import { createPlaceholderRangeValue } from "../../RangeDateField/utils/index.js";
import { mergeDateTime } from "../../utils/dates.js";
export const useRangeDatePickerState = datePickerStateFactory({
    getPlaceholderTime,
    mergeDateTime: mergeRangeDateTime,
    setTimezone,
    getDateTime: getDateTimeValue,
    useDateFieldState: useRangeDateFieldState,
    adjustDateToFormat: adjustRangeToFormat,
});
function getPlaceholderTime(placeholderValue, timeZone) {
    return createPlaceholderRangeValue({ placeholderValue, timeZone });
}
function mergeRangeDateTime(date, time) {
    const start = mergeDateTime(date.start, time.start);
    const end = mergeDateTime(date.end, time.end);
    return { start, end };
}
function setTimezone(date, timeZone) {
    const start = date.start.timeZone(timeZone);
    const end = date.end.timeZone(timeZone);
    return { start, end };
}
function adjustRangeToFormat(date, sectionsInfo) {
    const start = adjustDateToFormat(date.start, sectionsInfo, 'startOf');
    const end = adjustDateToFormat(date.end, sectionsInfo, 'endOf');
    return { start, end };
}
