// @gravity-ui/table/build/cjs/hooks/useSortableList.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSortableList = void 0;
const React = __importStar(require("react"));
const core_1 = require("@dnd-kit/core");
const useSortableList = ({ items, onDragStart, onDragEnd, enableNesting, childModeOffset = 20, nextChildModeOffset = 10, }) => {
    const [activeItemKey, setActiveItemKey] = React.useState(null);
    const [targetItemIndex, setTargetItemIndex] = React.useState(-1);
    const [isParentMode, setIsParentMode] = React.useState(false);
    const [isChildMode, setIsChildMode] = React.useState(false);
    const [isNextChildMode, setIsNextChildMode] = React.useState(false);
    const itemIndexMap = React.useMemo(() => {
        const map = new Map();
        items.forEach((item, index) => {
            map.set(item, index);
        });
        return map;
    }, [items]);
    const getItemIndex = (item) => {
        if (typeof item !== 'undefined' && item !== null) {
            return itemIndexMap.get(item) ?? -1;
        }
        return -1;
    };
    const resetState = () => {
        setActiveItemKey(null);
        setTargetItemIndex(-1);
        setIsParentMode(false);
        setIsChildMode(false);
        setIsNextChildMode(false);
    };
    const getTargetItemKey = (activeId, overId) => {
        if (overId) {
            const overItemIndex = getItemIndex(overId);
            const activeItemIndex = getItemIndex(activeId);
            if (overItemIndex <= activeItemIndex) {
                return items[overItemIndex - 1];
            }
        }
        return overId;
    };
    (0, core_1.useDndMonitor)({
        onDragStart: (event) => {
            setActiveItemKey(event.active.id);
            onDragStart?.(event.active.id);
            document.body.style.setProperty('cursor', 'grabbing');
        },
        onDragMove: (event) => {
            if (enableNesting) {
                setIsParentMode(event.delta.x < -childModeOffset);
                setIsChildMode(event.delta.x > childModeOffset);
                setIsNextChildMode(event.delta.x > nextChildModeOffset && event.delta.x <= childModeOffset);
            }
        },
        onDragOver: (event) => {
            const targetItemKey = getTargetItemKey(event.active.id, event.over?.id);
            setTargetItemIndex(targetItemKey && targetItemKey !== event.active.id
                ? getItemIndex(targetItemKey)
                : -1);
        },
        onDragEnd: (event) => {
            document.body.style.setProperty('cursor', 'default');
            if (!event.over) {
                resetState();
                return;
            }
            const draggedItemKey = event.active.id;
            const targetItemKey = getTargetItemKey(draggedItemKey, event.over.id);
            if (targetItemKey === draggedItemKey) {
                resetState();
                return;
            }
            if (isChildMode) {
                onDragEnd?.({
                    draggedItemKey,
                    targetItemKey,
                    enableNesting,
                });
            }
            else {
                onDragEnd?.({
                    draggedItemKey,
                    baseItemKey: targetItemKey,
                    baseNextItemKey: items[targetItemKey ? getItemIndex(targetItemKey) + 1 : 0],
                    enableNesting,
                    nextChild: isNextChildMode,
                    pullFromParent: isParentMode,
                });
            }
            resetState();
        },
        onDragCancel: resetState,
    });
    return {
        activeItemKey,
        activeItemIndex: getItemIndex(activeItemKey),
        targetItemIndex,
        isParentMode,
        isChildMode,
        isNextChildMode,
    };
};
exports.useSortableList = useSortableList;
