// @gravity-ui/table/build/cjs/hooks/useTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTable = void 0;
const react_table_1 = require("@tanstack/react-table");
const useTable = (options) => {
    const tableOptions = {
        ...options,
        enableColumnPinning: options.enableColumnPinning ?? false,
        enableColumnResizing: options.enableColumnResizing ?? false,
        enableExpanding: options.enableExpanding ?? false,
        enableGrouping: options.enableGrouping ?? false,
        enableMultiRowSelection: options.enableMultiRowSelection ?? false,
        enableRowSelection: options.enableRowSelection ?? false,
        enableSorting: options.enableSorting ?? false,
        getCoreRowModel: options.getCoreRowModel ?? (0, react_table_1.getCoreRowModel)(),
        getExpandedRowModel: options.enableExpanding
            ? (options.getExpandedRowModel ?? (0, react_table_1.getExpandedRowModel)())
            : undefined,
        getGroupedRowModel: options.enableGrouping
            ? (options.getGroupedRowModel ?? (0, react_table_1.getGroupedRowModel)())
            : undefined,
        getSortedRowModel: options.enableSorting
            ? (options.getSortedRowModel ?? (0, react_table_1.getSortedRowModel)())
            : undefined,
        manualGrouping: options.manualGrouping ?? false,
        manualSorting: options.manualSorting ?? false,
    };
    return (0, react_table_1.useReactTable)(tableOptions);
};
exports.useTable = useTable;
