// @gravity-ui/table/build/cjs/hooks/useTableSettings.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTableSettings = void 0;
const React = __importStar(require("react"));
const useTableSettings = ({ initialOrdering = [], initialVisibility = {}, onVisibilityChange, onOrderingChange } = {
    initialOrdering: [],
    initialVisibility: {},
}) => {
    const [visibilityState, setVisibilityState] = React.useState(initialVisibility);
    const [orderingState, setOrderState] = React.useState(initialOrdering);
    const onColumnVisibilityChange = React.useCallback((value) => {
        setVisibilityState(value);
        if (onVisibilityChange) {
            const newValue = typeof value === 'function' ? value(visibilityState) : value;
            onVisibilityChange(newValue);
        }
    }, [onVisibilityChange, visibilityState]);
    const onColumnOrderChange = React.useCallback((value) => {
        setOrderState(value);
        if (onOrderingChange) {
            const newValue = typeof value === 'function' ? value(orderingState) : value;
            onOrderingChange(newValue);
        }
    }, [onOrderingChange, orderingState]);
    return {
        state: {
            columnVisibility: visibilityState,
            columnOrder: orderingState,
        },
        callbacks: {
            onColumnVisibilityChange,
            onColumnOrderChange,
        },
    };
};
exports.useTableSettings = useTableSettings;
