// @gravity-ui/date-components/dist/esm/components/RelativeRangeDatePicker/RelativeRangeDatePicker.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { useControlledState, useFocusWithin, useMobile } from '@gravity-ui/uikit';
import { block } from "../../utils/cn.js";
import { HiddenInput } from "../HiddenInput/HiddenInput.js";
import { filterDOMProps } from "../utils/filterDOMProps.js";
import { Control } from "./components/Control/Control.js";
import { PickerDialog } from "./components/PickerDialog/PickerDialog.js";
import { useRelativeRangeDatePickerState } from "./hooks/useRelativeRangeDatePickerState.js";
import "./RelativeRangeDatePicker.css";
const b = block('relative-range-date-picker');
export function RelativeRangeDatePicker(props) {
    const state = useRelativeRangeDatePickerState(props);
    const isMobile = useMobile();
    const [anchor, setAnchor] = React.useState(null);
    const dialogClosing = React.useRef(false);
    const [open, setOpen] = useControlledState(undefined, false, props.onOpenChange);
    const { focusWithinProps } = useFocusWithin({
        isDisabled: props.disabled,
        onFocusWithin: props.onFocus,
        onBlurWithin: props.onBlur,
        onFocusWithinChange: (isFocusWithin) => {
            var _a;
            if (isFocusWithin &&
                !dialogClosing.current &&
                !props.renderControl &&
                ((_a = document.activeElement) === null || _a === void 0 ? void 0 : _a.tagName) !== 'BUTTON') {
                setOpen(true);
            }
        },
    });
    const DOMProps = filterDOMProps(props);
    delete DOMProps.id;
    return (_jsxs("div", Object.assign({}, DOMProps, { ref: setAnchor }, focusWithinProps, { className: b(null, props.className), style: props.style, children: [_jsx(Control, { props: props, state: state, open: open, setOpen: setOpen, isMobile: isMobile, onClick: () => {
                    if (props.disabled) {
                        return;
                    }
                    setOpen(true);
                }, onKeyDown: (e) => {
                    if (props.disabled) {
                        return;
                    }
                    if (e.altKey && (e.key === 'ArrowDown' || e.key === 'ArrowUp')) {
                        e.preventDefault();
                        setOpen(true);
                    }
                }, onClickCalendar: () => {
                    setOpen(!open);
                }, onUpdate: (v) => {
                    if (!props.readOnly && !v) {
                        state.setValue(null, 'default');
                    }
                } }), _jsx(HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a, _b; return (_b = (_a = v === null || v === void 0 ? void 0 : v.start) === null || _a === void 0 ? void 0 : _a.type) !== null && _b !== void 0 ? _b : ''; }, disabled: props.disabled }), _jsx(HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a; return getNativeValue((_a = v === null || v === void 0 ? void 0 : v.start) !== null && _a !== void 0 ? _a : null); }, disabled: props.disabled }), _jsx(HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a, _b; return (_b = (_a = v === null || v === void 0 ? void 0 : v.end) === null || _a === void 0 ? void 0 : _a.type) !== null && _b !== void 0 ? _b : ''; }, disabled: props.disabled }), _jsx(HiddenInput, { name: props.name, form: props.form, value: state.value, toStringValue: (v) => { var _a; return getNativeValue((_a = v === null || v === void 0 ? void 0 : v.end) !== null && _a !== void 0 ? _a : null); }, disabled: props.disabled }), _jsx(HiddenInput, { name: props.name, form: props.form, onReset: (v) => {
                    state.setValue(v.value, v.timeZone);
                }, value: { value: state.value, timeZone: state.timeZone }, toStringValue: (v) => v.timeZone, disabled: props.disabled }), _jsx(PickerDialog, { value: state.value, onUpdate: state.setValue, timeZone: state.timeZone, open: open, onClose: () => {
                    setOpen(false);
                    dialogClosing.current = true;
                }, onRemove: () => {
                    setTimeout(() => {
                        dialogClosing.current = false;
                    });
                }, anchor: anchor, modal: true, returnFocus: true, popupClassName: props.popupClassName, popupStyle: props.popupStyle, popupPlacement: props.popupPlacement, popupOffset: props.popupOffset, isMobile: isMobile, size: props.size, format: props.format, readOnly: props.readOnly, minValue: props.minValue, maxValue: props.maxValue, allowNullableValues: props.allowNullableValues, isDateUnavailable: props.isDateUnavailable, placeholderValue: props.placeholderValue, withPresets: props.withPresets, applyPresetsImmediately: props.applyPresetsImmediately, presetTabs: props.presetTabs, docs: props.docs, withApplyButton: props.withApplyButton, withZonesList: props.withZonesList, withHeader: props.withHeader })] })));
}
function getNativeValue(value) {
    if (!value) {
        return '';
    }
    if (value.type === 'relative') {
        return value.value;
    }
    return value.value.toISOString();
}
