// @gravity-ui/date-components/dist/cjs/hooks/useMove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useMove = useMove;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const useEventHandler_1 = require("./useEventHandler.js");
function useMove(props) {
    const { onMoveStart, onMove, onMoveEnd } = props;
    const state = react_1.default.useRef({
        isMoving: false,
        lastPosition: null,
        id: null,
    });
    const move = (0, useEventHandler_1.useEventHandler)((originalEvent, deltaX, deltaY, pointerType) => {
        if (deltaX === 0 && deltaY === 0) {
            return;
        }
        if (!state.current.isMoving) {
            state.current.isMoving = true;
            onMoveStart === null || onMoveStart === void 0 ? void 0 : onMoveStart({
                type: 'movestart',
                pointerType,
                shiftKey: originalEvent.shiftKey,
                ctrlKey: originalEvent.ctrlKey,
                metaKey: originalEvent.metaKey,
                altKey: originalEvent.altKey,
            });
        }
        onMove === null || onMove === void 0 ? void 0 : onMove({
            type: 'move',
            pointerType,
            deltaX,
            deltaY,
            shiftKey: originalEvent.shiftKey,
            ctrlKey: originalEvent.ctrlKey,
            metaKey: originalEvent.metaKey,
            altKey: originalEvent.altKey,
        });
    });
    const end = (0, useEventHandler_1.useEventHandler)((originalEvent, pointerType) => {
        if (state.current.isMoving) {
            state.current.isMoving = false;
            onMoveEnd === null || onMoveEnd === void 0 ? void 0 : onMoveEnd({
                type: 'moveend',
                pointerType,
                shiftKey: originalEvent.shiftKey,
                ctrlKey: originalEvent.ctrlKey,
                metaKey: originalEvent.metaKey,
                altKey: originalEvent.altKey,
            });
        }
    });
    return react_1.default.useMemo(() => {
        const moveProps = {};
        if (typeof PointerEvent === 'undefined') {
            moveProps.onMouseDown = (e) => {
                if (e.button === 0) {
                    state.current.isMoving = false;
                    e.preventDefault();
                    state.current.lastPosition = { pageX: e.pageX, pageY: e.pageY };
                    const handleMouseMove = (ev) => {
                        var _a, _b;
                        if (ev.buttons === 0) {
                            move(ev, ev.pageX - (((_a = state.current.lastPosition) === null || _a === void 0 ? void 0 : _a.pageX) || 0), ev.pageY - (((_b = state.current.lastPosition) === null || _b === void 0 ? void 0 : _b.pageY) || 0), 'mouse');
                            state.current.lastPosition = { pageX: ev.pageX, pageY: ev.pageY };
                        }
                    };
                    const handleMouseUp = (ev) => {
                        if (ev.button === 0) {
                            end(ev, 'mouse');
                            window.removeEventListener('mousemove', handleMouseMove);
                            window.removeEventListener('mouseup', handleMouseUp);
                        }
                    };
                    window.addEventListener('mousemove', handleMouseMove);
                    window.addEventListener('mouseup', handleMouseUp);
                }
            };
            moveProps.onTouchStart = (e) => {
                if (e.touches.length === 0 || state.current.id !== null) {
                    return;
                }
                const { pageX, pageY, identifier } = e.touches[0];
                state.current.id = identifier;
                state.current.lastPosition = { pageX, pageY };
                e.preventDefault();
                const handleTouchMove = (ev) => {
                    var _a, _b;
                    const touch = [...ev.changedTouches].findIndex((el) => el.identifier === state.current.id);
                    if (touch !== -1) {
                        const { pageX, pageY } = e.changedTouches[touch];
                        move(ev, pageX - (((_a = state.current.lastPosition) === null || _a === void 0 ? void 0 : _a.pageX) || 0), pageY - (((_b = state.current.lastPosition) === null || _b === void 0 ? void 0 : _b.pageY) || 0), 'touch');
                        state.current.lastPosition = { pageX, pageY };
                    }
                };
                const handleTouchEnd = (ev) => {
                    const touch = [...ev.changedTouches].findIndex((el) => el.identifier === state.current.id);
                    if (touch !== -1) {
                        end(ev, 'touch');
                        state.current.id = null;
                        window.removeEventListener('touchmove', handleTouchMove);
                        window.removeEventListener('touchend', handleTouchEnd);
                        window.removeEventListener('touchcancel', handleTouchEnd);
                    }
                };
                window.addEventListener('touchmove', handleTouchMove);
                window.addEventListener('touchend', handleTouchEnd);
                window.addEventListener('touchcancel', handleTouchEnd);
            };
        }
        else {
            moveProps.onPointerDown = (e) => {
                if (e.button === 0 && state.current.id === null) {
                    e.preventDefault();
                    state.current.isMoving = false;
                    state.current.id = e.pointerId;
                    state.current.lastPosition = { pageX: e.pageX, pageY: e.pageY };
                    const handlePointerMove = (ev) => {
                        var _a, _b;
                        if (ev.pointerId === state.current.id) {
                            move(ev, ev.pageX - (((_a = state.current.lastPosition) === null || _a === void 0 ? void 0 : _a.pageX) || 0), ev.pageY - (((_b = state.current.lastPosition) === null || _b === void 0 ? void 0 : _b.pageY) || 0), (ev.pointerType || 'mouse'));
                            state.current.lastPosition = { pageX: ev.pageX, pageY: ev.pageY };
                        }
                    };
                    const handlePointerUp = (ev) => {
                        if (ev.pointerId === state.current.id) {
                            end(ev, (ev.pointerType || 'mouse'));
                            state.current.id = null;
                            window.removeEventListener('pointermove', handlePointerMove);
                            window.removeEventListener('pointerup', handlePointerUp);
                            window.removeEventListener('pointercancel', handlePointerUp);
                        }
                    };
                    window.addEventListener('pointermove', handlePointerMove);
                    window.addEventListener('pointerup', handlePointerUp);
                    window.addEventListener('pointercancel', handlePointerUp);
                }
            };
        }
        const handleKeyboardMove = (e, deltaX, deltaY) => {
            state.current.isMoving = false;
            move(e, deltaX, deltaY, 'keyboard');
            end(e, 'keyboard');
        };
        moveProps.onKeyDown = (e) => {
            switch (e.key) {
                case 'ArrowLeft': {
                    e.preventDefault();
                    handleKeyboardMove(e, -1, 0);
                    break;
                }
                case 'ArrowRight': {
                    e.preventDefault();
                    handleKeyboardMove(e, 1, 0);
                    break;
                }
                case 'ArrowUp': {
                    e.preventDefault();
                    handleKeyboardMove(e, 0, -1);
                    break;
                }
                case 'ArrowDown': {
                    e.preventDefault();
                    handleKeyboardMove(e, 0, 1);
                    break;
                }
            }
        };
        return { moveProps };
    }, [end, move]);
}
